// @siemens/ix/components/ix-modal-loading.js
import { p as proxyCustomElement, H, h, d as Host } from './p-D-__gyet.js';
import { d as defineCustomElement$2 } from './p-Ds0M0G_e.js';

const modalLoadingCss = ":host{display:flex;justify-content:flex-start;align-items:center;gap:0.5rem;overflow:hidden}:host .loading-text{display:block;position:relative;overflow:hidden;white-space:nowrap;text-overflow:ellipsis}";

const ModalLoading = /*@__PURE__*/ proxyCustomElement(class ModalLoading extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
    }
    render() {
        return (h(Host, { key: 'd323b6a4a1650629e0ac5338c97fd7b08929807f' }, h("ix-spinner", { key: '94eab908aa232f0f022da10da6313cbe07388a3c', variant: "primary" }), h("span", { key: 'a1d7732187718f8775d626e0a0be49fb7578bec6', class: 'loading-text' }, h("slot", { key: 'bdbc66e91cc7f9ada60a3cda2f0347b40237d739' }))));
    }
    static get style() { return modalLoadingCss; }
}, [257, "ix-modal-loading"]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-modal-loading", "ix-spinner"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-modal-loading":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModalLoading);
            }
            break;
        case "ix-spinner":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}

const IxModalLoading = ModalLoading;
const defineCustomElement = defineCustomElement$1;

export { IxModalLoading, defineCustomElement };
