// @siemens/ix/components/p-CwfaZReW.js
import { p as proxyCustomElement, H, h, d as Host } from './p-D-__gyet.js';

const rowCss = ":host{display:flex;flex-wrap:wrap}:host(:not(:first-of-type)){margin-block-start:var(--ix-layout-grid-row-margin, 0)}";

const Row = /*@__PURE__*/ proxyCustomElement(class Row extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
    }
    render() {
        return (h(Host, { key: 'edd5d38945ebc6f0b97cc6cf3ae0a9230f7dc2f2' }, h("slot", { key: 'd2556a5cc69da0d7fc5f0779e8453761d45a1de0' })));
    }
    static get style() { return rowCss; }
}, [257, "ix-row"]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-row"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-row":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, Row);
            }
            break;
    } });
}

export { Row as R, defineCustomElement as d };
