// @siemens/ix/components/p-Djmkeupf.js
import { p as proxyCustomElement, H, e as createEvent, h, d as Host } from './p-D-__gyet.js';
import { f as flip, i as inline, s as shift, o as offset, a as autoUpdate, c as computePosition } from './p-CAqtPJ4-.js';
import { a as addDisposableEventListener } from './p-CKoABG1h.js';
import { f as findElement } from './p-CFRrPFxi.js';
import { m as makeRef } from './p-bcj7UEIC.js';
import { r as requestAnimationFrameNoNgZone } from './p-BEuV0Xpe.js';
import { d as dropdownController, h as hasDropdownItemWrapperImplemented } from './p-D3K9vmFp.js';

/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
class ArrowFocusController {
    constructor(items, container, callback) {
        this.wrap = false;
        this.items = items;
        this.container = container;
        this.callback = callback;
        this.keyListener = addDisposableEventListener(container, 'keydown', (e) => this.onKeyDown(e));
    }
    getActiveIndex() {
        if (!document.activeElement) {
            return -1;
        }
        return this.items.indexOf(document.activeElement);
    }
    onKeyDown(e) {
        var _a, _b, _c, _d;
        const activeIndex = this.getActiveIndex();
        if (activeIndex < 0) {
            return;
        }
        switch (e.key) {
            case 'ArrowDown':
                if (activeIndex < this.items.length - 1) {
                    e.preventDefault();
                    (_a = this.callback) === null || _a === void 0 ? void 0 : _a.call(this, activeIndex + 1);
                }
                else if (this.wrap) {
                    e.preventDefault();
                    (_b = this.callback) === null || _b === void 0 ? void 0 : _b.call(this, 0);
                }
                break;
            case 'ArrowUp':
                {
                    if (activeIndex > 0) {
                        e.preventDefault();
                        (_c = this.callback) === null || _c === void 0 ? void 0 : _c.call(this, activeIndex - 1);
                    }
                    else if (this.wrap && activeIndex === 0) {
                        e.preventDefault();
                        (_d = this.callback) === null || _d === void 0 ? void 0 : _d.call(this, this.items.length - 1);
                    }
                }
                break;
        }
    }
    disconnect() {
        if (this.keyListener) {
            this.keyListener();
        }
        this.container = undefined;
        this.callback = undefined;
    }
}

const dropdownCss = ":host{background-color:var(--theme-color-2);border-radius:var(--theme-default-border-radius);min-width:0px;z-index:var(--theme-z-index-dropdown);box-shadow:var(--theme-shadow-4);padding:0.25rem 0px}:host *,:host *::after,:host *::before{box-sizing:border-box}:host ::-webkit-scrollbar-button{display:none}@-moz-document url-prefix(){:host *{scrollbar-color:var(--theme-scrollbar-thumb--background) var(--theme-scrollbar-track--background);scrollbar-width:thin}}:host{}:host ::-webkit-scrollbar{width:0.5rem;height:0.5rem}:host{}:host ::-webkit-scrollbar-track{border-radius:5px;background:var(--theme-scrollbar-track--background)}:host ::-webkit-scrollbar-track:hover{background:var(--theme-scrollbar-track--background--hover)}:host{}:host ::-webkit-scrollbar-thumb{border-radius:5px;background:var(--theme-scrollbar-thumb--background)}:host{}:host ::-webkit-scrollbar-thumb:hover{background:var(--theme-scrollbar-thumb--background--hover)}:host ::-webkit-scrollbar-corner{display:none}:host .dropdown-header{display:flex;align-items:center;height:2.5rem;color:var(--theme-menu-header--color);padding:0 1rem}:host(.overflow){max-height:calc(50vh - 3rem);overflow-y:auto}:host(:not(.show)){display:none}.dialog{margin:0;border:none;outline:none;padding:0.25rem 0px;min-width:0;max-width:100vw;width:-moz-max-content;width:max-content;height:-moz-fit-content;height:fit-content;background-color:var(--theme-color-2);border-radius:var(--theme-default-border-radius);box-shadow:var(--theme-shadow-4);overflow-x:visible;overflow-y:visible;inset:unset;color-scheme:inherit;color:var(--theme-color-std-text);box-sizing:border-box}.dialog *,.dialog *::after,.dialog *::before{box-sizing:border-box}.dialog .dropdown-container{display:block;position:relative}.dialog .dropdown-header{display:flex;align-items:center;height:2.5rem;color:var(--theme-menu-header--color);padding:0 1rem}.dialog.overflow{max-height:calc(50vh - 3rem);overflow-y:auto}";

let sequenceId = 0;
const Dropdown = /*@__PURE__*/ proxyCustomElement(class Dropdown extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
        this.showChanged = createEvent(this, "showChanged", 7);
        /**
         * Suppress the automatic placement of the dropdown.
         */
        this.suppressAutomaticPlacement = false;
        /**
         * Show dropdown
         */
        this.show = false;
        /**
         * Controls if the dropdown will be closed in response to a click event depending on the position of the event relative to the dropdown.
         * If the dropdown is a child of another one, it will be closed with the parent, regardless of its own close behavior.
         */
        this.closeBehavior = 'both';
        /**
         * Placement of the dropdown
         */
        this.placement = 'bottom-start';
        /**
         * Position strategy
         */
        this.positioningStrategy = 'fixed';
        /**
         * @internal
         * If initialization of this dropdown is expected to be deferred submenu discovery will have to be re-run globally by the controller.
         * This property indicates the need for that to the controller.
         */
        this.discoverAllSubmenus = false;
        /** @internal */
        this.ignoreRelatedSubmenu = false;
        /** @internal */
        this.suppressOverflowBehavior = false;
        /**
         * Enable Popover API rendering for top-layer positioning.
         *
         * @default false in v4.x, will default to true in v5.0.0
         */
        this.enableTopLayer = false;
        this.dialogRef = makeRef();
        this.localUId = `dropdown-${sequenceId++}`;
        this.assignedSubmenu = [];
        this.itemObserver = new MutationObserver(() => {
            if (this.arrowFocusController) {
                this.arrowFocusController.items = this.dropdownItems;
            }
        });
    }
    connectedCallback() {
        dropdownController.connected(this);
        if (this.trigger != undefined) {
            this.registerListener(this.trigger);
        }
    }
    cacheSubmenuId(event) {
        event.stopImmediatePropagation();
        event.preventDefault();
        const { detail } = event;
        if (this.assignedSubmenu.indexOf(detail) === -1) {
            this.assignedSubmenu.push(detail);
        }
    }
    disconnectedCallback() {
        var _a;
        dropdownController.dismiss(this);
        dropdownController.disconnected(this);
        if (this.arrowFocusController) {
            (_a = this.arrowFocusController) === null || _a === void 0 ? void 0 : _a.disconnect();
            this.arrowFocusController = undefined;
        }
        if (this.itemObserver) {
            this.itemObserver.disconnect();
            this.itemObserver = undefined;
        }
        if (this.disposeClickListener) {
            this.disposeClickListener();
            this.disposeClickListener = undefined;
        }
        if (this.disposeKeyListener) {
            this.disposeKeyListener();
            this.disposeKeyListener = undefined;
        }
        if (this.autoUpdateCleanup) {
            this.autoUpdateCleanup();
            this.autoUpdateCleanup = undefined;
        }
    }
    getAssignedSubmenuIds() {
        return this.assignedSubmenu;
    }
    isPresent() {
        return this.show;
    }
    present() {
        this.show = true;
    }
    dismiss() {
        this.show = false;
    }
    getId() {
        return this.localUId;
    }
    willDismiss() {
        const { defaultPrevented } = this.showChanged.emit(false);
        return !defaultPrevented;
    }
    willPresent() {
        const { defaultPrevented } = this.showChanged.emit(true);
        return !defaultPrevented;
    }
    get dropdownItems() {
        return Array.from(this.hostElement.querySelectorAll('ix-dropdown-item'));
    }
    get slotElement() {
        return this.hostElement.shadowRoot.querySelector('slot');
    }
    addEventListenersFor() {
        var _a, _b, _c;
        (_a = this.disposeClickListener) === null || _a === void 0 ? void 0 : _a.call(this);
        (_b = this.disposeKeyListener) === null || _b === void 0 ? void 0 : _b.call(this);
        const toggleController = () => {
            if (!this.isPresent()) {
                dropdownController.present(this);
            }
            else {
                dropdownController.dismiss(this);
            }
            dropdownController.dismissOthers(this.getId());
        };
        if (!this.triggerElement) {
            return;
        }
        this.disposeClickListener = addDisposableEventListener(this.triggerElement, 'click', (event) => {
            if (!event.defaultPrevented) {
                toggleController();
            }
        });
        (_c = this.triggerElement) === null || _c === void 0 ? void 0 : _c.setAttribute('data-ix-dropdown-trigger', this.localUId);
    }
    /** @internal */
    async discoverSubmenu() {
        var _a;
        (_a = this.triggerElement) === null || _a === void 0 ? void 0 : _a.dispatchEvent(new CustomEvent('ix-assign-sub-menu', {
            bubbles: true,
            composed: true,
            cancelable: true,
            detail: this.localUId,
        }));
    }
    registerKeyListener() {
        if (!this.triggerElement) {
            return;
        }
        this.disposeKeyListener = addDisposableEventListener(this.triggerElement, 'keydown', ((event) => {
            if (event.key !== 'ArrowDown') {
                return;
            }
            if (document.activeElement !== this.triggerElement) {
                return;
            }
            dropdownController.present(this);
            setTimeout(() => {
                this.focusDropdownItem(0);
            });
        }));
    }
    async registerListener(element) {
        this.triggerElement = await this.resolveElement(element);
        if (this.triggerElement) {
            this.addEventListenersFor();
            this.discoverSubmenu();
        }
    }
    async resolveElement(element) {
        const el = await findElement(element);
        return this.checkForSubmenuAnchor(el);
    }
    async checkForSubmenuAnchor(element) {
        if (!element) {
            return undefined;
        }
        if (hasDropdownItemWrapperImplemented(element)) {
            const dropdownItem = await element.getDropdownItemElement();
            dropdownItem.isSubMenu = true;
            this.hostElement.style.zIndex = `var(--theme-z-index-dropdown)`;
        }
        if (element.tagName === 'IX-DROPDOWN-ITEM') {
            element.isSubMenu = true;
            this.hostElement.style.zIndex = `var(--theme-z-index-dropdown)`;
        }
        return element;
    }
    async resolveAnchorElement() {
        if (this.anchor) {
            this.anchorElement = await this.resolveElement(this.anchor);
        }
        else if (this.trigger) {
            this.anchorElement = await this.resolveElement(this.trigger);
        }
    }
    async changedShow(newShow) {
        var _a;
        if (!newShow) {
            this.cleanupOnHide();
            if (this.enableTopLayer) {
                await this.hideDropdownAsync();
            }
            return;
        }
        this.arrowFocusController = new ArrowFocusController(this.dropdownItems, this.hostElement, (index) => this.focusDropdownItem(index));
        (_a = this.itemObserver) === null || _a === void 0 ? void 0 : _a.observe(this.hostElement, {
            childList: true,
            subtree: true,
        });
        this.registerKeyListener();
        if (this.enableTopLayer) {
            await this.showDropdownAsync();
            return;
        }
        await this.resolveAnchorElement();
        if (this.anchorElement) {
            this.applyDropdownPosition();
        }
    }
    changedTrigger(newTriggerValue) {
        this.registerListener(newTriggerValue);
    }
    destroyAutoUpdate() {
        if (this.autoUpdateCleanup) {
            this.autoUpdateCleanup();
            this.autoUpdateCleanup = undefined;
        }
    }
    isAnchorSubmenu() {
        var _a;
        if (!hasDropdownItemWrapperImplemented(this.anchorElement)) {
            // Is no official dropdown-item, but check if any dropdown-item
            // is placed somewhere up the DOM
            return !!((_a = this.anchorElement) === null || _a === void 0 ? void 0 : _a.closest('ix-dropdown-item'));
        }
        return true;
    }
    async showDropdownAsync() {
        const popover = await this.dialogRef.waitForCurrent();
        if (!popover) {
            return;
        }
        popover.showPopover();
        await this.resolveAnchorElement();
        if (this.anchorElement) {
            this.applyDropdownPosition();
        }
        else {
            requestAnimationFrameNoNgZone(() => {
                const referenceElement = this.hostElement.parentElement || this.hostElement;
                const refRect = referenceElement.getBoundingClientRect();
                const transform = `translate(${Math.round(refRect.left)}px, ${Math.round(refRect.top)}px)`;
                Object.assign(popover.style, {
                    top: '0',
                    left: '0',
                    transform,
                });
            });
        }
    }
    async hideDropdownAsync() {
        const popover = await this.dialogRef.waitForCurrent();
        if (popover === null || popover === void 0 ? void 0 : popover.matches(':popover-open')) {
            popover.hidePopover();
        }
    }
    cleanupOnHide() {
        var _a, _b, _c;
        this.destroyAutoUpdate();
        (_a = this.arrowFocusController) === null || _a === void 0 ? void 0 : _a.disconnect();
        (_b = this.itemObserver) === null || _b === void 0 ? void 0 : _b.disconnect();
        (_c = this.disposeKeyListener) === null || _c === void 0 ? void 0 : _c.call(this);
    }
    async applyDropdownPosition() {
        var _a, _b;
        if (!this.show) {
            return;
        }
        if (!this.anchorElement) {
            return;
        }
        const referenceElement = this.anchorElement;
        const isSubmenu = this.isAnchorSubmenu();
        let targetElement = this.hostElement;
        let strategy = this.positioningStrategy;
        if (this.enableTopLayer) {
            const dialog = await this.dialogRef.waitForCurrent();
            if (!dialog) {
                return;
            }
            targetElement = dialog;
            strategy = 'fixed';
        }
        let positionConfig = {
            strategy,
            middleware: [],
        };
        if (!this.suppressAutomaticPlacement) {
            (_a = positionConfig.middleware) === null || _a === void 0 ? void 0 : _a.push(flip({ fallbackStrategy: 'initialPlacement' }));
        }
        positionConfig.placement = isSubmenu ? 'right-start' : this.placement;
        positionConfig.middleware = [
            ...(((_b = positionConfig.middleware) === null || _b === void 0 ? void 0 : _b.filter(Boolean)) || []),
            inline(),
            shift(),
        ];
        if (this.offset) {
            positionConfig.middleware.push(offset(this.offset));
        }
        this.destroyAutoUpdate();
        this.autoUpdateCleanup = autoUpdate(referenceElement, targetElement, async () => {
            const computeResponse = await computePosition(referenceElement, targetElement, positionConfig);
            Object.assign(targetElement.style, {
                top: '0',
                left: '0',
                transform: `translate(${Math.round(computeResponse.x)}px,${Math.round(computeResponse.y)}px)`,
            });
            if (this.overwriteDropdownStyle) {
                const overwriteStyle = await this.overwriteDropdownStyle({
                    dropdownRef: targetElement,
                    triggerRef: this.triggerElement,
                });
                Object.assign(targetElement.style, overwriteStyle);
            }
        }, {
            ancestorResize: true,
            ancestorScroll: true,
            elementResize: true,
        });
    }
    focusDropdownItem(index) {
        requestAnimationFrameNoNgZone(() => {
            var _a, _b;
            const button = (_b = (_a = this.dropdownItems[index]) === null || _a === void 0 ? void 0 : _a.shadowRoot) === null || _b === void 0 ? void 0 : _b.querySelector('button');
            if (button) {
                button.focus();
            }
        });
    }
    async componentDidLoad() {
        if (this.show) {
            this.changedShow(true);
        }
        if (!this.trigger) {
            return;
        }
        this.changedTrigger(this.trigger);
    }
    async componentDidRender() {
        await this.applyDropdownPosition();
        await this.resolveAnchorElement();
    }
    isTriggerElement(element) {
        const trigger = !!element.hasAttribute('data-ix-dropdown-trigger');
        return trigger;
    }
    onDropdownClick(event) {
        const target = dropdownController.pathIncludesTrigger(event.composedPath());
        if (target) {
            if (target !== this.triggerElement) {
                event.preventDefault();
            }
            if (this.isTriggerElement(target)) {
                if (this.closeBehavior === 'outside') {
                    event.preventDefault();
                }
                return;
            }
        }
        if (!event.defaultPrevented &&
            (this.closeBehavior === 'inside' || this.closeBehavior === 'both')) {
            dropdownController.dismissAll([this.getId()], this.ignoreRelatedSubmenu);
            return;
        }
        dropdownController.dismissOthers(this.getId());
    }
    /**
     * Update position of dropdown
     */
    async updatePosition() {
        this.applyDropdownPosition();
    }
    render() {
        if (this.enableTopLayer) {
            return (h(Host, { "data-ix-dropdown": this.localUId, class: {
                    'dropdown-menu': true,
                    show: this.show,
                } }, h("dialog", { ref: this.dialogRef, class: {
                    dialog: true,
                    overflow: !this.suppressOverflowBehavior,
                }, popover: "manual", "aria-modal": "true", "aria-label": this.header, tabindex: -1, onClick: (event) => this.onDropdownClick(event) }, h("div", { class: "dropdown-container" }, this.header && h("div", { class: "dropdown-header" }, this.header), this.show && h("slot", null)))));
        }
        return (h(Host, { "data-ix-dropdown": this.localUId, class: {
                'dropdown-menu': true,
                show: this.show,
                overflow: !this.suppressOverflowBehavior,
            }, style: {
                margin: '0',
                minWidth: '0px',
                position: this.positioningStrategy,
            }, role: "list", onClick: (event) => this.onDropdownClick(event) }, h("div", { style: { display: 'contents' } }, this.header && h("div", { class: "dropdown-header" }, this.header), this.show && h("slot", null))));
    }
    get hostElement() { return this; }
    static get watchers() { return {
        "show": ["changedShow"],
        "trigger": ["changedTrigger"]
    }; }
    static get style() { return dropdownCss; }
}, [257, "ix-dropdown", {
        "suppressAutomaticPlacement": [4, "suppress-automatic-placement"],
        "show": [1540],
        "trigger": [1],
        "anchor": [1],
        "closeBehavior": [8, "close-behavior"],
        "placement": [1],
        "positioningStrategy": [1, "positioning-strategy"],
        "header": [1],
        "offset": [16],
        "overwriteDropdownStyle": [16, "overwrite-dropdown-style"],
        "discoverAllSubmenus": [4, "discover-all-submenus"],
        "ignoreRelatedSubmenu": [4, "ignore-related-submenu"],
        "suppressOverflowBehavior": [4, "suppress-overflow-behavior"],
        "enableTopLayer": [4, "enable-top-layer"],
        "discoverSubmenu": [64],
        "updatePosition": [64]
    }, [[0, "ix-assign-sub-menu", "cacheSubmenuId"]], {
        "show": ["changedShow"],
        "trigger": ["changedTrigger"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-dropdown"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-dropdown":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, Dropdown);
            }
            break;
    } });
}

export { ArrowFocusController as A, Dropdown as D, defineCustomElement as d };
