// @siemens/ix/dist/cjs/ix-dropdown.cjs.entry.js
'use strict';

var index = require('./index-C8Xo8L1k.js');
var floatingUi_dom = require('./floating-ui.dom-CMzYsRR7.js');
var disposableEventListener = require('./disposable-event-listener-BALEIeXq.js');
var findElement = require('./find-element-D7yD1Ukn.js');
var focus = require('./focus-Cl8_98b5.js');
var makeRef = require('./make-ref-BaH6_YJk.js');
var requestAnimationFrame = require('./requestAnimationFrame-B0rWQ3G-.js');
var dropdownController = require('./dropdown-controller-BLOjtc6w.js');

const dropdownCss = ":host{background-color:var(--theme-color-2);border-radius:var(--theme-default-border-radius);min-width:0px;z-index:var(--theme-z-index-dropdown);box-shadow:var(--theme-shadow-4);padding:0.25rem 0px}:host *,:host *::after,:host *::before{box-sizing:border-box}:host ::-webkit-scrollbar-button{display:none}@-moz-document url-prefix(){:host *{scrollbar-color:var(--theme-scrollbar-thumb--background) var(--theme-scrollbar-track--background);scrollbar-width:thin}}:host{}:host ::-webkit-scrollbar{width:0.5rem;height:0.5rem}:host{}:host ::-webkit-scrollbar-track{border-radius:5px;background:var(--theme-scrollbar-track--background)}:host ::-webkit-scrollbar-track:hover{background:var(--theme-scrollbar-track--background--hover)}:host{}:host ::-webkit-scrollbar-thumb{border-radius:5px;background:var(--theme-scrollbar-thumb--background)}:host{}:host ::-webkit-scrollbar-thumb:hover{background:var(--theme-scrollbar-thumb--background--hover)}:host ::-webkit-scrollbar-corner{display:none}:host .dropdown-header{display:flex;align-items:center;height:2.5rem;color:var(--theme-menu-header--color);padding:0 1rem}:host(.overflow){max-height:calc(50vh - 3rem);overflow-y:auto}:host(:not(.show)){display:none}.dialog{margin:0;border:none;outline:none;padding:0.25rem 0px;min-width:0;max-width:100vw;width:-moz-max-content;width:max-content;height:-moz-fit-content;height:fit-content;background-color:var(--theme-color-2);border-radius:var(--theme-default-border-radius);box-shadow:var(--theme-shadow-4);overflow-x:visible;overflow-y:visible;inset:unset;color-scheme:inherit;color:var(--theme-color-std-text);box-sizing:border-box}.dialog *,.dialog *::after,.dialog *::before{box-sizing:border-box}.dialog .dropdown-container{display:block;position:relative}.dialog .dropdown-header{display:flex;align-items:center;height:2.5rem;color:var(--theme-menu-header--color);padding:0 1rem}.dialog.overflow{max-height:calc(50vh - 3rem);overflow-y:auto}";

let sequenceId = 0;
const Dropdown = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.showChanged = index.createEvent(this, "showChanged", 7);
        /**
         * Suppress the automatic placement of the dropdown.
         */
        this.suppressAutomaticPlacement = false;
        /**
         * Show dropdown
         */
        this.show = false;
        /**
         * Controls if the dropdown will be closed in response to a click event depending on the position of the event relative to the dropdown.
         * If the dropdown is a child of another one, it will be closed with the parent, regardless of its own close behavior.
         */
        this.closeBehavior = 'both';
        /**
         * Placement of the dropdown
         */
        this.placement = 'bottom-start';
        /**
         * Position strategy
         */
        this.positioningStrategy = 'fixed';
        /**
         * @internal
         * If initialization of this dropdown is expected to be deferred submenu discovery will have to be re-run globally by the controller.
         * This property indicates the need for that to the controller.
         */
        this.discoverAllSubmenus = false;
        /** @internal */
        this.ignoreRelatedSubmenu = false;
        /** @internal */
        this.suppressOverflowBehavior = false;
        /**
         * Enable Popover API rendering for top-layer positioning.
         *
         * @default false in v4.x, will default to true in v5.0.0
         */
        this.enableTopLayer = false;
        this.dialogRef = makeRef.makeRef();
        this.localUId = `dropdown-${sequenceId++}`;
        this.assignedSubmenu = [];
        this.itemObserver = new MutationObserver(() => {
            if (this.arrowFocusController) {
                this.arrowFocusController.items = this.dropdownItems;
            }
        });
    }
    connectedCallback() {
        dropdownController.dropdownController.connected(this);
        if (this.trigger != undefined) {
            this.registerListener(this.trigger);
        }
    }
    cacheSubmenuId(event) {
        event.stopImmediatePropagation();
        event.preventDefault();
        const { detail } = event;
        if (this.assignedSubmenu.indexOf(detail) === -1) {
            this.assignedSubmenu.push(detail);
        }
    }
    disconnectedCallback() {
        var _a;
        dropdownController.dropdownController.dismiss(this);
        dropdownController.dropdownController.disconnected(this);
        if (this.arrowFocusController) {
            (_a = this.arrowFocusController) === null || _a === void 0 ? void 0 : _a.disconnect();
            this.arrowFocusController = undefined;
        }
        if (this.itemObserver) {
            this.itemObserver.disconnect();
            this.itemObserver = undefined;
        }
        if (this.disposeClickListener) {
            this.disposeClickListener();
            this.disposeClickListener = undefined;
        }
        if (this.disposeKeyListener) {
            this.disposeKeyListener();
            this.disposeKeyListener = undefined;
        }
        if (this.autoUpdateCleanup) {
            this.autoUpdateCleanup();
            this.autoUpdateCleanup = undefined;
        }
    }
    getAssignedSubmenuIds() {
        return this.assignedSubmenu;
    }
    isPresent() {
        return this.show;
    }
    present() {
        this.show = true;
    }
    dismiss() {
        this.show = false;
    }
    getId() {
        return this.localUId;
    }
    willDismiss() {
        const { defaultPrevented } = this.showChanged.emit(false);
        return !defaultPrevented;
    }
    willPresent() {
        const { defaultPrevented } = this.showChanged.emit(true);
        return !defaultPrevented;
    }
    get dropdownItems() {
        return Array.from(this.hostElement.querySelectorAll('ix-dropdown-item'));
    }
    get slotElement() {
        return this.hostElement.shadowRoot.querySelector('slot');
    }
    addEventListenersFor() {
        var _a, _b, _c;
        (_a = this.disposeClickListener) === null || _a === void 0 ? void 0 : _a.call(this);
        (_b = this.disposeKeyListener) === null || _b === void 0 ? void 0 : _b.call(this);
        const toggleController = () => {
            if (!this.isPresent()) {
                dropdownController.dropdownController.present(this);
            }
            else {
                dropdownController.dropdownController.dismiss(this);
            }
            dropdownController.dropdownController.dismissOthers(this.getId());
        };
        if (!this.triggerElement) {
            return;
        }
        this.disposeClickListener = disposableEventListener.addDisposableEventListener(this.triggerElement, 'click', (event) => {
            if (!event.defaultPrevented) {
                toggleController();
            }
        });
        (_c = this.triggerElement) === null || _c === void 0 ? void 0 : _c.setAttribute('data-ix-dropdown-trigger', this.localUId);
    }
    /** @internal */
    async discoverSubmenu() {
        var _a;
        (_a = this.triggerElement) === null || _a === void 0 ? void 0 : _a.dispatchEvent(new CustomEvent('ix-assign-sub-menu', {
            bubbles: true,
            composed: true,
            cancelable: true,
            detail: this.localUId,
        }));
    }
    registerKeyListener() {
        if (!this.triggerElement) {
            return;
        }
        this.disposeKeyListener = disposableEventListener.addDisposableEventListener(this.triggerElement, 'keydown', ((event) => {
            if (event.key !== 'ArrowDown') {
                return;
            }
            if (document.activeElement !== this.triggerElement) {
                return;
            }
            dropdownController.dropdownController.present(this);
            setTimeout(() => {
                this.focusDropdownItem(0);
            });
        }));
    }
    async registerListener(element) {
        this.triggerElement = await this.resolveElement(element);
        if (this.triggerElement) {
            this.addEventListenersFor();
            this.discoverSubmenu();
        }
    }
    async resolveElement(element) {
        const el = await findElement.findElement(element);
        return this.checkForSubmenuAnchor(el);
    }
    async checkForSubmenuAnchor(element) {
        if (!element) {
            return undefined;
        }
        if (dropdownController.hasDropdownItemWrapperImplemented(element)) {
            const dropdownItem = await element.getDropdownItemElement();
            dropdownItem.isSubMenu = true;
            this.hostElement.style.zIndex = `var(--theme-z-index-dropdown)`;
        }
        if (element.tagName === 'IX-DROPDOWN-ITEM') {
            element.isSubMenu = true;
            this.hostElement.style.zIndex = `var(--theme-z-index-dropdown)`;
        }
        return element;
    }
    async resolveAnchorElement() {
        if (this.anchor) {
            this.anchorElement = await this.resolveElement(this.anchor);
        }
        else if (this.trigger) {
            this.anchorElement = await this.resolveElement(this.trigger);
        }
    }
    async changedShow(newShow) {
        var _a;
        if (!newShow) {
            this.cleanupOnHide();
            if (this.enableTopLayer) {
                await this.hideDropdownAsync();
            }
            return;
        }
        this.arrowFocusController = new focus.ArrowFocusController(this.dropdownItems, this.hostElement, (index) => this.focusDropdownItem(index));
        (_a = this.itemObserver) === null || _a === void 0 ? void 0 : _a.observe(this.hostElement, {
            childList: true,
            subtree: true,
        });
        this.registerKeyListener();
        if (this.enableTopLayer) {
            await this.showDropdownAsync();
            return;
        }
        await this.resolveAnchorElement();
        if (this.anchorElement) {
            this.applyDropdownPosition();
        }
    }
    changedTrigger(newTriggerValue) {
        this.registerListener(newTriggerValue);
    }
    destroyAutoUpdate() {
        if (this.autoUpdateCleanup) {
            this.autoUpdateCleanup();
            this.autoUpdateCleanup = undefined;
        }
    }
    isAnchorSubmenu() {
        var _a;
        if (!dropdownController.hasDropdownItemWrapperImplemented(this.anchorElement)) {
            // Is no official dropdown-item, but check if any dropdown-item
            // is placed somewhere up the DOM
            return !!((_a = this.anchorElement) === null || _a === void 0 ? void 0 : _a.closest('ix-dropdown-item'));
        }
        return true;
    }
    async showDropdownAsync() {
        const popover = await this.dialogRef.waitForCurrent();
        if (!popover) {
            return;
        }
        popover.showPopover();
        await this.resolveAnchorElement();
        if (this.anchorElement) {
            this.applyDropdownPosition();
        }
        else {
            requestAnimationFrame.requestAnimationFrameNoNgZone(() => {
                const referenceElement = this.hostElement.parentElement || this.hostElement;
                const refRect = referenceElement.getBoundingClientRect();
                const transform = `translate(${Math.round(refRect.left)}px, ${Math.round(refRect.top)}px)`;
                Object.assign(popover.style, {
                    top: '0',
                    left: '0',
                    transform,
                });
            });
        }
    }
    async hideDropdownAsync() {
        const popover = await this.dialogRef.waitForCurrent();
        if (popover === null || popover === void 0 ? void 0 : popover.matches(':popover-open')) {
            popover.hidePopover();
        }
    }
    cleanupOnHide() {
        var _a, _b, _c;
        this.destroyAutoUpdate();
        (_a = this.arrowFocusController) === null || _a === void 0 ? void 0 : _a.disconnect();
        (_b = this.itemObserver) === null || _b === void 0 ? void 0 : _b.disconnect();
        (_c = this.disposeKeyListener) === null || _c === void 0 ? void 0 : _c.call(this);
    }
    async applyDropdownPosition() {
        var _a, _b;
        if (!this.show) {
            return;
        }
        if (!this.anchorElement) {
            return;
        }
        const referenceElement = this.anchorElement;
        const isSubmenu = this.isAnchorSubmenu();
        let targetElement = this.hostElement;
        let strategy = this.positioningStrategy;
        if (this.enableTopLayer) {
            const dialog = await this.dialogRef.waitForCurrent();
            if (!dialog) {
                return;
            }
            targetElement = dialog;
            strategy = 'fixed';
        }
        let positionConfig = {
            strategy,
            middleware: [],
        };
        if (!this.suppressAutomaticPlacement) {
            (_a = positionConfig.middleware) === null || _a === void 0 ? void 0 : _a.push(floatingUi_dom.flip({ fallbackStrategy: 'initialPlacement' }));
        }
        positionConfig.placement = isSubmenu ? 'right-start' : this.placement;
        positionConfig.middleware = [
            ...(((_b = positionConfig.middleware) === null || _b === void 0 ? void 0 : _b.filter(Boolean)) || []),
            floatingUi_dom.inline(),
            floatingUi_dom.shift(),
        ];
        if (this.offset) {
            positionConfig.middleware.push(floatingUi_dom.offset(this.offset));
        }
        this.destroyAutoUpdate();
        this.autoUpdateCleanup = floatingUi_dom.autoUpdate(referenceElement, targetElement, async () => {
            const computeResponse = await floatingUi_dom.computePosition(referenceElement, targetElement, positionConfig);
            Object.assign(targetElement.style, {
                top: '0',
                left: '0',
                transform: `translate(${Math.round(computeResponse.x)}px,${Math.round(computeResponse.y)}px)`,
            });
            if (this.overwriteDropdownStyle) {
                const overwriteStyle = await this.overwriteDropdownStyle({
                    dropdownRef: targetElement,
                    triggerRef: this.triggerElement,
                });
                Object.assign(targetElement.style, overwriteStyle);
            }
        }, {
            ancestorResize: true,
            ancestorScroll: true,
            elementResize: true,
        });
    }
    focusDropdownItem(index) {
        requestAnimationFrame.requestAnimationFrameNoNgZone(() => {
            var _a, _b;
            const button = (_b = (_a = this.dropdownItems[index]) === null || _a === void 0 ? void 0 : _a.shadowRoot) === null || _b === void 0 ? void 0 : _b.querySelector('button');
            if (button) {
                button.focus();
            }
        });
    }
    async componentDidLoad() {
        if (this.show) {
            this.changedShow(true);
        }
        if (!this.trigger) {
            return;
        }
        this.changedTrigger(this.trigger);
    }
    async componentDidRender() {
        await this.applyDropdownPosition();
        await this.resolveAnchorElement();
    }
    isTriggerElement(element) {
        const trigger = !!element.hasAttribute('data-ix-dropdown-trigger');
        return trigger;
    }
    onDropdownClick(event) {
        const target = dropdownController.dropdownController.pathIncludesTrigger(event.composedPath());
        if (target) {
            if (target !== this.triggerElement) {
                event.preventDefault();
            }
            if (this.isTriggerElement(target)) {
                if (this.closeBehavior === 'outside') {
                    event.preventDefault();
                }
                return;
            }
        }
        if (!event.defaultPrevented &&
            (this.closeBehavior === 'inside' || this.closeBehavior === 'both')) {
            dropdownController.dropdownController.dismissAll([this.getId()], this.ignoreRelatedSubmenu);
            return;
        }
        dropdownController.dropdownController.dismissOthers(this.getId());
    }
    /**
     * Update position of dropdown
     */
    async updatePosition() {
        this.applyDropdownPosition();
    }
    render() {
        if (this.enableTopLayer) {
            return (index.h(index.Host, { "data-ix-dropdown": this.localUId, class: {
                    'dropdown-menu': true,
                    show: this.show,
                } }, index.h("dialog", { ref: this.dialogRef, class: {
                    dialog: true,
                    overflow: !this.suppressOverflowBehavior,
                }, popover: "manual", "aria-modal": "true", "aria-label": this.header, tabindex: -1, onClick: (event) => this.onDropdownClick(event) }, index.h("div", { class: "dropdown-container" }, this.header && index.h("div", { class: "dropdown-header" }, this.header), this.show && index.h("slot", null)))));
        }
        return (index.h(index.Host, { "data-ix-dropdown": this.localUId, class: {
                'dropdown-menu': true,
                show: this.show,
                overflow: !this.suppressOverflowBehavior,
            }, style: {
                margin: '0',
                minWidth: '0px',
                position: this.positioningStrategy,
            }, role: "list", onClick: (event) => this.onDropdownClick(event) }, index.h("div", { style: { display: 'contents' } }, this.header && index.h("div", { class: "dropdown-header" }, this.header), this.show && index.h("slot", null))));
    }
    get hostElement() { return index.getElement(this); }
    static get watchers() { return {
        "show": ["changedShow"],
        "trigger": ["changedTrigger"]
    }; }
};
Dropdown.style = dropdownCss;

exports.ix_dropdown = Dropdown;
