// @siemens/ix@4.3.0/dist/collection/components/application-header/application-header.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { iconApps, iconMoreMenu } from "@siemens/ix-icons/icons";
import { h, Host, readTask, } from "@stencil/core";
import { a11yBoolean } from "../utils/a11y";
import { showAppSwitch } from "../utils/app-switch";
import { applicationLayoutService } from "../utils/application-layout";
import { ApplicationLayoutContext, } from "../utils/application-layout/context";
import { getCurrentBreakpoint } from "../utils/breakpoints";
import { useContextConsumer } from "../utils/context";
import { menuController } from "../utils/menu-service/menu-service";
import { hasSlottedElements } from "../utils/shadow-dom";
/**
 * @slot default - Place items on the right side of the header. If the screen size is small, the items will be shown inside a dropdown.
 * @slot secondary - Place additional items inside the header. They will appear after logo and name. If the screen size is small, the items will be shown inside a dropdown.
 * @slot overflow - Use this slot to display additional items that do not fit in the default or secondary slot.
 * @slot logo - Place a company logo inside the header. Alternatively the companyLogo property can be set.
 */
export class ApplicationHeader {
    constructor() {
        /**
         * Render subtle outline around app icon to ensure proper contrast.
         *
         * @since 4.0.0
         */
        this.appIconOutline = false;
        /**
         * Hides the bottom border of the header
         *
         * @since 4.0.0
         */
        this.hideBottomBorder = false;
        /**
         * Controls the visibility of the menu toggle button based on the context of the application header.
         *
         * When the application header is utilized outside the application frame, the menu toggle button is displayed.
         * Conversely, if the header is within the application frame, this property is ineffective.
         */
        this.showMenu = false;
        /**
         * Enable Popover API rendering for dropdown.
         *
         * @default false
         * @since 4.3.0
         */
        this.enableTopLayer = false;
        this.breakpoint = 'lg';
        this.menuExpanded = false;
        this.suppressResponsive = false;
        this.hasSlottedLogo = false;
        this.hasOverflowContextMenu = false;
        this.hasSecondarySlotElements = false;
        this.hasDefaultSlotElements = false;
        this.hasOverflowSlotElements = false;
    }
    get contentBackground() {
        return this.hostElement.shadowRoot.querySelector('.dropdown-content');
    }
    componentWillLoad() {
        this.breakpoint = getCurrentBreakpoint();
        useContextConsumer(this.hostElement, ApplicationLayoutContext, (ctx) => {
            this.breakpoint = applicationLayoutService.breakpoint;
            this.applicationLayoutContext = ctx;
            this.tryUpdateAppSwitch();
        }, true);
        this.menuDisposable = menuController.expandChange.on((show) => {
            this.menuExpanded = show;
        });
        this.modeDisposable = applicationLayoutService.onChange.on((mode) => {
            if (this.suppressResponsive) {
                return;
            }
            this.breakpoint = mode;
        });
        this.updateHasSlotAssignedElementsStates();
    }
    componentDidLoad() {
        this.attachSiemensLogoIfLoaded();
    }
    disconnectedCallback() {
        var _a, _b;
        (_a = this.menuDisposable) === null || _a === void 0 ? void 0 : _a.dispose();
        (_b = this.modeDisposable) === null || _b === void 0 ? void 0 : _b.dispose();
    }
    watchApplicationLayoutContext() {
        if (this.applicationLayoutContext) {
            this.showMenu = false;
        }
    }
    watchSuppressResponsive() {
        this.breakpoint = 'md';
    }
    checkLogoSlot() {
        const slotElement = this.hostElement.shadowRoot.querySelector('slot[name="logo"]');
        const isSiemensLogoDefined = window.customElements.get('ix-siemens-logo') !== undefined;
        if (isSiemensLogoDefined) {
            return hasSlottedElements(slotElement);
        }
        let assignedElements = slotElement === null || slotElement === void 0 ? void 0 : slotElement.assignedElements({ flatten: true });
        assignedElements = assignedElements === null || assignedElements === void 0 ? void 0 : assignedElements.filter((element) => element.tagName !== 'IX-SIEMENS-LOGO');
        return (assignedElements === null || assignedElements === void 0 ? void 0 : assignedElements.length) !== 0;
    }
    attachSiemensLogoIfLoaded() {
        if (this.companyLogo) {
            return;
        }
        if (!this.checkLogoSlot()) {
            const logoElement = document.createElement('ix-siemens-logo');
            logoElement.slot = 'logo';
            this.hostElement.appendChild(logoElement);
        }
    }
    async onMenuClick() {
        if (this.applicationLayoutContext) {
            menuController.toggle();
        }
        else {
            this.menuExpanded = !this.menuExpanded;
        }
        this.menuToggle.emit(this.menuExpanded);
    }
    resolveContextMenuButton() {
        return new Promise((resolve) => readTask(() => resolve(this.hostElement.shadowRoot.querySelector('[data-context-menu]'))));
    }
    tryUpdateAppSwitch() {
        var _a, _b;
        if (!this.callbackUpdateAppSwitchModal ||
            !((_a = this.applicationLayoutContext) === null || _a === void 0 ? void 0 : _a.appSwitchConfig)) {
            return;
        }
        this.callbackUpdateAppSwitchModal((_b = this.applicationLayoutContext) === null || _b === void 0 ? void 0 : _b.appSwitchConfig);
    }
    async showAppSwitch() {
        var _a, _b;
        const { defaultPrevented } = this.openAppSwitch.emit();
        if (defaultPrevented) {
            return;
        }
        if (!((_a = this.applicationLayoutContext) === null || _a === void 0 ? void 0 : _a.appSwitchConfig)) {
            return;
        }
        this.callbackUpdateAppSwitchModal = await showAppSwitch((_b = this.applicationLayoutContext) === null || _b === void 0 ? void 0 : _b.appSwitchConfig);
    }
    updateHasSlotAssignedElementsStates() {
        const defaultSlot = this.hostElement.shadowRoot.querySelector('.content slot:not([name])');
        const secondarySlot = this.hostElement.shadowRoot.querySelector('.content slot[name="secondary"]');
        const overflowSlot = this.hostElement.shadowRoot.querySelector('.content slot[name="overflow"]');
        this.hasDefaultSlotElements = hasSlottedElements(defaultSlot);
        this.hasSecondarySlotElements = hasSlottedElements(secondarySlot);
        this.hasOverflowSlotElements = hasSlottedElements(overflowSlot);
        this.hasOverflowContextMenu =
            this.hasDefaultSlotElements || this.hasSecondarySlotElements;
    }
    onContentBgClick(e) {
        if (e.target === this.contentBackground) {
            e.preventDefault();
        }
    }
    render() {
        var _a;
        const hasApplicationContextAvailable = !!this.applicationLayoutContext;
        const showMenuByApplicationFrame = this.breakpoint === 'sm' &&
            this.suppressResponsive === false &&
            hasApplicationContextAvailable;
        const showApplicationSwitch = ((_a = this.applicationLayoutContext) === null || _a === void 0 ? void 0 : _a.appSwitchConfig) &&
            this.breakpoint !== 'sm' &&
            this.suppressResponsive === false;
        const showCompanyLogoByProperty = this.breakpoint !== 'sm' && !!this.companyLogo;
        return (h(Host, { key: 'f70c16368414080ba826e2009b1e0f3f65cb2ff1', class: {
                [`breakpoint-${this.breakpoint}`]: true,
                'hide-bottom-border': this.hideBottomBorder,
            }, slot: "application-header" }, h("div", { key: 'aa68f775dc1b58aafdf4b4d37700113fbc509e2d', class: "left-side" }, this.appIcon && this.breakpoint !== 'sm' && (h("div", { key: 'ae9b4672c1348d21c9c48421e867634af213a5c7', class: {
                'app-icon': true,
                'app-icon-outline': this.appIconOutline,
            } }, h("img", { key: '2bf6e28d9867f08a6d954e132ece2c149e20f1b5', src: this.appIcon, alt: this.appIconAlt }))), (this.showMenu || showMenuByApplicationFrame) && (h("ix-menu-expand-icon", { key: '64ba854155f4ab55ff7aa2f2b15b2feb19145a4e', onClick: () => this.onMenuClick(), expanded: this.menuExpanded })), showApplicationSwitch && (h("ix-icon-button", { key: '265abaec15030ee7dc41a1e0535f09e32b99ff01', onClick: () => this.showAppSwitch(), icon: iconApps, variant: "subtle-tertiary", class: {
                'app-switch': true,
                'without-app-icon': !this.appIcon,
            }, "aria-label": this.ariaLabelAppSwitchIconButton })), showCompanyLogoByProperty && (h("div", { key: '34e1ee388c68ec1e5a4c5ab0a67ae07f66d3b2ef', class: "logo" }, h("img", { key: '86bb6a77c06e0f30ce41de66a4fe813bc236b536', src: this.companyLogo, alt: this.companyLogoAlt }))), h("div", { key: '3391b44a0cdaaa1b8b60e29fe71c9f9a04473919', class: {
                logo: true,
                'hide-logo': !this.hasSlottedLogo,
            } }, h("slot", { key: 'e0e554af7fe4746a6b12ada015f0f129b96bdd76', name: "logo", onSlotchange: () => (this.hasSlottedLogo = this.checkLogoSlot()) })), h("div", { key: 'cf3bca529813307e538a13ddc454d01cce5a8689', class: "name" }, h("ix-typography", { key: '9e4d9c3b6a3f45e6ceba918bc17fd4ff4a5453c3', format: "body-lg", class: "application-name" }, this.name), this.nameSuffix && this.breakpoint !== 'sm' && (h("ix-typography", { key: '9f7780499e2fb19e622e16a4f7e46084e5713f68', format: "body-xs", class: "application-name-suffix" }, this.nameSuffix)))), h("div", { key: '3a4dee26b48e356c2652517dd88887d936f926d5', class: { 'right-side': true, sm: this.breakpoint === 'sm' } }, this.breakpoint !== 'sm' && (h("div", { key: 'fd961eea6ec62a745dc2b6c9842a7dcd777c9218', class: "secondary" }, h("slot", { key: 'c214a8d62f28268acb56f1ecb88115d82f14cf0a', name: "secondary" }))), h("div", { key: '48bb5fd5dbb93f25c933436ce5033fbb75db6962', class: "content" }, this.breakpoint !== 'sm' && h("slot", { key: '868289b10946d895af657afb22d047a2864a9e8a' }), h("ix-icon-button", { key: '6b83cf6c6f04a1435c21a6f1a61a54f82531db48', class: {
                'context-menu': true,
                'context-menu-visible': this.hasOverflowContextMenu || this.hasOverflowSlotElements,
            }, "data-context-menu": true, "data-testid": "show-more", icon: iconMoreMenu, variant: "subtle-tertiary", "aria-label": this.ariaLabelMoreMenuIconButton, "aria-hidden": a11yBoolean(!(this.hasOverflowContextMenu || this.hasOverflowSlotElements)) }), h("ix-dropdown", { key: 'e0c44a160fccfc5d02672b09abff3ea209df4056', "data-overflow-dropdown": true, class: "dropdown", discoverAllSubmenus: true, trigger: this.resolveContextMenuButton(), "aria-hidden": a11yBoolean(!(this.hasOverflowContextMenu || this.hasOverflowSlotElements)), enableTopLayer: this.enableTopLayer }, h("div", { key: 'd17c4e3695fdcc517272552b4420175dfbd2f2c8', class: "dropdown-content", onClick: (e) => this.onContentBgClick(e) }, this.breakpoint === 'sm' && (h("div", { key: '6e96017787c861308cd4dbcaca0312dd0601746b', class: {
                'slot-content': true,
                'slot-content-active': this.hasSecondarySlotElements,
            } }, h("slot", { key: 'd078fd257e64da944799047951c8b9ab2d762e0d', name: "secondary", onSlotchange: () => this.updateHasSlotAssignedElementsStates() }))), this.breakpoint === 'sm' && (h("div", { key: '64d3c501c11859ed9c27d3a8f64294a080925ca4', class: {
                'slot-content': true,
                'slot-content-active': this.hasDefaultSlotElements,
            } }, h("slot", { key: 'aa759b7edbfa50848887dc23e649ffe85fd296fa', onSlotchange: () => this.updateHasSlotAssignedElementsStates() }))), h("div", { key: 'caf2086cb213c9650e52df3433eac548a7a5510e', class: {
                'slot-content': true,
                'slot-content-active': this.hasOverflowSlotElements,
            } }, h("slot", { key: 'e4bf46ff49a8416e405513115db147994a2c6df6', name: "overflow", onSlotchange: () => this.updateHasSlotAssignedElementsStates() })))), h("slot", { key: '17038a34906a10b5f0014a3f9f0e4bf8362f5695', name: "ix-application-header-avatar" })))));
    }
    static get is() { return "ix-application-header"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["application-header.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["application-header.css"]
        };
    }
    static get properties() {
        return {
            "name": {
                "type": "string",
                "attribute": "name",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Application name"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "nameSuffix": {
                "type": "string",
                "attribute": "name-suffix",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "Define a suffix which will be displayed next to the application name"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "companyLogo": {
                "type": "string",
                "attribute": "company-logo",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "Company logo will be show on the left side of the application name.\nIt will be hidden on smaller screens."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "companyLogoAlt": {
                "type": "string",
                "attribute": "company-logo-alt",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "Alt text for the company logo"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "appIcon": {
                "type": "string",
                "attribute": "app-icon",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "The app icon will be shown as the first element inside the header.\nIt will be hidden on smaller screens."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "appIconAlt": {
                "type": "string",
                "attribute": "app-icon-alt",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "Alt text for the app icon"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "appIconOutline": {
                "type": "boolean",
                "attribute": "app-icon-outline",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "Render subtle outline around app icon to ensure proper contrast."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "hideBottomBorder": {
                "type": "boolean",
                "attribute": "hide-bottom-border",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "Hides the bottom border of the header"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "showMenu": {
                "type": "boolean",
                "attribute": "show-menu",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Controls the visibility of the menu toggle button based on the context of the application header.\n\nWhen the application header is utilized outside the application frame, the menu toggle button is displayed.\nConversely, if the header is within the application frame, this property is ineffective."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "ariaLabelMenuExpandIconButton": {
                "type": "string",
                "attribute": "aria-label-menu-expand-icon-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }, {
                            "name": "deprecated",
                            "text": "This prop is no longer used as the menu expand button is hidden from screen readers. Will be removed in 5.0.0"
                        }],
                    "text": "ARIA label for the menu expand icon button"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "ariaLabelAppSwitchIconButton": {
                "type": "string",
                "attribute": "aria-label-app-switch-icon-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label for the app switch icon button"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "ariaLabelMoreMenuIconButton": {
                "type": "string",
                "attribute": "aria-label-more-menu-icon-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label for the more menu icon button"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "enableTopLayer": {
                "type": "boolean",
                "attribute": "enable-top-layer",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "default",
                            "text": "false"
                        }, {
                            "name": "since",
                            "text": "4.3.0"
                        }],
                    "text": "Enable Popover API rendering for dropdown."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get states() {
        return {
            "breakpoint": {},
            "menuExpanded": {},
            "suppressResponsive": {},
            "hasSlottedLogo": {},
            "hasOverflowContextMenu": {},
            "hasSecondarySlotElements": {},
            "hasDefaultSlotElements": {},
            "hasOverflowSlotElements": {},
            "applicationLayoutContext": {}
        };
    }
    static get events() {
        return [{
                "method": "menuToggle",
                "name": "menuToggle",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the menu toggle button is clicked"
                },
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                }
            }, {
                "method": "openAppSwitch",
                "name": "openAppSwitch",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.0.0"
                        }],
                    "text": "Event emitted when the app switch button is clicked"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "hostElement"; }
    static get watchers() {
        return [{
                "propName": "applicationLayoutContext",
                "methodName": "watchApplicationLayoutContext"
            }, {
                "propName": "suppressResponsive",
                "methodName": "watchSuppressResponsive"
            }];
    }
}
