// @siemens/ix@4.3.0/dist/collection/components/flip-tile/flip-tile.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { iconEye } from "@siemens/ix-icons/icons";
import { h, Host, } from "@stencil/core";
import { animate } from "animejs";
import Animation from "../utils/animation";
import { createMutationObserver } from "../utils/mutation-observer";
import { hasSlottedElements } from "../utils/shadow-dom";
export class FlipTile {
    constructor() {
        /**
         * Variation of the Flip
         *
         * @since 4.0.0
         */
        this.variant = 'filled';
        /**
         * Height interpreted as REM
         */
        this.height = 15.125;
        /**
         * Width interpreted as REM
         */
        this.width = 16;
        /**
         * Index of the currently visible content
         * @since 3.0.0
         */
        this.index = 0;
        this.isFlipAnimationActive = false;
        this.hasFooterSlot = false;
        this.contentItems = [];
    }
    watchIndex(newIndex, oldIndex) {
        if (newIndex === oldIndex) {
            return;
        }
        this.doFlipAnimation(newIndex);
    }
    componentDidLoad() {
        this.observer = createMutationObserver(() => this.updateContentItems());
        this.observer.observe(this.hostElement, {
            childList: true,
        });
    }
    componentWillLoad() {
        this.updateContentItems();
        this.updateContentVisibility(this.index);
    }
    disconnectedCallback() {
        if (this.observer) {
            this.observer.disconnect();
        }
    }
    handleFooterSlotChange(event) {
        const { target } = event;
        const slot = target;
        this.hasFooterSlot = hasSlottedElements(slot);
    }
    updateContentItems() {
        this.contentItems = Array.from(this.hostElement.querySelectorAll('ix-flip-tile-content'));
    }
    updateContentVisibility(indexVisible) {
        this.contentItems.forEach((content, index) => (content.contentVisible = index === indexVisible));
    }
    toggleIndex() {
        let newIndex;
        const oldIndex = this.index;
        if (this.index >= this.contentItems.length - 1) {
            newIndex = 0;
        }
        else {
            newIndex = this.index + 1;
        }
        const { defaultPrevented } = this.toggle.emit(newIndex);
        if (defaultPrevented) {
            this.index = oldIndex;
            return;
        }
        this.doFlipAnimation(newIndex);
    }
    doFlipAnimation(index) {
        if (this.isFlipAnimationActive) {
            return;
        }
        this.isFlipAnimationActive = true;
        animate(this.hostElement.shadowRoot.querySelector('.flip-tile-container'), {
            keyframes: {
                '0%': {
                    transform: 'rotateY(0)',
                },
                '50%': {
                    transform: 'rotateY(90deg)',
                },
                '51%': {
                    transform: 'rotateY(270deg)',
                },
                '100%': {
                    transform: 'rotateY(360deg)',
                },
            },
            duration: Animation.defaultTime,
            easing: 'ease-in-out',
            onComplete: () => {
                this.index = index;
                this.updateContentVisibility(this.index);
            },
        });
        setTimeout(() => {
            this.isFlipAnimationActive = false;
        }, 2 * Animation.defaultTime);
    }
    render() {
        return (h(Host, { key: 'a54038f2dbaeffac796822404e2d186c742a5357', class: {
                [`flip-tile-variant-${this.variant}`]: true,
            }, style: {
                height: `${this.height}${this.height === 'auto' ? '' : 'rem'}`,
                'min-height': `${this.height}${this.height === 'auto' ? '' : 'rem'}`,
                'max-height': `${this.height}${this.height === 'auto' ? '' : 'rem'}`,
                width: `${this.width}${this.width === 'auto' ? '' : 'rem'}`,
                'min-width': `${this.width}${this.width === 'auto' ? '' : 'rem'}`,
                'max-width': `${this.width}${this.width === 'auto' ? '' : 'rem'}`,
            } }, h("div", { key: '0bf12231b1cdeb0afe41d9d45c06fdef4cdedaf4', class: "flip-tile-container" }, h("div", { key: '3e9da89ed0b9cf7595e4c1d317d21ed1b55aa9d7', class: "flip-tile-header" }, h("div", { key: 'd4e7cc2a8162254ea3c7c4a0a031346b7a35e3d9', class: "header-slot-container text-l-title" }, h("slot", { key: 'dd9e2244a75179ceb4ac9647b403a5e2b01fd2ae', name: "header" })), h("ix-icon-button", { key: '74a5abdba3bf0c90222fc7bf7064742c0ea85b80', icon: iconEye, variant: "tertiary", onClick: () => this.toggleIndex(), "aria-label": this.ariaLabelEyeIconButton })), h("div", { key: '636884dbf0b0143e4f30061cd6e0472c83fe698a', class: "content-container" }, h("slot", { key: 'b94b3a488e7a137b06505290ab351724c5f01fca' })), h("div", { key: 'a15153c86848df2dd14b1423c686b9655d6fcdc1', class: {
                footer: true,
                'show-footer': this.hasFooterSlot,
            } }, h("slot", { key: '68d0585ff19d8a29fc8ccfdb2ec6302f767e3a94', name: "footer", onSlotchange: (event) => this.handleFooterSlotChange(event) })))));
    }
    static get is() { return "ix-flip-tile"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["flip-tile.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["flip-tile.css"]
        };
    }
    static get properties() {
        return {
            "variant": {
                "type": "string",
                "attribute": "variant",
                "mutable": false,
                "complexType": {
                    "original": "FlipTileVariant",
                    "resolved": "\"alarm\" | \"filled\" | \"info\" | \"outline\" | \"primary\" | \"warning\"",
                    "references": {
                        "FlipTileVariant": {
                            "location": "import",
                            "path": "./flip-tile.types",
                            "id": "src/components/flip-tile/flip-tile.types.ts::FlipTileVariant"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "Variation of the Flip"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'filled'"
            },
            "height": {
                "type": "any",
                "attribute": "height",
                "mutable": false,
                "complexType": {
                    "original": "number | 'auto'",
                    "resolved": "\"auto\" | number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Height interpreted as REM"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "15.125"
            },
            "width": {
                "type": "any",
                "attribute": "width",
                "mutable": false,
                "complexType": {
                    "original": "number | 'auto'",
                    "resolved": "\"auto\" | number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Width interpreted as REM"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "16"
            },
            "index": {
                "type": "number",
                "attribute": "index",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.0.0"
                        }],
                    "text": "Index of the currently visible content"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "0"
            },
            "ariaLabelEyeIconButton": {
                "type": "string",
                "attribute": "aria-label-eye-icon-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label for the eye icon button\nWill be set as aria-label on the nested HTML button element"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            }
        };
    }
    static get states() {
        return {
            "isFlipAnimationActive": {},
            "hasFooterSlot": {}
        };
    }
    static get events() {
        return [{
                "method": "toggle",
                "name": "toggle",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.0.0"
                        }],
                    "text": "Event emitted when the index changes"
                },
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "hostElement"; }
    static get watchers() {
        return [{
                "propName": "index",
                "methodName": "watchIndex"
            }];
    }
}
