// @siemens/ix@4.3.0/dist/cjs/index.cjs.js
'use strict';

var logicalFilterOperator = require('./logical-filter-operator-q-2KOPJs.js');
var tree_types = require('./tree.types-BygtRFW3.js');
var uploadFileState = require('./upload-file-state-BvLCvnjZ.js');
var modal = require('./modal-CUsbxNFf.js');
var animation = require('./animation-DtTuOXbI.js');
var typedEvent = require('./typed-event-DxWATHPL.js');
var rwd_util = require('./rwd.util-GTn-vgCJ.js');
var themeSwitcher = require('./theme-switcher-76Td30il.js');
var setup = require('./setup-Do7b4BhK.js');
require('./index-Dyz3z1FD.js');
require('./index-C8Xo8L1k.js');

/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
class FilterState {
    constructor() {
        this.tokens = [];
        this.categories = [];
    }
}

/**
 * Gets the toast container.
 */
function getToastContainer() {
    const containerList = Array.from(document.querySelectorAll('ix-toast-container'));
    const [container] = containerList;
    if (containerList.length > 1) {
        console.warn('Multiple toast containers were found. Only the first one will be used.');
        return container;
    }
    if (!container) {
        const toastContainer = document.createElement('ix-toast-container');
        document.body.appendChild(toastContainer);
        return toastContainer;
    }
    return container;
}
/**
 * Sets the position of the toast container
 */
function setToastPosition(position) {
    const container = getToastContainer();
    container.setAttribute('position', position);
}
/**
 * Display a toast based on the provided configuration
 */
function toast(config) {
    const container = getToastContainer();
    return container.showToast(config);
}
/**
 * Display an info toast based on the provided configuration
 */
toast.info = (config) => {
    return toast(Object.assign(Object.assign({}, config), { type: 'info' }));
};
/**
 * Display an error toast based on the provided configuration
 */
toast.error = (config) => {
    return toast(Object.assign(Object.assign({}, config), { type: 'error' }));
};
/**
 * Display a success toast based on the provided configuration
 */
toast.success = (config) => {
    return toast(Object.assign(Object.assign({}, config), { type: 'success' }));
};
/**
 * Display a warning toast based on the provided configuration
 */
toast.warning = (config) => {
    return toast(Object.assign(Object.assign({}, config), { type: 'warning' }));
};

/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
/**
 * Displays a loading modal with a message
 */
function showModalLoading(message) {
    const modal$1 = document.createElement('ix-modal');
    modal$1.disableEscapeClose = true;
    const loading = document.createElement('ix-modal-loading');
    loading.innerText = message;
    modal$1.appendChild(loading);
    modal.getCoreDelegate().attachView(modal$1);
    modal$1.showModal();
    return {
        update: (text) => (loading.innerHTML = text),
        finish: (text, timeout = 250) => {
            if (text !== undefined) {
                loading.innerText = text;
            }
            else {
                timeout = 0;
            }
            setTimeout(() => {
                modal$1.closeModal(null);
                setTimeout(async () => await modal.getCoreDelegate().removeView(modal$1), animation.Animation.mediumTime);
            }, timeout);
        },
    };
}

exports.InputState = logicalFilterOperator.InputState;
Object.defineProperty(exports, "LogicalFilterOperator", {
    enumerable: true,
    get: function () { return logicalFilterOperator.LogicalFilterOperator; }
});
exports.defaultRefreshTreeOptions = tree_types.defaultRefreshTreeOptions;
exports.renderDefaultItem = tree_types.renderDefaultItem;
Object.defineProperty(exports, "UploadFileState", {
    enumerable: true,
    get: function () { return uploadFileState.UploadFileState; }
});
exports.closeModal = modal.closeModal;
exports.dismissModal = modal.dismissModal;
exports.getCoreDelegate = modal.getCoreDelegate;
exports.registerFrameworkDelegate = modal.registerFrameworkDelegate;
exports.resolveDelegate = modal.resolveDelegate;
exports.showMessage = modal.showMessage;
exports.showModal = modal.showModal;
exports.TypedEvent = typedEvent.TypedEvent;
exports.convertToAbbreviationString = rwd_util.convertToAbbreviationString;
exports.convertToRemString = rwd_util.convertToRemString;
exports.getCurrentSystemAppearance = themeSwitcher.getCurrentSystemAppearance;
exports.themeSwitcher = themeSwitcher.themeSwitcher;
exports.handlePlatformHelpers = setup.handlePlatformHelpers;
exports.FilterState = FilterState;
exports.getToastContainer = getToastContainer;
exports.setToastPosition = setToastPosition;
exports.showModalLoading = showModalLoading;
exports.toast = toast;
