// @siemens/ix@4.3.0/dist/cjs/ix-breadcrumb.cjs.entry.js
'use strict';

var index = require('./index-C8Xo8L1k.js');
var a11y = require('./a11y-B_6chCvu.js');
var makeRef = require('./make-ref-BaH6_YJk.js');
var mutationObserver = require('./mutation-observer-DE0YMfYu.js');

const breadcrumbCss = ":host{display:flex;justify-content:flex-start;height:2.5rem;align-items:center;background-color:transparent;overflow:hidden}:host *,:host *::after,:host *::before{box-sizing:border-box}:host ::-webkit-scrollbar-button{display:none}@-moz-document url-prefix(){:host *{scrollbar-color:var(--theme-scrollbar-thumb--background) var(--theme-scrollbar-track--background);scrollbar-width:thin}}:host{}:host ::-webkit-scrollbar{width:0.5rem;height:0.5rem}:host{}:host ::-webkit-scrollbar-track{border-radius:5px;background:var(--theme-scrollbar-track--background)}:host ::-webkit-scrollbar-track:hover{background:var(--theme-scrollbar-track--background--hover)}:host{}:host ::-webkit-scrollbar-thumb{border-radius:5px;background:var(--theme-scrollbar-thumb--background)}:host{}:host ::-webkit-scrollbar-thumb:hover{background:var(--theme-scrollbar-thumb--background--hover)}:host ::-webkit-scrollbar-corner{display:none}:host .previous-button{width:3rem;min-width:0px}:host .crumb-dropdown{overflow:visible}:host .remove-anchor::after{display:none}:host .more-text{display:flex}:host .more-text .more-text-ellipsis{width:1rem;display:inline-block;font-weight:700}:host .more-text ix-icon{padding-top:2px}:host nav,:host ol,:host .crumb-items{display:contents}";

let sequenceId = 0;
const createId = (prefix = 'breadcrumb-') => {
    return `${prefix}-${sequenceId++}`;
};
const Breadcrumb = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.itemClick = index.createEvent(this, "itemClick", 7);
        this.nextClick = index.createEvent(this, "nextClick", 7);
        /**
         * Excess items will get hidden inside of dropdown
         */
        this.visibleItemCount = 9;
        /**
         * Items will be accessible through a dropdown
         */
        this.nextItems = [];
        /**
         * Ghost breadcrumbs will not show solid backgrounds on individual crumbs unless there is a mouse event (e.g. hover)
         */
        this.subtle = false;
        /**
         * Accessibility label for the dropdown button (ellipsis icon) used to access the dropdown list
         * with conditionally hidden previous items
         */
        this.ariaLabelPreviousButton = 'previous';
        /**
         * Enable Popover API rendering for dropdown.
         *
         * @default false
         * @since 4.3.0
         */
        this.enableTopLayer = false;
        this.previousButtonRef = makeRef.makeRef();
        this.nextButtonRef = makeRef.makeRef();
        this.items = [];
        this.isPreviousDropdownExpanded = false;
        this.previousButtonId = createId();
        this.previousDropdownId = createId();
    }
    onNextItemsChange() {
        this.onChildMutation();
    }
    onItemClick(item) {
        this.itemClick.emit(item);
    }
    componentDidLoad() {
        this.mutationObserver = mutationObserver.createMutationObserver(() => this.onChildMutation());
        this.mutationObserver.observe(this.hostElement, {
            subtree: true,
            childList: true,
        });
    }
    componentWillLoad() {
        this.onChildMutation();
    }
    disconnectedCallback() {
        var _a;
        (_a = this.mutationObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
    }
    async onChildMutation() {
        const updatedItems = this.getItems();
        updatedItems.forEach((bc, index) => {
            const shouldShowDropdown = this.nextItems.length !== 0 && updatedItems.length - 1 === index;
            bc.subtle = this.subtle;
            bc.hideChevron = updatedItems.length - 1 === index && !shouldShowDropdown;
            bc.isDropdownTrigger = shouldShowDropdown;
            if (shouldShowDropdown) {
                this.nextButtonRef(bc);
            }
            if (updatedItems.length < this.visibleItemCount) {
                return;
            }
            bc.invisible = index < updatedItems.length - this.visibleItemCount;
        });
        this.items = updatedItems;
    }
    getItems() {
        return Array.from(this.hostElement.querySelectorAll('ix-breadcrumb-item'));
    }
    render() {
        var _a, _b, _c, _d;
        const a11y$1 = a11y.a11yHostAttributes(this.hostElement);
        return (index.h(index.Host, { key: '6c7e20e30924d04854ca78ead9fd98d5ebe346bf' }, index.h("ix-dropdown", { key: '822c078f94f89e66db61e65f77d5e7bc6fb40aa2', id: this.previousDropdownId, "aria-label": this.ariaLabelPreviousButton, trigger: ((_a = this.items) === null || _a === void 0 ? void 0 : _a.length) > this.visibleItemCount
                ? this.previousButtonRef.waitForCurrent()
                : undefined, enableTopLayer: this.enableTopLayer, onShowChanged: ({ detail }) => {
                this.isPreviousDropdownExpanded = detail;
                const previousButton = this.hostElement.shadowRoot.getElementById(this.previousButtonId);
                // Need to force update previous button to change `aria-expanded`
                if (previousButton) {
                    index.forceUpdate(previousButton);
                }
            } }, this.items
            .slice(0, this.items.length - this.visibleItemCount)
            .map((item) => {
            var _a;
            const label = (_a = item.label) !== null && _a !== void 0 ? _a : item.innerText;
            return (index.h("ix-dropdown-item", { label: label, onClick: () => {
                    this.onItemClick(label);
                }, onItemClick: (event) => event.stopPropagation() }));
        })), ((_b = this.items) === null || _b === void 0 ? void 0 : _b.length) > this.visibleItemCount ? (index.h("ix-breadcrumb-item", { id: this.previousButtonId, ref: this.previousButtonRef, label: "...", tabIndex: 1, onItemClick: (event) => event.stopPropagation(), "aria-describedby": this.previousDropdownId, "aria-controls": this.previousDropdownId, "aria-expanded": a11y.a11yBoolean(this.isPreviousDropdownExpanded), class: 'previous-button' })) : null, index.h("nav", { key: 'e9f69c0666458f34d930bc5e19663fc9a8e69d90', class: "crumb-items", "aria-label": (_c = a11y$1['aria-label']) !== null && _c !== void 0 ? _c : 'breadcrumbs' }, index.h("ol", { key: '2ad3f37c9baffbd9eb8789dc8304dceb6772e0d4' }, index.h("slot", { key: '2bd136c9714401ddf2ea8bc5e3cb783b128ce795' }))), index.h("ix-dropdown", { key: 'ce382ec6bd3590e678a03517b31ec15bb7f2528f', trigger: this.nextButtonRef.waitForCurrent(), enableTopLayer: this.enableTopLayer }, (_d = this.nextItems) === null || _d === void 0 ? void 0 : _d.map((item) => (index.h("ix-dropdown-item", { label: item, onClick: (e) => {
                this.nextClick.emit({
                    event: e,
                    item,
                });
            }, onItemClick: (event) => event.stopPropagation() }))))));
    }
    get hostElement() { return index.getElement(this); }
    static get watchers() { return {
        "nextItems": ["onNextItemsChange"]
    }; }
};
Breadcrumb.style = breadcrumbCss;

exports.ix_breadcrumb = Breadcrumb;
