// @siemens/ix@4.3.0/dist/cjs/ix-split-button.cjs.entry.js
'use strict';

var index = require('./index-C8Xo8L1k.js');
var index$1 = require('./index-Dyz3z1FD.js');
var makeRef = require('./make-ref-BaH6_YJk.js');

const splitButtonCss = ".btn-group ix-button:first-child .btn{border-top-right-radius:0px;border-bottom-right-radius:0px;border-right:0px}.btn-group ix-button:last-child .btn{border-top-left-radius:0px;border-bottom-left-radius:0px}.btn-group ix-button:not(:first-child):not(:last-child) .btn{border-radius:0px;border-right:0px}:host{display:inline-block}:host *,:host *::after,:host *::before{box-sizing:border-box}:host ::-webkit-scrollbar-button{display:none}@-moz-document url-prefix(){:host *{scrollbar-color:var(--theme-scrollbar-thumb--background) var(--theme-scrollbar-track--background);scrollbar-width:thin}}:host{}:host ::-webkit-scrollbar{width:0.5rem;height:0.5rem}:host{}:host ::-webkit-scrollbar-track{border-radius:5px;background:var(--theme-scrollbar-track--background)}:host ::-webkit-scrollbar-track:hover{background:var(--theme-scrollbar-track--background--hover)}:host{}:host ::-webkit-scrollbar-thumb{border-radius:5px;background:var(--theme-scrollbar-thumb--background)}:host{}:host ::-webkit-scrollbar-thumb:hover{background:var(--theme-scrollbar-thumb--background--hover)}:host ::-webkit-scrollbar-corner{display:none}:host .btn-group{position:relative;display:inline-flex;vertical-align:middle;width:100%}:host .btn-group>ix-button:nth-child(1),:host .btn-group>ix-icon-button:nth-child(1){width:calc(100% - 2rem);--ix-button-border-radius-right:0}:host .btn-group>ix-icon-button:nth-child(2){width:2rem;--ix-button-border-radius-left:0}:host .middle-gap{gap:0.125rem}:host .left-button-border{border-top-width:0.125rem;border-right-width:0;border-left-width:0.125rem;border-bottom-width:0.125rem}";

const SplitButton = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.buttonClick = index.createEvent(this, "buttonClick", 7);
        /**
         * Color variant of button
         */
        this.variant = 'primary';
        /**
         * Controls if the dropdown will be closed in response to a click event depending on the position of the event relative to the dropdown.
         */
        this.closeBehavior = 'both';
        /**
         * Disabled
         */
        this.disabled = false;
        /**
         * Disables only the main button while keeping the dropdown trigger enabled
         *
         *  @since 4.1.0
         */
        this.disableButton = false;
        /**
         * Disables only the dropdown trigger while keeping the main button enabled
         *
         * @since 4.1.0
         */
        this.disableDropdownButton = false;
        /**
         * Placement of the dropdown
         */
        this.placement = 'bottom-start';
        /**
         * Enable Popover API rendering for dropdown.
         *
         * @default false
         * @since 4.3.0
         */
        this.enableTopLayer = false;
        this.toggle = false;
        this.triggerElementRef = makeRef.makeRef();
    }
    get isDisabledButton() {
        return this.disabled || this.disableButton;
    }
    get isDisabledIcon() {
        return this.disabled || this.disableDropdownButton;
    }
    render() {
        var _a;
        const hasOutline = this.variant.toLocaleLowerCase().includes('secondary');
        const baseAttributes = {
            variant: this.variant,
        };
        const buttonAttributes = Object.assign(Object.assign({}, baseAttributes), { disabled: this.isDisabledButton, class: {
                'left-button-border': !hasOutline,
            } });
        const dropdownButtonAttributes = Object.assign(Object.assign({}, baseAttributes), { disabled: this.isDisabledIcon });
        return (index.h(index.Host, { key: '034329433940b847951719d6df094e7e799c9b2b' }, index.h("div", { key: 'e08144dd52c338d0cb3a271a95442709a4bd10ba', class: { 'btn-group': true, 'middle-gap': !hasOutline } }, this.label ? (index.h("ix-button", Object.assign({}, buttonAttributes, { icon: this.icon, onClick: (e) => this.buttonClick.emit(e), "aria-label": this.ariaLabelButton }), this.label)) : (index.h("ix-icon-button", Object.assign({}, buttonAttributes, { icon: this.icon, onClick: (e) => this.buttonClick.emit(e), "aria-label": this.ariaLabelButton }))), index.h("ix-icon-button", Object.assign({ key: '4b7776b9328af463c873eb524e8808966b61fff1' }, dropdownButtonAttributes, { ref: this.triggerElementRef, class: 'anchor', icon: (_a = this.splitIcon) !== null && _a !== void 0 ? _a : index$1.iconContextMenu, "aria-label": this.ariaLabelSplitIconButton }))), index.h("ix-dropdown", { key: 'fbc53f2d73b27ed465a8dc4ae42357a61a6300c7', closeBehavior: this.closeBehavior, trigger: this.triggerElementRef.waitForCurrent(), enableTopLayer: this.enableTopLayer }, index.h("slot", { key: 'e776d44c8f78e89bed7f065edd55168563b0f23c' }))));
    }
    get hostElement() { return index.getElement(this); }
};
SplitButton.style = splitButtonCss;

exports.ix_split_button = SplitButton;
