// @siemens/ix@4.3.0/dist/cjs/ix-date-dropdown.cjs.entry.js
'use strict';

var index = require('./index-C8Xo8L1k.js');
var index$1 = require('./index-Dyz3z1FD.js');
var datetime = require('./datetime-Dvtv5O8f.js');
var makeRef = require('./make-ref-BaH6_YJk.js');

const dateDropdownCss = ":host{display:inline-block;position:relative}:host *,:host *::after,:host *::before{box-sizing:border-box}:host ::-webkit-scrollbar-button{display:none}@-moz-document url-prefix(){:host *{scrollbar-color:var(--theme-scrollbar-thumb--background) var(--theme-scrollbar-track--background);scrollbar-width:thin}}:host{}:host ::-webkit-scrollbar{width:0.5rem;height:0.5rem}:host{}:host ::-webkit-scrollbar-track{border-radius:5px;background:var(--theme-scrollbar-track--background)}:host ::-webkit-scrollbar-track:hover{background:var(--theme-scrollbar-track--background--hover)}:host{}:host ::-webkit-scrollbar-thumb{border-radius:5px;background:var(--theme-scrollbar-thumb--background)}:host{}:host ::-webkit-scrollbar-thumb:hover{background:var(--theme-scrollbar-thumb--background--hover)}:host ::-webkit-scrollbar-corner{display:none}@media (max-width: 48em){:host .min-width{width:-moz-min-content;width:min-content}}@media (min-width: 48em){:host .border-right{border-right:var(--theme-x-weak-bdr-2)}}:host ix-button{width:100%}:host .max-height{max-height:-moz-max-content;max-height:max-content}:host .no-margin{margin:0;padding:0}:host .pull-right{float:right;padding:0 1rem 0.75rem 1rem}";

const DateDropdown = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.dateRangeChange = index.createEvent(this, "dateRangeChange", 7);
        /**
         * Disable the button that opens the dropdown containing the date picker.
         */
        this.disabled = false;
        /**
         * Date format string.
         * See {@link https://moment.github.io/luxon/#/formatting?id=table-of-tokens} for all available tokens.
         */
        this.format = 'yyyy/LL/dd';
        /**
         * If true disables date range selection (from/to).
         */
        this.singleSelection = false;
        /**
         * Picker date. If the picker is in range mode this property is the start date.
         * If set to `null` no default start date will be pre-selected.
         *
         * Format is based on `format`
         */
        this.from = '';
        /**
         * Picker date. If the picker is in range mode this property is the end date.
         * If the picker is not in range mode leave this value `null`
         *
         * Format is based on `format`
         */
        this.to = '';
        /**
         * The earliest date that can be selected by the date picker.
         * If not set there will be no restriction.
         */
        this.minDate = '';
        /**
         * The latest date that can be selected by the date picker.
         * If not set there will be no restriction.
         */
        this.maxDate = '';
        /**
         * Used to set the initial select date range as well as the button name,
         * if not set or no according date range label is found, nothing will be selected
         */
        this.dateRangeId = 'custom';
        /**
         * Button variant
         */
        this.variant = 'primary';
        /**
         * Loading button
         */
        this.loading = false;
        /**
         * Shows week numbers displayed on the left side of the date picker
         *
         * @since 3.0.0
         */
        this.showWeekNumbers = false;
        /**
         * Controls whether custom date range selection is disabled in the component.
         * When set to 'false', the user can select a custom date range using the date picker.
         * When set to 'true', only predefined time date ranges are available for selection.
         */
        this.customRangeDisabled = false;
        /**
         * An array of predefined date range options for the date picker.
         * Each option is an object with a label describing the range and a function
         * that returns the start and end dates of the range as a DateRangeOption object.
         *
         * Example format:
         *   {
         *     id: 'some unique id',
         *     label: 'Name of the range',
         *     from: undefined, to: '2023/03/29'
         *   },
         *   // ... other predefined date range options ...
         */
        this.dateRangeOptions = [];
        /**
         * The index of which day to start the week on, based on the Locale#weekdays array.
         * E.g. if the locale is en-us, weekStartIndex = 1 results in starting the week on monday.
         */
        this.weekStartIndex = 0;
        /**
         * Text for custom dropdown item. Will be used for translation.
         */
        this.i18nCustomItem = 'Custom...';
        /**
         * Text for the done button. Will be used for translation.
         */
        this.i18nDone = 'Done';
        /**
         * Text for the done button. Will be used for translation.
         */
        this.i18nNoRange = 'No range set';
        /** @internal */
        this.today = datetime.DateTime.now().toISO();
        /**
         * Enable Popover API rendering for dropdown.
         *
         * @default false
         * @since 4.3.0
         */
        this.enableTopLayer = false;
        this.selectedDateRangeId = '';
        this.triggerRef = makeRef.makeRef();
        this.datePickerTouched = false;
    }
    onDateRangeIdChange() {
        this.onRangeListSelect(this.dateRangeId);
        this.updateCurrentDate();
        this.setDateRangeSelection(this.dateRangeId);
        if (!this.currentRangeValue) {
            return;
        }
        this.onDateSelect({
            from: this.currentRangeValue.from,
            to: this.currentRangeValue.to,
            id: this.currentRangeValue.id,
        });
    }
    onDateRangeOptionsChange() {
        this.initialize();
        this.onDateRangeIdChange();
    }
    onDisabledChange() {
        if (this.disabled) {
            this.closeDropdown();
        }
    }
    componentWillLoad() {
        this.initialize();
        this.setDateRangeSelection(this.dateRangeId);
    }
    /**
     * Retrieves the currently selected date range from the component.
     * This method returns the selected date range as a `DateChangeEvent` object.
     */
    async getDateRange() {
        return this.currentRangeValue || { id: '', from: '', to: '' };
    }
    initialize() {
        const isCustomRange = this.dateRangeId === 'custom' ||
            (!this.dateRangeId && !!this.from && !!this.to);
        if (isCustomRange && !this.customRangeDisabled) {
            this.selectedDateRangeId = 'custom';
            this.updateCurrentDate();
            return;
        }
        const isValidConfiguration = !isCustomRange && !this.from;
        if (!isValidConfiguration) {
            console.warn('"from" and "range-date-id" is provided this is an invalid combination. Using "custom".');
            this.selectedDateRangeId = 'custom';
            this.updateCurrentDate();
            return;
        }
    }
    updateCurrentDate() {
        this.currentRangeValue = {
            id: this.selectedDateRangeId,
            from: this.from,
            to: this.to,
        };
    }
    onDateSelect(rangeValue, preserveDropdown = true) {
        this.dateRangeChange.emit(rangeValue);
        if (preserveDropdown) {
            this.closeDropdown();
        }
        this.datePickerTouched = false;
    }
    onRangeListSelect(id) {
        if (this.setDateRangeSelection(id) && this.currentRangeValue) {
            this.onDateSelect(this.currentRangeValue);
        }
    }
    setDateRangeSelection(id) {
        this.selectedDateRangeId = id;
        const option = this.dateRangeOptions.find((range) => range.id === id);
        if (option) {
            this.currentRangeValue = option;
        }
        return option;
    }
    closeDropdown() {
        const dropdown = this.hostElement.shadowRoot.querySelector('ix-dropdown');
        if (dropdown) {
            dropdown.show = false;
        }
    }
    getButtonLabel() {
        var _a, _b;
        if (this.selectedDateRangeId === 'custom' && ((_a = this.currentRangeValue) === null || _a === void 0 ? void 0 : _a.from)) {
            let range = this.currentRangeValue.from;
            if (this.currentRangeValue.to) {
                range += ` - ${this.currentRangeValue.to}`;
            }
            return range;
        }
        if (!this.selectedDateRangeId) {
            return this.i18nNoRange;
        }
        if (!this.dateRangeOptions || ((_b = this.dateRangeOptions) === null || _b === void 0 ? void 0 : _b.length) === 0) {
            return this.i18nNoRange;
        }
        const option = this.dateRangeOptions.find((option) => option.id === this.selectedDateRangeId);
        if (!option) {
            console.error(`Cannot find range option with id ${this.selectedDateRangeId}`);
            return this.i18nNoRange;
        }
        return option.label;
    }
    render() {
        var _a, _b, _c;
        return (index.h(index.Host, { key: '34b878eb8221b6c13f235d1b42ec63d3b6dce25c' }, index.h("ix-button", { key: 'c40cff0539f79b19c0af903edce79de7d52e704e', "data-testid": "date-dropdown-trigger", "data-date-dropdown-trigger": true, variant: this.variant, loading: this.loading, icon: index$1.iconHistory, ref: this.triggerRef, disabled: this.disabled, ariaLabel: this.ariaLabelDropdownButton }, this.getButtonLabel()), index.h("ix-dropdown", { key: '3a74d5b8d2f4e7c7487b0bac9af67c3d25d6b75d', "data-testid": "date-dropdown", "data-date-dropdown": true, class: "min-width max-height", trigger: this.triggerRef.waitForCurrent(), closeBehavior: "outside", placement: "bottom-start", enableTopLayer: this.enableTopLayer, onShowChanged: ({ detail: show }) => {
                if (!show &&
                    this.selectedDateRangeId === 'custom' &&
                    this.datePickerTouched &&
                    this.currentRangeValue) {
                    this.onDateSelect(this.currentRangeValue);
                }
            } }, index.h("ix-layout-grid", { key: '0aeda647289d651805bdd7eaac3528b5ee6b649b', noMargin: true }, index.h("ix-row", { key: '43bed15b884d437de29d9879be99e1cc4b48d41a' }, ((_a = this.dateRangeOptions) === null || _a === void 0 ? void 0 : _a.length) > 1 && (index.h("ix-col", { key: '3beb46a324a3d12bcb13be28956d7219f708b528', class: {
                'no-margin': true,
                'border-right': this.selectedDateRangeId === 'custom',
            } }, this.dateRangeOptions.map((dateRangeOption) => (index.h("ix-dropdown-item", { label: dateRangeOption.label, onClick: () => this.onRangeListSelect(dateRangeOption.id), checked: this.selectedDateRangeId === dateRangeOption.id }))), index.h("div", { key: 'f0e7910986569a41b7d0ca0419361e7c9e81be2d', hidden: this.customRangeDisabled }, index.h("ix-dropdown-item", { key: 'c1af0808c290045824cbdba19dc72d46d23e2d66', label: this.i18nCustomItem, checked: this.selectedDateRangeId === 'custom', onClick: () => this.onRangeListSelect('custom') })))), index.h("ix-col", { key: 'b3b1b70c86b309f2c8572751a30007ef1da9e9f3', class: "no-margin" }, this.selectedDateRangeId === 'custom' && (index.h(index.Fragment, { key: '70e07f3dec874213296af70b8ca63e631a247541' }, index.h("ix-date-picker", { key: '386b034422347098aff8b09255e83368f0d8d106', embedded: true, locale: this.locale, onDateChange: (e) => {
                e.stopPropagation();
                this.currentRangeValue = Object.assign(Object.assign({}, e.detail), { id: 'custom' });
                this.datePickerTouched = true;
            }, onDateRangeChange: (e) => e.stopPropagation(), format: this.format, singleSelection: this.singleSelection, from: this.from || ((_b = this.currentRangeValue) === null || _b === void 0 ? void 0 : _b.from), to: this.to || ((_c = this.currentRangeValue) === null || _c === void 0 ? void 0 : _c.to), minDate: this.minDate, maxDate: this.maxDate, today: this.today, weekStartIndex: this.weekStartIndex, showWeekNumbers: this.showWeekNumbers }), index.h("div", { key: '3f3f115659920a42b192c9f10cdd81af5a0c04ae', class: "pull-right" }, index.h("ix-button", { key: '2fb0e8b8d6de6244a1d9d60d175226a24c2103ad', onClick: () => {
                if (this.currentRangeValue) {
                    this.onDateSelect(this.currentRangeValue);
                }
            } }, this.i18nDone))))))))));
    }
    get hostElement() { return index.getElement(this); }
    static get watchers() { return {
        "dateRangeId": ["onDateRangeIdChange"],
        "to": ["onDateRangeIdChange"],
        "from": ["onDateRangeIdChange"],
        "dateRangeOptions": ["onDateRangeOptionsChange"],
        "disabled": ["onDisabledChange"]
    }; }
};
DateDropdown.style = dateDropdownCss;

exports.ix_date_dropdown = DateDropdown;
