// @siemens/ix@4.3.0/dist/cjs/ix-helper-text.cjs.entry.js
'use strict';

var index = require('./index-C8Xo8L1k.js');
var validation = require('./validation-DTmSqO2d.js');
var helperTextUtil = require('./helper-text-util-D2W-yNmK.js');
require('./index-Dyz3z1FD.js');

const helperTextCss = ":host{display:block}:host .bottom-text{display:flex;position:relative;align-items:flex-start;justify-content:flex-start;gap:0.25rem;margin-right:0.25rem}:host .text-icon{margin:0.125rem}:host .text-icon.invalid{color:var(--theme-helper-icon--color--invalid)}:host .text-icon.info{color:var(--theme-helper-icon--color--info)}:host .text-icon.warning{color:var(--theme-helper-icon--color--warning)}:host .text-icon.valid{color:var(--theme-helper-icon--color--valid)}";

const HelperText = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.validationResults = {
            isInfo: false,
            isInvalid: false,
            isValid: false,
            isWarning: false,
            isInvalidByRequired: false,
        };
        this.observer = new MutationObserver(() => this.checkForRequired());
    }
    connectedCallback() {
        this.observer.observe(window.document, {
            childList: true,
            subtree: true,
        });
    }
    disconnectedCallback() {
        if (this.observer) {
            this.observer.disconnect();
        }
    }
    componentWillRender() {
        this.checkForRequired();
    }
    async checkForRequired() {
        if (!this.htmlFor) {
            return;
        }
        const forElement = document.getElementById(this.htmlFor);
        if (!forElement) {
            return;
        }
        if (this.classObserver) {
            this.classObserver.destroy();
        }
        this.classObserver = validation.createClassMutationObserver(forElement, () => {
            this.validationResults = validation.checkFieldClasses(forElement);
        });
        this.validationResults = validation.checkFieldClasses(forElement);
    }
    render() {
        return (index.h(index.Host, { key: 'aa99794dc7b9c59abf2fdd8f2ac8112cf705a5ad' }, index.h(helperTextUtil.HelperText, Object.assign({ key: 'c3e36ebbae8617aa43cf2616242b035739ab4cc7', helperText: this.helperText,
            invalidText: this.invalidText,
            validText: this.validText,
            infoText: this.infoText,
            warningText: this.warningText }, this.validationResults))));
    }
    get hostElement() { return index.getElement(this); }
};
HelperText.style = helperTextCss;

exports.ix_helper_text = HelperText;
