// @siemens/ix@4.3.0/dist/collection/components/utils/modal/message.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { iconError, iconInfo, iconQuestion, iconSuccess, iconWarning, } from "@siemens/ix-icons/icons";
import { getCoreDelegate } from "../delegate";
import { TypedEvent } from "../typed-event";
function setA11yAttributes(element, config) {
    const ariaDescribedby = config.ariaDescribedby;
    const ariaLabelledby = config.ariaLabelledby;
    delete config['ariaDescribedby'];
    delete config['ariaLabelledby'];
    if (ariaDescribedby) {
        element.setAttribute('aria-describedby', ariaDescribedby);
    }
    if (ariaLabelledby) {
        element.setAttribute('aria-labelledby', ariaLabelledby);
    }
}
function createConfirmButtons(textOkay, textCancel, payloadOkay, payloadCancel) {
    let actions = [];
    if (textCancel !== undefined) {
        actions = [
            ...actions,
            {
                id: 'cancel',
                text: textCancel,
                type: 'cancel',
                payload: payloadCancel,
            },
        ];
    }
    return [
        ...actions,
        {
            id: 'okay',
            text: textOkay,
            type: 'okay',
            payload: payloadOkay,
        },
    ];
}
/**
 * Displays a message modal with configurable actions and returns an event emitter for action responses
 */
export async function showMessage(config) {
    const onMessageAction = new TypedEvent();
    const dialog = document.createElement('ix-modal');
    const header = document.createElement('ix-modal-header');
    const content = document.createElement('ix-modal-content');
    const footer = document.createElement('ix-modal-footer');
    setA11yAttributes(dialog, config);
    Object.assign(header, config);
    Object.assign(content, config);
    Object.assign(footer, config);
    header.innerText = config.messageTitle;
    content.innerText = config.message;
    config.actions.forEach(({ id, text, type, payload }) => {
        const button = document.createElement('ix-button');
        button.innerText = text;
        footer.appendChild(button);
        if (type === 'okay') {
            button.variant = 'primary';
            button.addEventListener('click', () => dialog.closeModal({
                actionId: id,
                payload,
            }));
            return;
        }
        else if (type === 'cancel') {
            button.variant = 'secondary';
            button.addEventListener('click', () => dialog.dismissModal({
                actionId: id,
                payload,
            }));
            return;
        }
    });
    dialog.appendChild(header);
    dialog.appendChild(content);
    dialog.appendChild(footer);
    const dialogRef = await getCoreDelegate().attachView(dialog);
    dialogRef.addEventListener('dialogClose', (event) => {
        onMessageAction.emit(event.detail);
        dialogRef.remove();
    });
    dialogRef.addEventListener('dialogDismiss', (event) => {
        onMessageAction.emit(event.detail);
        dialogRef.remove();
    });
    setA11yAttributes(dialogRef, config);
    Object.assign(dialogRef, config);
    dialogRef.showModal();
    return onMessageAction;
}
/**
 * Displays an info message modal with an info icon
 */
showMessage.info = (title, message, textOkay, textCancel, payloadOkay, payloadCancel) => {
    return showMessage({
        message,
        messageTitle: title,
        icon: iconInfo,
        actions: createConfirmButtons(textOkay, textCancel, payloadOkay, payloadCancel),
    });
};
/**
 * Displays a warning message modal with a warning icon
 */
showMessage.warning = (title, message, textOkay, textCancel, payloadOkay, payloadCancel) => {
    return showMessage({
        message,
        messageTitle: title,
        icon: iconWarning,
        //TODO(IX-3400): Remove warning-text when proper CSS variable is available
        iconColor: 'color-warning-text',
        actions: createConfirmButtons(textOkay, textCancel, payloadOkay, payloadCancel),
    });
};
/**
 * Displays an error message modal with an error icon
 */
showMessage.error = (title, message, textOkay, textCancel, payloadOkay, payloadCancel) => {
    return showMessage({
        message,
        messageTitle: title,
        icon: iconError,
        iconColor: 'color-alarm',
        actions: createConfirmButtons(textOkay, textCancel, payloadOkay, payloadCancel),
    });
};
/**
 * Displays a success message modal with a success icon
 */
showMessage.success = (title, message, textOkay, textCancel, payloadOkay, payloadCancel) => {
    return showMessage({
        message,
        messageTitle: title,
        icon: iconSuccess,
        iconColor: 'color-success',
        actions: createConfirmButtons(textOkay, textCancel, payloadOkay, payloadCancel),
    });
};
/**
 * Displays a question message modal with a question icon
 */
showMessage.question = (title, message, textOkay, textCancel, payloadOkay, payloadCancel) => {
    return showMessage({
        message,
        messageTitle: title,
        icon: iconQuestion,
        actions: createConfirmButtons(textOkay, textCancel, payloadOkay, payloadCancel),
    });
};
