// @siemens/ix@4.3.0/dist/collection/components/group/group.js
/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host, } from "@stencil/core";
import { createMutationObserver } from "../utils/mutation-observer";
import { hasSlottedElements } from "../utils/shadow-dom";
import { iconChevronDownSmall, iconChevronUpSmall, } from "@siemens/ix-icons/icons";
export class Group {
    constructor() {
        /**
         * Prevent header from being selectable
         */
        this.suppressHeaderSelection = false;
        /**
         * Whether the group is expanded or collapsed. Defaults to false.
         */
        this.expanded = false;
        /**
         * Whether the group is selected.
         */
        this.selected = false;
        /**
         * Expand the group if the header is clicked
         */
        this.expandOnHeaderClick = false;
        this.itemSelected = false;
        this.slotSize = this.groupItems.length;
        this.footerVisible = false;
        this.showExpandCollapsedIcon = false;
        this.hasDropdown = false;
        this.observer = null;
    }
    selectedChanged(newSelected) {
        if (newSelected === false) {
            this.changeItemIndex();
        }
    }
    get dropdownItems() {
        return Array.from(this.hostElement.querySelectorAll('ix-group-dropdown-item'));
    }
    get groupItems() {
        return Array.from(this.hostElement.querySelectorAll('ix-group-item:not(.footer)'));
    }
    get groupContent() {
        var _a;
        return (_a = this.hostElement.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('.group-content');
    }
    onExpandClick(event) {
        const oldExpanded = this.expanded;
        this.expanded = !this.expanded;
        const { defaultPrevented } = this.expandedChanged.emit(this.expanded);
        event.stopPropagation();
        if (defaultPrevented) {
            this.expanded = oldExpanded;
        }
    }
    onHeaderClick(event) {
        if (this.suppressHeaderSelection) {
            this.onExpandClick(event);
            return;
        }
        this.changeHeaderSelection(!this.selected);
        this.changeItemIndex();
    }
    changeHeaderSelection(newSelection) {
        const oldIsHeaderSelected = this.selected;
        const newIsHeaderSelected = newSelection;
        this.selected = newIsHeaderSelected;
        const { defaultPrevented } = this.selectGroup.emit(newIsHeaderSelected);
        if (defaultPrevented) {
            this.selected = oldIsHeaderSelected;
            return;
        }
    }
    changeItemIndex(index) {
        const oldIndex = this.index;
        const newIndex = index === this.index ? undefined : index;
        if (this.index === newIndex) {
            return;
        }
        this.index = newIndex;
        const { defaultPrevented } = this.selectItem.emit(newIndex);
        if (defaultPrevented) {
            this.index = oldIndex;
            return;
        }
        const items = this.groupItems;
        items.forEach((item, i) => {
            item.selected = i === this.index;
        });
        this.itemSelected = items.some((item) => item.selected);
    }
    onSlotChange() {
        var _a;
        const slot = (_a = this.hostElement.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('slot[name="footer"]');
        if (slot) {
            this.footerVisible = hasSlottedElements(slot);
        }
    }
    checkDropdownSlot() {
        this.hasDropdown = !!this.hostElement.querySelector('[slot="dropdown"]');
    }
    componentWillRender() {
        this.groupItems.forEach((item, index) => {
            item.selected = index === this.index;
            item.index = index;
        });
        this.checkDropdownSlot();
    }
    componentDidLoad() {
        this.observer = createMutationObserver(() => {
            this.slotSize = this.groupItems.length;
        });
        if (!this.groupContent) {
            return;
        }
        this.observer.observe(this.groupContent, {
            childList: true,
        });
        this.checkDropdownSlot();
    }
    disconnectedCallback() {
        if (this.observer) {
            this.observer.disconnect();
        }
    }
    onItemClicked(event) {
        if (event.target instanceof HTMLElement) {
            const item = event.target;
            const index = this.groupItems.indexOf(item);
            this.changeItemIndex(index);
        }
    }
    render() {
        return (h(Host, { key: '99b00ca8c88730036688974abd67ff0534dd4d17' }, h("div", { key: '47384431d36ce7cdc5a2e81ecd682d6a0dd13b53', class: {
                'group-header': true,
                expand: this.expanded,
                selected: this.selected,
            }, tabindex: "0" }, h("div", { key: '68968fc5816ca0fef0ae658a6925cc6b8ee3479a', class: "group-header-clickable", onClick: (e) => this.onHeaderClick(e) }, h("div", { key: 'a26153957aa76d882d95a3f88318c01517078a64', class: {
                'group-header-selection-indicator': true,
                'group-header-selection-indicator-item-selected': this.itemSelected,
            } }), h("div", { key: '01427b28c1b873a1875bd4c2093fee6d900bdf89', class: "btn-expand-header" }, h("ix-icon", { key: '1feb76e98b4ebda01d902b76065cdcefa0dde5a4', "data-testid": "expand-collapsed-icon", class: {
                hidden: !this.showExpandCollapsedIcon,
            }, name: this.expanded ? iconChevronUpSmall : iconChevronDownSmall, onClick: (event) => this.onExpandClick(event) })), h("div", { key: '5c6232f9b9a36ccbfa2f0e9352a5e739e202e275', class: "group-header-content" }, this.header ? (h("div", { class: "group-header-props-container" }, h("div", { class: "group-header-title" }, h("span", { title: this.header }, this.header)), h("div", { class: "group-subheader", title: this.subHeader }, this.subHeader))) : null, h("slot", { key: '592d03499639a45183c8837b421a33bbffbae954', name: "header" }))), this.hasDropdown && (h("ix-group-context-menu", { key: 'f666ce19d8441bd3afc40800128e6326da14615e' }, h("slot", { key: '96d9509e35fbf388d31fbc832fc31dd2167842a8', name: "dropdown" })))), h("div", { key: '3d6f649530d00fd19761bb84ed4bf8769b0bec14', class: {
                'group-content': true,
            } }, h("div", { key: 'cad4a7c42f5402400add10d1d5ac57174aa34b6e', style: {
                display: this.expanded ? 'contents' : 'none',
            } }, h("slot", { key: '305a97c24f58db1e1d1264876c99316aceda6e8d', onSlotchange: () => {
                var _a;
                const slot = (_a = this.hostElement.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('slot:not([name])');
                this.showExpandCollapsedIcon = hasSlottedElements(slot);
            } }), h("ix-group-item", { key: 'f19cce4aad2fc8465f57d94bb6a6b228204120e6', class: {
                footer: true,
                'footer-visible': this.footerVisible,
            }, groupFooter: true, suppressSelection: true }, h("slot", { key: 'c073314bc4528f7559c8c1cfa43b88881d7a4e49', name: "footer", onSlotchange: () => this.onSlotChange() }))))));
    }
    static get is() { return "ix-group"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["group.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["group.css"]
        };
    }
    static get properties() {
        return {
            "suppressHeaderSelection": {
                "type": "boolean",
                "attribute": "suppress-header-selection",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Prevent header from being selectable"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "header": {
                "type": "string",
                "attribute": "header",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Group header"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "subHeader": {
                "type": "string",
                "attribute": "sub-header",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Group header subtitle"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "expanded": {
                "type": "boolean",
                "attribute": "expanded",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Whether the group is expanded or collapsed. Defaults to false."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "defaultValue": "false"
            },
            "selected": {
                "type": "boolean",
                "attribute": "selected",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Whether the group is selected."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "defaultValue": "false"
            },
            "index": {
                "type": "number",
                "attribute": "index",
                "mutable": true,
                "complexType": {
                    "original": "number",
                    "resolved": "number | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The index of the selected group entry.\nIf undefined no group item is selected."
                },
                "getter": false,
                "setter": false,
                "reflect": true
            },
            "expandOnHeaderClick": {
                "type": "boolean",
                "attribute": "expand-on-header-click",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Expand the group if the header is clicked"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get states() {
        return {
            "itemSelected": {},
            "slotSize": {},
            "footerVisible": {},
            "showExpandCollapsedIcon": {},
            "hasDropdown": {}
        };
    }
    static get events() {
        return [{
                "method": "selectGroup",
                "name": "selectGroup",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emits when whole group gets selected."
                },
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                }
            }, {
                "method": "selectItem",
                "name": "selectItem",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emits when group item gets selected."
                },
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                }
            }, {
                "method": "expandedChanged",
                "name": "expandedChanged",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Group expanded"
                },
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "hostElement"; }
    static get watchers() {
        return [{
                "propName": "selected",
                "methodName": "selectedChanged"
            }];
    }
    static get listeners() {
        return [{
                "name": "selectedChanged",
                "method": "onItemClicked",
                "target": undefined,
                "capture": false,
                "passive": false
            }];
    }
}
