// @siemens/ix/dist/collection/components/tabs/tabs.js
/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host, } from "@stencil/core";
import { requestAnimationFrameNoNgZone } from "../utils/requestAnimationFrame";
import { iconChevronLeftSmall, iconChevronRightSmall, } from "@siemens/ix-icons/icons";
const TAB_MANAGED_CLASSES = {
    SELECTED: 'selected',
    DISABLED: 'disabled',
    SMALL_TAB: 'small-tab',
    ICON: 'icon',
    STRETCHED: 'stretched',
    BOTTOM: 'bottom',
    TOP: 'top',
    CIRCLE: 'circle',
    HYDRATED: 'hydrated',
};
const MANAGED_CLASSES_SET = new Set(Object.values(TAB_MANAGED_CLASSES));
export class Tabs {
    constructor() {
        /**
         * Set tab items to small size
         */
        this.small = false;
        /**
         * Set rounded tabs
         */
        this.rounded = false;
        /**
         * Set default selected tab by index
         */
        this.selected = 0;
        /**
         * Set layout width style
         */
        this.layout = 'auto';
        /**
         * Set placement style
         */
        this.placement = 'bottom';
        this.totalItems = 0;
        this.currentScrollAmount = 0;
        this.scrollAmount = 100;
        this.scrollActionAmount = 0;
        this.showArrowPrevious = false;
        this.showArrowNext = false;
        this.windowStartSize = window.innerWidth;
        this.ARROW_WIDTH = 32;
        this.updateScheduled = false;
        this.clickAction = {
            timeout: null,
            isClick: true,
        };
    }
    onWindowResize() {
        this.totalItems = 0;
        this.totalItems = this.getTabs().length;
        if (this.windowStartSize === 0)
            return (this.windowStartSize = window.innerWidth);
        this.move(this.windowStartSize - window.innerWidth);
        this.windowStartSize = window.innerWidth;
    }
    getTabs() {
        return Array.from(this.hostElement.querySelectorAll('ix-tab-item'));
    }
    getTab(tabIndex) {
        return this.getTabs()[tabIndex];
    }
    getTabsWrapper() {
        var _a;
        return (_a = this.hostElement.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('.items-content');
    }
    getTabsContainer() {
        var _a;
        return (_a = this.hostElement.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('.tab-items');
    }
    initResizeObserver() {
        const parentElement = this.hostElement.parentElement;
        if (!parentElement)
            return;
        this.resizeObserver = new ResizeObserver(() => {
            this.renderArrows();
        });
        this.resizeObserver.observe(parentElement);
    }
    observeSlotChanges() {
        var _a;
        (_a = this.classObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
        this.classObserver = new MutationObserver(() => {
            this.scheduleTabUpdate();
        });
        this.classObserver.observe(this.hostElement, {
            childList: true,
            subtree: true,
            attributes: true,
            attributeFilter: ['class'],
        });
    }
    scheduleTabUpdate() {
        if (this.updateScheduled)
            return;
        this.updateScheduled = true;
        requestAnimationFrame(() => {
            this.updateTabAttributes();
            this.updateScheduled = false;
        });
    }
    setTabAttributes(element, index) {
        const isSelected = index === this.selected;
        const isDisabled = element.disabled;
        if (this.small)
            element.setAttribute('small', 'true');
        if (this.rounded)
            element.setAttribute('rounded', 'true');
        element.setAttribute('layout', this.layout);
        element.setAttribute('selected', isSelected.toString());
        element.setAttribute('placement', this.placement);
        element.toggleAttribute('disabled', isDisabled);
        this.applyRequiredClasses(element, isSelected, isDisabled);
    }
    applyRequiredClasses(element, isSelected, isDisabled) {
        const requiredClasses = new Set(this.buildRequiredClasses(isSelected, isDisabled));
        const { classList } = element;
        for (const cls of requiredClasses) {
            classList.add(cls);
        }
        for (const managedClass of MANAGED_CLASSES_SET) {
            if (!requiredClasses.has(managedClass)) {
                classList.remove(managedClass);
            }
        }
    }
    buildRequiredClasses(isSelected, isDisabled) {
        const classConditions = {
            [TAB_MANAGED_CLASSES.HYDRATED]: true,
            [TAB_MANAGED_CLASSES.SELECTED]: isSelected,
            [TAB_MANAGED_CLASSES.DISABLED]: isDisabled,
            [TAB_MANAGED_CLASSES.SMALL_TAB]: this.small,
            [TAB_MANAGED_CLASSES.STRETCHED]: this.layout === 'stretched',
            [TAB_MANAGED_CLASSES.BOTTOM]: this.placement === 'bottom',
            [TAB_MANAGED_CLASSES.TOP]: this.placement === 'top',
            [TAB_MANAGED_CLASSES.CIRCLE]: this.rounded,
        };
        return Object.entries(classConditions)
            .filter(([, condition]) => condition)
            .map(([className]) => className);
    }
    ensureSelectedIndex() {
        if (this.totalItems === 0) {
            console.warn('ix-tabs: No tabs available for selection');
            this.selected = -1;
            return;
        }
        if (this.selected < this.totalItems) {
            return;
        }
        const originalIndex = this.selected;
        const previousIndex = originalIndex - 1;
        if (previousIndex >= 0 && previousIndex < this.totalItems) {
            this.updateSelected(previousIndex);
            return;
        }
        if (this.totalItems > 0) {
            this.updateSelected(0);
        }
    }
    updateSelected(index) {
        this.selected = index;
        this.selectedChange.emit(index);
    }
    updateTabAttributes() {
        const tabs = this.getTabs();
        this.totalItems = tabs.length;
        this.ensureSelectedIndex();
        for (const [index, element] of tabs.entries()) {
            this.setTabAttributes(element, index);
        }
        this.renderArrows();
    }
    showArrows() {
        try {
            const tabWrapper = this.getTabsWrapper();
            return (tabWrapper &&
                tabWrapper.scrollWidth >
                    Math.ceil(tabWrapper.getBoundingClientRect().width) &&
                this.layout === 'auto');
        }
        catch (error) {
            return false;
        }
    }
    showPreviousArrow() {
        try {
            return this.showArrows() === true && this.scrollActionAmount < 0;
        }
        catch (error) {
            return false;
        }
    }
    showNextArrow() {
        try {
            const tabWrapper = this.getTabsWrapper();
            if (!tabWrapper) {
                return false;
            }
            const tabWrapperRect = tabWrapper.getBoundingClientRect();
            return (this.showArrows() === true &&
                this.scrollActionAmount >
                    (tabWrapper.scrollWidth - tabWrapperRect.width) * -1);
        }
        catch (error) {
            return false;
        }
    }
    move(amount, click = false) {
        const tabsWrapper = this.getTabsWrapper();
        if (!tabsWrapper) {
            return;
        }
        const tabsWrapperVisibleWidth = tabsWrapper.getBoundingClientRect().width;
        const maxScrollWidth = -this.currentScrollAmount +
            tabsWrapperVisibleWidth -
            tabsWrapper.scrollWidth;
        amount = amount < maxScrollWidth ? maxScrollWidth : amount;
        amount += this.currentScrollAmount;
        amount = Math.min(amount, 0);
        const styles = [
            `transform: translateX(${amount}px);`,
            click ? 'transition: all ease-in-out 400ms;' : '',
        ].join('');
        tabsWrapper.setAttribute('style', styles);
        if (click)
            this.currentScrollAmount = this.scrollActionAmount = amount;
        else
            this.scrollActionAmount = amount;
    }
    onSelectedChange(newValue) {
        if (!this.showArrows())
            return;
        const tab = this.getTab(newValue);
        const container = this.getTabsContainer();
        if (!tab || !container)
            return;
        const containerRect = container.getBoundingClientRect();
        const tabRect = tab.getBoundingClientRect();
        const tabLeftRelative = tabRect.left - containerRect.left;
        const tabRightRelative = tabLeftRelative + tabRect.width;
        if (tabLeftRelative < this.ARROW_WIDTH) {
            this.move(-tabLeftRelative + this.ARROW_WIDTH, true);
        }
        else if (tabRightRelative > containerRect.width - this.ARROW_WIDTH) {
            this.move(containerRect.width - tabRightRelative - this.ARROW_WIDTH, true);
        }
    }
    setSelected(index) {
        this.selected = index;
    }
    clickTab(index) {
        if (!this.clickAction.isClick || this.dragStop()) {
            return;
        }
        const { defaultPrevented } = this.selectedChange.emit(index);
        if (defaultPrevented) {
            return;
        }
        this.setSelected(index);
    }
    dragStart(element, event) {
        if (!this.showArrows())
            return;
        if (event.button > 0)
            return;
        this.clickAction.timeout =
            this.clickAction.timeout === null
                ? setTimeout(() => (this.clickAction.isClick = false), 300)
                : null;
        const tabPositionX = parseFloat(window.getComputedStyle(element).left);
        const mousedownPositionX = event.clientX;
        const move = (event) => this.dragMove(event, tabPositionX, mousedownPositionX);
        const windowClick = () => {
            window.removeEventListener('mousemove', move, false);
            window.removeEventListener('click', windowClick, false);
            this.dragStop();
        };
        window.addEventListener('click', windowClick);
        window.addEventListener('mousemove', move, false);
    }
    dragMove(event, tabX, mousedownX) {
        this.move(event.clientX + tabX - mousedownX);
    }
    dragStop() {
        if (this.clickAction.timeout) {
            clearTimeout(this.clickAction.timeout);
            this.clickAction.timeout = null;
        }
        if (this.clickAction.isClick)
            return false;
        this.currentScrollAmount = this.scrollActionAmount;
        this.clickAction.isClick = true;
        return true;
    }
    componentWillLoad() {
        this.initResizeObserver();
    }
    componentDidRender() {
        this.updateTabAttributes();
    }
    componentWillRender() {
        this.renderArrows();
    }
    renderArrows() {
        requestAnimationFrameNoNgZone(() => {
            this.showArrowNext = this.showNextArrow();
            this.showArrowPrevious = this.showPreviousArrow();
        });
    }
    componentDidLoad() {
        const tabs = this.getTabs();
        tabs.forEach((element) => {
            element.addEventListener('mousedown', (event) => this.dragStart(element, event));
        });
        this.observeSlotChanges();
    }
    disconnectedCallback() {
        var _a, _b;
        (_a = this.resizeObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
        (_b = this.classObserver) === null || _b === void 0 ? void 0 : _b.disconnect();
    }
    onTabClick(event) {
        if (event.defaultPrevented) {
            return;
        }
        const target = event.target;
        const tabs = this.getTabs();
        tabs.forEach((tab, index) => {
            if (!tab.disabled && tab === target) {
                this.clickTab(index);
            }
        });
    }
    render() {
        return (h(Host, { key: 'db27e336ea0432630306faa2e51f2799f1340d2c' }, this.showArrowPrevious && (h("button", { key: 'c9edd1bab0a829dd853a8dc227b59f068ea8b24a', class: "arrow", onClick: () => this.move(this.scrollAmount, true), "aria-label": this.ariaLabelChevronLeftIconButton }, h("ix-icon", { key: '6271e2d4ccbc8d94793b0226c7117c8f1a24561a', name: iconChevronLeftSmall }))), h("div", { key: 'd1ff277b75c488f25821296fc5a7320f822f1bbd', class: {
                'tab-items': true,
                'overflow-shadow': true,
                'shadow-left': this.showArrowPrevious,
                'shadow-right': this.showArrowNext,
                'shadow-both': this.showArrowNext && this.showArrowPrevious,
            } }, h("div", { key: '9f929a8ba8f76c0efe8604218df1fe207eeff9ff', class: "items-content" }, h("slot", { key: 'a906661489fdee6484c871c7c7c46f0f1c5153fe' }))), this.showArrowNext && (h("button", { key: '632c5d16531fe4071368a94cdef569e43fd2ec71', class: "arrow right", onClick: () => this.move(-this.scrollAmount, true), "aria-label": this.ariaLabelChevronRightIconButton }, h("ix-icon", { key: '7105649d79c2a5a6d2c8fd3e6e9a0db540a61767', name: iconChevronRightSmall })))));
    }
    static get is() { return "ix-tabs"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tabs.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tabs.css"]
        };
    }
    static get properties() {
        return {
            "small": {
                "type": "boolean",
                "attribute": "small",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Set tab items to small size"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "rounded": {
                "type": "boolean",
                "attribute": "rounded",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Set rounded tabs"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "selected": {
                "type": "number",
                "attribute": "selected",
                "mutable": true,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Set default selected tab by index"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "0"
            },
            "layout": {
                "type": "string",
                "attribute": "layout",
                "mutable": false,
                "complexType": {
                    "original": "'auto' | 'stretched'",
                    "resolved": "\"auto\" | \"stretched\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Set layout width style"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'auto'"
            },
            "placement": {
                "type": "string",
                "attribute": "placement",
                "mutable": false,
                "complexType": {
                    "original": "'bottom' | 'top'",
                    "resolved": "\"bottom\" | \"top\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Set placement style"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'bottom'"
            },
            "ariaLabelChevronLeftIconButton": {
                "type": "string",
                "attribute": "aria-label-chevron-left-icon-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label for the chevron left icon button"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "ariaLabelChevronRightIconButton": {
                "type": "string",
                "attribute": "aria-label-chevron-right-icon-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label for the chevron right icon button"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            }
        };
    }
    static get states() {
        return {
            "totalItems": {},
            "currentScrollAmount": {},
            "scrollAmount": {},
            "scrollActionAmount": {},
            "showArrowPrevious": {},
            "showArrowNext": {}
        };
    }
    static get events() {
        return [{
                "method": "selectedChange",
                "name": "selectedChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "`selected` property changed"
                },
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "hostElement"; }
    static get watchers() {
        return [{
                "propName": "selected",
                "methodName": "onSelectedChange"
            }];
    }
    static get listeners() {
        return [{
                "name": "resize",
                "method": "onWindowResize",
                "target": "window",
                "capture": false,
                "passive": true
            }, {
                "name": "tabClick",
                "method": "onTabClick",
                "target": undefined,
                "capture": false,
                "passive": false
            }];
    }
}
