// @siemens/ix@4.3.0/dist/collection/components/slider/slider.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
        r = Reflect.decorate(decorators, target, key, desc);
    else
        for (var i = decorators.length - 1; i >= 0; i--)
            if (d = decorators[i])
                r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { h, Host, } from "@stencil/core";
import { a11yHostAttributes } from "../utils/a11y";
import { OnListener } from "../utils/listener";
import { makeRef } from "../utils/make-ref";
import { HookValidationLifecycle, } from "../utils/input";
function between(min, value, max) {
    if (value < min) {
        return min;
    }
    else if (value > max) {
        return max;
    }
    else {
        return value;
    }
}
/**
 * @form-ready
 * @slot label-start - Element will be displayed at the start of the slider
 * @slot label-end - Element will be displayed at the end of the slider
 */
export class Slider {
    constructor() {
        /**
         * Show helper, info, warning, error and valid text as tooltip
         * @since 4.3.0
         */
        this.showTextAsTooltip = false;
        /**
         * Legal number intervals
         *
         * {@link https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/range#step}
         */
        this.step = 1;
        /**
         * Minimum slider value
         */
        this.min = 0;
        /**
         * Maximum slider value
         */
        this.max = 100;
        /**
         * Current value of the slider
         */
        this.value = 0;
        /**
         * Show a trace line
         */
        this.trace = false;
        /**
         * Define the start point of the trace line
         */
        this.traceReference = 0;
        /**
         * Show control as disabled
         */
        this.disabled = false;
        this.rangeInput = 0;
        this.rangeMin = 0;
        this.rangeMax = 100;
        this.rangeTraceReference = 0;
        this.showTooltip = false;
        this.isInvalid = false;
        this.isValid = false;
        this.isInfo = false;
        this.isWarning = false;
        this.touched = false;
        this.controlRef = makeRef();
        this.thumbRef = makeRef();
        this.tooltipRef = makeRef();
    }
    get hasLabels() {
        return !!this.hostElement.querySelector('[slot="label-start"], [slot="label-end"]');
    }
    get tooltip() {
        return this.tooltipRef.current;
    }
    get pseudoThumb() {
        return this.thumbRef.current;
    }
    get slider() {
        var _a;
        return (_a = this.hostElement.shadowRoot) === null || _a === void 0 ? void 0 : _a.getElementById('slider');
    }
    onShowTooltipChange() {
        var _a, _b;
        if (this.showTooltip && this.pseudoThumb) {
            (_a = this.tooltip) === null || _a === void 0 ? void 0 : _a.showTooltip(this.pseudoThumb);
            return;
        }
        (_b = this.tooltip) === null || _b === void 0 ? void 0 : _b.hideTooltip();
    }
    onClassField({ isInvalid, isInfo, isValid, isWarning, isInvalidByRequired, }) {
        this.isInvalid = isInvalid || isInvalidByRequired;
        this.isInfo = isInfo;
        this.isValid = isValid;
        this.isWarning = isWarning;
    }
    componentWillLoad() {
        this.a11yAttributes = a11yHostAttributes(this.hostElement, [
            'role',
            'aria-valuemin',
            'aria-valuemax',
            'aria-valuenow',
        ]);
        this.updateRangeVariables();
        this.setFormValueIfChanged(this.value);
    }
    updateRangeVariables() {
        this.rangeInput = between(this.min, this.value, this.max);
        this.rangeTraceReference = between(this.min, this.traceReference, this.max);
        this.rangeMin = Math.min(this.min, this.max);
        this.rangeMax = Math.max(this.min, this.max);
        if (this.value !== undefined) {
            this.setFormValueIfChanged(this.value);
        }
    }
    updateFormInternalValue(value) {
        this.setFormValueIfChanged(value);
        this.valueChange.emit(value);
    }
    setFormValueIfChanged(value) {
        const valueStr = value.toString();
        if (this.lastFormValue !== valueStr) {
            this.formInternals.setFormValue(valueStr);
            this.lastFormValue = valueStr;
        }
    }
    onInput(event) {
        event.stopPropagation();
        const value = parseFloat(this.slider.value);
        if (!isNaN(value)) {
            const oldValue = this.rangeInput;
            this.rangeInput = value;
            const { defaultPrevented } = this.emitInputEvent();
            if (defaultPrevented) {
                this.rangeInput = oldValue;
                this.slider.value = oldValue.toString();
            }
            else {
                this.updateFormInternalValue(value);
            }
        }
    }
    emitInputEvent() {
        return this.valueChange.emit(this.rangeInput);
    }
    isMarkerActive(markerValue) {
        const start = Math.min(this.traceReference, this.rangeInput);
        const end = Math.max(this.traceReference, this.rangeInput);
        const value = markerValue;
        return value >= start && value <= end;
    }
    // Listen globally on window because sometimes the event listener
    // of the DOM element input itself is not called if the release
    // click is not inside the element anymore
    onPointerUp() {
        this.showTooltip = false;
    }
    /** @internal */
    hasValidValue() {
        return Promise.resolve(true);
    }
    /** @internal */
    isTouched() {
        return Promise.resolve(this.touched);
    }
    render() {
        var _a;
        const range = this.rangeMax - this.rangeMin;
        let traceReferenceInPercentage = (this.rangeTraceReference - this.rangeMin) / range;
        let valueInPercentage = (this.rangeInput - this.rangeMin) / range;
        const distance = valueInPercentage - traceReferenceInPercentage;
        let traceStart = traceReferenceInPercentage;
        let traceEnd = valueInPercentage;
        if (distance <= 0) {
            traceStart = valueInPercentage;
            traceEnd = traceReferenceInPercentage;
        }
        return (h(Host, { key: 'dcce2da89192816ddc61f9f1950a399ebdf53766', class: {
                disabled: this.disabled,
                error: !!this.error,
                invalid: this.isInvalid,
                info: this.isInfo,
                valid: this.isValid,
                warning: this.isWarning,
            }, onPointerDown: () => setTimeout(() => (this.showTooltip = true)) }, h("ix-field-wrapper", { key: 'a3ce51bbc6f18cfbc99c3ef821016c6421217ffe', label: this.label, helperText: this.helperText, invalidText: this.invalidText, infoText: this.infoText, warningText: this.warningText, validText: this.validText, showTextAsTooltip: this.showTextAsTooltip, isValid: this.isValid, isInfo: this.isInfo, isWarning: this.isWarning, isInvalid: this.isInvalid, controlRef: this.controlRef }, h("div", { key: '602a467d24592696b8f5ebaf13f1b7d5e3451e38', class: "slider-container" }, h("div", { key: 'b9ded80de026dfc23f6892ef5fab71bcf9444011', class: "slider" }, h("div", { key: '78472ae8734b51390efeac850ffe2400e8fd67be', class: "track" }, h("div", { key: '0183018c9421d77caa857d6e8ff83dce0b7714f1', ref: this.thumbRef, class: "thumb", style: {
                left: `calc(${valueInPercentage} * (100% - 1rem))`,
            } }), h("svg", { key: '3a43071be5fb88e7ae8d0f8abd34f7cef203fb54', class: "ticks", xmlns: "http://www.w3.org/2000/svg" }, (_a = this.marker) === null || _a === void 0 ? void 0 : _a.filter((markerValue) => markerValue >= this.min && markerValue <= this.max).map((markerValue) => {
            const markerPosition = (markerValue - this.rangeMin) / range;
            return (h("circle", { class: {
                    tick: true,
                    'tick-active': this.isMarkerActive(markerValue) && this.trace,
                    'tick-at-min': markerPosition === 0,
                    'tick-at-max': markerPosition === 1,
                }, cx: "0", cy: "0", style: {
                    '--tick-value': `${markerPosition}`,
                } }));
        }))), h("input", Object.assign({ key: '1f1304342e72682bc040994e96297a31a61f4d00', id: "slider", ref: this.controlRef, type: "range", list: this.marker ? 'markers' : undefined, step: this.step, min: this.min, max: this.max, value: this.rangeInput, tabindex: this.disabled ? -1 : 0, onInput: (event) => this.onInput(event), onFocus: () => {
                this.showTooltip = true;
                this.touched = true;
            }, onBlur: () => {
                this.showTooltip = false;
            }, style: {
                '--value': `${valueInPercentage}`,
                '--trace-reference': `${traceReferenceInPercentage}`,
                '--trace-start': `${traceStart}`,
                '--trace-end': `${traceEnd}`,
            }, class: {
                trace: this.trace &&
                    traceReferenceInPercentage !== valueInPercentage,
                'hide-trace-reference': this.trace &&
                    (this.traceReference <= this.min ||
                        this.traceReference >= this.max),
            }, role: "slider", "aria-valuenow": this.rangeInput, "aria-valuemin": this.min, "aria-valuemax": this.max }, this.a11yAttributes)), h("ix-tooltip", { key: 'e0c3c6b098a2112cbcc27a90a8cdb696148109dc', ref: this.tooltipRef, class: {
                'hide-tooltip': !this.showTooltip,
            }, animationFrame: true, for: this.thumbRef.waitForCurrent() }, this.rangeInput)), this.hasLabels && (h("div", { key: '0a04945ac1055fbcd7e4b05b1188071e5e58180b', class: "label" }, h("div", { key: 'e1d1f0963b6e56947afeb9e371c1404b9b0fb819' }, h("slot", { key: '16e408212c6193e840f399343cfe2d3a8e260e30', name: "label-start" })), h("div", { key: '569530a9d3f987a0993ea8f79709b5f2252e0956' }, h("slot", { key: 'aba45e17633571d07a3dc95d46502dd6ebdc0aae', name: "label-end" })))), this.error ? (h("ix-typography", { textColor: "alarm" }, this.error)) : null))));
    }
    static get is() { return "ix-slider"; }
    static get encapsulation() { return "shadow"; }
    static get formAssociated() { return true; }
    static get originalStyleUrls() {
        return {
            "$": ["slider.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["slider.css"]
        };
    }
    static get properties() {
        return {
            "helperText": {
                "type": "string",
                "attribute": "helper-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.3.0"
                        }],
                    "text": "Show text below the field component"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "label": {
                "type": "string",
                "attribute": "label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.3.0"
                        }],
                    "text": "Label for the field component"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "invalidText": {
                "type": "string",
                "attribute": "invalid-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.3.0"
                        }],
                    "text": "Error text for the field component"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "infoText": {
                "type": "string",
                "attribute": "info-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.3.0"
                        }],
                    "text": "Info text for the field component"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "warningText": {
                "type": "string",
                "attribute": "warning-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.3.0"
                        }],
                    "text": "Warning text for the field component"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "validText": {
                "type": "string",
                "attribute": "valid-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.3.0"
                        }],
                    "text": "Valid text for the field component"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "showTextAsTooltip": {
                "type": "boolean",
                "attribute": "show-text-as-tooltip",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.3.0"
                        }],
                    "text": "Show helper, info, warning, error and valid text as tooltip"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "step": {
                "type": "number",
                "attribute": "step",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Legal number intervals\n\n{@link https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/range#step}"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "1"
            },
            "min": {
                "type": "number",
                "attribute": "min",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Minimum slider value"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "0"
            },
            "max": {
                "type": "number",
                "attribute": "max",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Maximum slider value"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "100"
            },
            "value": {
                "type": "number",
                "attribute": "value",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Current value of the slider"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "0"
            },
            "marker": {
                "type": "unknown",
                "attribute": "marker",
                "mutable": false,
                "complexType": {
                    "original": "SliderMarker",
                    "resolved": "number[] | undefined",
                    "references": {
                        "SliderMarker": {
                            "location": "import",
                            "path": "./slider.types",
                            "id": "src/components/slider/slider.types.ts::SliderMarker"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Define tick marker on the slider. Marker has to be within slider min/max"
                },
                "getter": false,
                "setter": false
            },
            "trace": {
                "type": "boolean",
                "attribute": "trace",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show a trace line"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "traceReference": {
                "type": "number",
                "attribute": "trace-reference",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Define the start point of the trace line"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "0"
            },
            "disabled": {
                "type": "boolean",
                "attribute": "disabled",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show control as disabled"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "error": {
                "type": "any",
                "attribute": "error",
                "mutable": false,
                "complexType": {
                    "original": "boolean | string",
                    "resolved": "boolean | string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "deprecated",
                            "text": "Will be removed in 5.0.0. Use invalid class instead."
                        }],
                    "text": "Show error state and message"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            }
        };
    }
    static get states() {
        return {
            "rangeInput": {},
            "rangeMin": {},
            "rangeMax": {},
            "rangeTraceReference": {},
            "showTooltip": {},
            "isInvalid": {},
            "isValid": {},
            "isInfo": {},
            "isWarning": {}
        };
    }
    static get events() {
        return [{
                "method": "valueChange",
                "name": "valueChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Will emit the value when it changes"
                },
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                }
            }];
    }
    static get methods() {
        return {
            "hasValidValue": {
                "complexType": {
                    "signature": "() => Promise<boolean>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<boolean>"
                },
                "docs": {
                    "text": "",
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }]
                }
            },
            "isTouched": {
                "complexType": {
                    "signature": "() => Promise<boolean>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<boolean>"
                },
                "docs": {
                    "text": "",
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }]
                }
            }
        };
    }
    static get elementRef() { return "hostElement"; }
    static get watchers() {
        return [{
                "propName": "showTooltip",
                "methodName": "onShowTooltipChange"
            }, {
                "propName": "value",
                "methodName": "updateRangeVariables"
            }, {
                "propName": "max",
                "methodName": "updateRangeVariables"
            }, {
                "propName": "min",
                "methodName": "updateRangeVariables"
            }, {
                "propName": "traceReference",
                "methodName": "updateRangeVariables"
            }];
    }
    static get attachInternalsMemberName() { return "formInternals"; }
}
__decorate([
    HookValidationLifecycle()
], Slider.prototype, "onClassField", null);
__decorate([
    OnListener('pointerup', (self) => self.showTooltip)
], Slider.prototype, "onPointerUp", null);
