// @siemens/ix/dist/collection/components/push-card/push-card.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host } from "@stencil/core";
export class PushCard {
    constructor() {
        /**
         * Card variant
         */
        this.variant = 'outline';
        /**
         * Expand the card
         */
        this.expanded = false;
        /**
         * If true, disables hover and active styles and changes cursor to default
         */
        this.passive = false;
    }
    render() {
        var _a;
        return (h(Host, { key: 'b5d6c4c34598670f4e8ae1228843b9afde2ef241', class: `card-${this.variant}` }, h("ix-card", { key: 'a62c8dfd617a8c141dd6ccf3c95da7877d44f7b7', variant: this.variant, passive: this.passive }, h("ix-card-content", { key: '4367ef2c91bf7b5ee9adef7c42b1a3f801b83bf9' }, h("ix-card-title", { key: 'e17700e10181981d14e90c83f7f7e400a618df6e' }, this.icon ? (h("ix-icon", { class: 'icon', name: this.icon, size: "32", "aria-label": this.ariaLabelIcon })) : null, h("ix-typography", { key: 'd10982fa32f721d09a4bb829dbf4ce7980ad1572', format: "display-xxl" }, (_a = this.notification) !== null && _a !== void 0 ? _a : 0), h("slot", { key: '2c5ba728f0ed723750fd3daa37c0c849ef791e52', name: "title-action" })), h("ix-typography", { key: 'b74801e57b670bdf5f1bfcfd3404502e118c00a5', format: "h4" }, this.heading), h("ix-typography", { key: '4869e7565d8e301551efae7ca8679bfb1df02d75' }, this.subheading)), h("ix-card-accordion", { key: '5fda212d77b6c8e1664d0c1a855b7b2e6a2ba26a', collapse: !this.expanded, variant: this.variant }, h("slot", { key: '9cfe767d08c2c49811d7d6f8ee1cba049606977b' })))));
    }
    static get is() { return "ix-push-card"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["push-card.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["push-card.css"]
        };
    }
    static get properties() {
        return {
            "icon": {
                "type": "string",
                "attribute": "icon",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Card icon"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "ariaLabelIcon": {
                "type": "string",
                "attribute": "aria-label-icon",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label for the icon"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "notification": {
                "type": "string",
                "attribute": "notification",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Card KPI value"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "heading": {
                "type": "string",
                "attribute": "heading",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Card heading"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "subheading": {
                "type": "string",
                "attribute": "subheading",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Card subheading"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "variant": {
                "type": "string",
                "attribute": "variant",
                "mutable": false,
                "complexType": {
                    "original": "PushCardVariant",
                    "resolved": "\"alarm\" | \"critical\" | \"filled\" | \"info\" | \"neutral\" | \"outline\" | \"primary\" | \"success\" | \"warning\"",
                    "references": {
                        "PushCardVariant": {
                            "location": "import",
                            "path": "./push-card.types",
                            "id": "src/components/push-card/push-card.types.ts::PushCardVariant"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Card variant"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'outline'"
            },
            "expanded": {
                "type": "boolean",
                "attribute": "expanded",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Expand the card"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "passive": {
                "type": "boolean",
                "attribute": "passive",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "If true, disables hover and active styles and changes cursor to default"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
}
