// @siemens/ix@4.3.0/components/index.js
import { s as setPlatformHelpers } from './p-D-__gyet.js';
export { g as getAssetPath, r as render, a as setAssetPath, b as setNonce, c as setPlatformOptions } from './p-D-__gyet.js';
export { I as InputState, L as LogicalFilterOperator } from './p-BH3f5fa3.js';
export { d as defaultRefreshTreeOptions, r as renderDefaultItem } from './p-DSTYytot.js';
export { U as UploadFileState } from './p-BGzrnl_l.js';
import { g as getCoreDelegate } from './p-D0T_1WDk.js';
export { c as closeModal, d as dismissModal, r as registerFrameworkDelegate, a as resolveDelegate, s as showMessage, b as showModal } from './p-D0T_1WDk.js';
import { A as Animation } from './p-C5MWUgDN.js';
export { T as TypedEvent } from './p-BdCnOrqW.js';
export { c as convertToAbbreviationString, a as convertToRemString } from './p-JJddxCCh.js';
export { g as getCurrentSystemAppearance, t as themeSwitcher } from './p-CsHYWHE6.js';

/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
function handlePlatformHelpers(config) {
    const platformHelpers = {};
    if (config.ael) {
        platformHelpers.ael = config.ael;
    }
    if (config.rel) {
        platformHelpers.rel = config.rel;
    }
    if (config.ce) {
        platformHelpers.ce = config.ce;
    }
    setPlatformHelpers(platformHelpers);
}

/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
class FilterState {
    constructor() {
        this.tokens = [];
        this.categories = [];
    }
}

/**
 * Gets the toast container.
 */
function getToastContainer() {
    const containerList = Array.from(document.querySelectorAll('ix-toast-container'));
    const [container] = containerList;
    if (containerList.length > 1) {
        console.warn('Multiple toast containers were found. Only the first one will be used.');
        return container;
    }
    if (!container) {
        const toastContainer = document.createElement('ix-toast-container');
        document.body.appendChild(toastContainer);
        return toastContainer;
    }
    return container;
}
/**
 * Sets the position of the toast container
 */
function setToastPosition(position) {
    const container = getToastContainer();
    container.setAttribute('position', position);
}
/**
 * Display a toast based on the provided configuration
 */
function toast(config) {
    const container = getToastContainer();
    return container.showToast(config);
}
/**
 * Display an info toast based on the provided configuration
 */
toast.info = (config) => {
    return toast(Object.assign(Object.assign({}, config), { type: 'info' }));
};
/**
 * Display an error toast based on the provided configuration
 */
toast.error = (config) => {
    return toast(Object.assign(Object.assign({}, config), { type: 'error' }));
};
/**
 * Display a success toast based on the provided configuration
 */
toast.success = (config) => {
    return toast(Object.assign(Object.assign({}, config), { type: 'success' }));
};
/**
 * Display a warning toast based on the provided configuration
 */
toast.warning = (config) => {
    return toast(Object.assign(Object.assign({}, config), { type: 'warning' }));
};

/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
/**
 * Displays a loading modal with a message
 */
function showModalLoading(message) {
    const modal = document.createElement('ix-modal');
    modal.disableEscapeClose = true;
    const loading = document.createElement('ix-modal-loading');
    loading.innerText = message;
    modal.appendChild(loading);
    getCoreDelegate().attachView(modal);
    modal.showModal();
    return {
        update: (text) => (loading.innerHTML = text),
        finish: (text, timeout = 250) => {
            if (text !== undefined) {
                loading.innerText = text;
            }
            else {
                timeout = 0;
            }
            setTimeout(() => {
                modal.closeModal(null);
                setTimeout(async () => await getCoreDelegate().removeView(modal), Animation.mediumTime);
            }, timeout);
        },
    };
}

export { FilterState, getCoreDelegate, getToastContainer, handlePlatformHelpers, setToastPosition, showModalLoading, toast };
