// @siemens/ix@4.3.0/components/ix-dropdown-quick-actions.js
import { p as proxyCustomElement, H, h, d as Host } from './p-D-__gyet.js';

const dropdownQuickActionsCss = ":host{display:flex;justify-content:center;align-items:center;margin-inline-start:1.5rem;margin-inline-end:1.5rem;margin-block-end:0.25rem}:host slot::slotted(*){display:flex;margin-inline-end:0.625rem}";

const DropdownQuickActions = /*@__PURE__*/ proxyCustomElement(class DropdownQuickActions extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
    }
    render() {
        return (h(Host, { key: '5b9938da726bfe31dffe1667ff81c1b3a5935e1e' }, h("slot", { key: 'a3f04dc61e5896ede45c0484506b3ee611b413c8' })));
    }
    static get style() { return dropdownQuickActionsCss; }
}, [257, "ix-dropdown-quick-actions"]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-dropdown-quick-actions"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-dropdown-quick-actions":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, DropdownQuickActions);
            }
            break;
    } });
}

const IxDropdownQuickActions = DropdownQuickActions;
const defineCustomElement = defineCustomElement$1;

export { IxDropdownQuickActions, defineCustomElement };
