// @siemens/ix@4.3.0/components/p-CGqmQsSj.js
import { p as proxyCustomElement, H, e as createEvent, h, d as Host } from './p-D-__gyet.js';
import { d as iconChevronRightSmall } from './p-BvsadYCu.js';

const treeItemCss = ":host{display:flex;align-items:center;height:32px;width:100%;cursor:pointer}:host *,:host *::after,:host *::before{box-sizing:border-box}:host ::-webkit-scrollbar-button{display:none}@-moz-document url-prefix(){:host *{scrollbar-color:var(--theme-scrollbar-thumb--background) var(--theme-scrollbar-track--background);scrollbar-width:thin}}:host{}:host ::-webkit-scrollbar{width:0.5rem;height:0.5rem}:host{}:host ::-webkit-scrollbar-track{border-radius:5px;background:var(--theme-scrollbar-track--background)}:host ::-webkit-scrollbar-track:hover{background:var(--theme-scrollbar-track--background--hover)}:host{}:host ::-webkit-scrollbar-thumb{border-radius:5px;background:var(--theme-scrollbar-thumb--background)}:host{}:host ::-webkit-scrollbar-thumb:hover{background:var(--theme-scrollbar-thumb--background--hover)}:host ::-webkit-scrollbar-corner{display:none}:host .tree-node-container{display:flex;align-items:center;height:2rem;flex-grow:1;align-items:center;overflow:hidden;text-overflow:ellipsis;white-space:nowrap}:host .tree-node-container .tree-node-text{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}:host .icon-toggle-container{display:flex;align-items:center;justify-content:center;width:2rem;height:2rem;min-width:2rem}:host .icon-toggle-container ix-icon{transition:transform var(--theme-default-time) ease-in-out}:host .icon-toggle-container ix-icon.icon-toggle-down{transform:rotate(90deg)}:host(:not(.disabled):not(:disabled):not(.selected).hover),:host(:not(.disabled):not(:disabled):not(.selected):hover){background-color:var(--theme-tree-item--background--hover)}:host(:not(.disabled):not(:disabled):not(.selected).active),:host(:not(.disabled):not(:disabled):not(.selected):active){background-color:var(--theme-tree-item--background--active)}:host(.selected){background-color:var(--theme-tree-item--background--selected)}:host(.selected.hover),:host(.selected:hover){background-color:var(--theme-tree-item--background--selected-hover)}:host(.selected.active),:host(.selected:active){background-color:var(--theme-tree-item--background--selected-active)}";

const TreeItem = /*@__PURE__*/ proxyCustomElement(class TreeItem extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
        this.toggle = createEvent(this, "toggle", 7);
        this.itemClick = createEvent(this, "itemClick", 7);
        /**
         * Has tree item children
         */
        this.hasChildren = false;
    }
    render() {
        var _a, _b;
        return (h(Host, { key: '0f615d72d15f5d10c9d44ea3bda96bb64494944a', class: {
                selected: !!((_a = this.context) === null || _a === void 0 ? void 0 : _a.isSelected),
            } }, h("div", { key: 'ea65a20109ff5347f1e6aa7da01f105f28594256', class: "icon-toggle-container" }, this.hasChildren ? (h("ix-icon", { name: iconChevronRightSmall, size: "24", class: {
                ['icon-toggle-down']: !!((_b = this.context) === null || _b === void 0 ? void 0 : _b.isExpanded),
            }, color: "color-std-text", onClick: (e) => {
                e.preventDefault();
                e.stopPropagation();
                this.toggle.emit();
            }, "aria-label": this.ariaLabelChevronIcon })) : null), h("div", { key: 'aba0519a308a75609a1339ec1d2ab7a6fcd4638e', class: "tree-node-container", onClick: () => {
                this.itemClick.emit();
            } }, h("div", { key: 'fa5985f1971f20cfb60f34e9c735c949e468ac4a', class: "tree-node-text" }, this.text), h("slot", { key: '9f1110d6d2d64af7cf2840fa429ccf7ed312e5fc' }))));
    }
    static get style() { return treeItemCss; }
}, [257, "ix-tree-item", {
        "text": [1],
        "hasChildren": [4, "has-children"],
        "context": [16],
        "ariaLabelChevronIcon": [1, "aria-label-chevron-icon"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-tree-item"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-tree-item":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, TreeItem);
            }
            break;
    } });
}

export { TreeItem as T, defineCustomElement as d };
