// @siemens/ix@4.3.0/components/p-D-nChuuk.js
import { p as proxyCustomElement, H, h, d as Host } from './p-D-__gyet.js';
import { m as makeRef } from './p-bcj7UEIC.js';
import { h as hasAnyText, H as HelperText } from './p-CrqgL2m1.js';
import { d as defineCustomElement$3 } from './p-BHD2m8UF.js';
import { d as defineCustomElement$2 } from './p-D8qc3Tjo.js';
import { d as defineCustomElement$1 } from './p-DBeR3Qpr.js';

const fieldWrapperCss = ":host{display:flex;position:relative;flex-direction:column}:host .slot-wrapper{display:flex;position:relative;align-items:center;justify-content:flex-start;gap:0.25rem}:host .field-bottom,:host .field-top{display:flex;flex-direction:row;position:relative;justify-content:space-between;gap:1rem;width:-moz-min-content;width:min-content;min-width:100%}:host .field-bottom .bottom-right{margin-left:auto;margin-right:0px}:host .bottom-text{display:flex;position:relative;align-items:flex-start;justify-content:flex-start;gap:0.25rem;margin-right:0.25rem}:host .text-icon{margin:0.125rem}:host .text-icon.invalid{color:var(--theme-helper-icon--color--invalid)}:host .text-icon.info{color:var(--theme-helper-icon--color--info)}:host .text-icon.warning{color:var(--theme-helper-icon--color--warning)}:host .text-icon.valid{color:var(--theme-helper-icon--color--valid)}:host .bottom-text{margin-top:0.25rem;margin-bottom:0.25rem}";

const FieldWrapper = /*@__PURE__*/ proxyCustomElement(class FieldWrapper extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
        /**
         * Is the field component invalid
         */
        this.isInvalid = false;
        /**
         * Is the field component valid
         */
        this.isValid = false;
        /**
         * Is the field component info
         */
        this.isInfo = false;
        /**
         * Is the field component warning
         */
        this.isWarning = false;
        /**
         * Show helper, error, info, warning text as tooltip
         */
        this.showTextAsTooltip = false;
        /**
         * Show label as required
         */
        this.required = false;
        this.slotRef = makeRef();
    }
    render() {
        const textOptions = {
            invalidText: this.invalidText,
            isInvalid: this.isInvalid,
            isValid: this.isValid,
            validText: this.validText,
            isWarning: this.isWarning,
            warningText: this.warningText,
            isInfo: this.isInfo,
            infoText: this.infoText,
            helperText: this.helperText,
        };
        return (h(Host, { key: '531b08942bf31c36043fb3d7cd3af489f2fe83b0' }, this.label && (h("div", { key: '49b9ef124fc2d69fb2528585d7b232b2e3d6d4b6', class: "field-top" }, h("ix-field-label", { key: '20a089a72de7ee3045c048c19d868ff3b4be2706', required: this.required, htmlFor: this.htmlForLabel, controlRef: this.controlRef, isInvalid: this.isInvalid }, this.label))), h("div", { key: '80c8ce58eb73379755757b2cef6b318cdc279b6b', class: {
                'slot-wrapper': true,
            }, ref: this.slotRef }, h("slot", { key: 'f41f0681742e9f783092bad7484c851e2b987c99' })), h("div", { key: '9db46e23621d508e08fe87cae69a3698491347dc', class: 'field-bottom' }, !this.showTextAsTooltip && h(HelperText, Object.assign({ key: '29802c92675a34cd215740a5df6fae77ed36368f' }, textOptions)), h("div", { key: 'b94a1cc2acd5f8379e768fda9f222af6cb85f2b7', class: "bottom-right" }, h("slot", { key: '3c5ba2d9a6d4671b6fadba6e06455ae229894f94', name: "bottom-right" }))), this.showTextAsTooltip === true && hasAnyText(textOptions) && (h("ix-tooltip", { key: '890c4a39781fbb14b96660691fd6f6fe96a30e0b', for: this.slotRef.waitForCurrent(), showDelay: 500, placement: "bottom" }, h(HelperText, Object.assign({ key: '638410fd78282cf38065635a79cceeb064a2a10e' }, textOptions))))));
    }
    get hostElement() { return this; }
    static get style() { return fieldWrapperCss; }
}, [257, "ix-field-wrapper", {
        "helperText": [1, "helper-text"],
        "label": [1],
        "invalidText": [1, "invalid-text"],
        "validText": [1, "valid-text"],
        "infoText": [1, "info-text"],
        "warningText": [1, "warning-text"],
        "isInvalid": [4, "is-invalid"],
        "isValid": [4, "is-valid"],
        "isInfo": [4, "is-info"],
        "isWarning": [4, "is-warning"],
        "showTextAsTooltip": [4, "show-text-as-tooltip"],
        "required": [4],
        "htmlForLabel": [1, "html-for-label"],
        "controlRef": [16, "control-ref"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-field-wrapper", "ix-field-label", "ix-tooltip", "ix-typography"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-field-wrapper":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, FieldWrapper);
            }
            break;
        case "ix-field-label":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "ix-tooltip":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
        case "ix-typography":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}

export { FieldWrapper as F, defineCustomElement as d };
