// @siemens/ix@4.3.0/dist/collection/components/field-wrapper/field-wrapper.js
/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Host, h } from "@stencil/core";
import { makeRef } from "../utils/make-ref";
import { hasAnyText, HelperText } from "./helper-text-util";
/** @internal */
export class FieldWrapper {
    constructor() {
        /**
         * Is the field component invalid
         */
        this.isInvalid = false;
        /**
         * Is the field component valid
         */
        this.isValid = false;
        /**
         * Is the field component info
         */
        this.isInfo = false;
        /**
         * Is the field component warning
         */
        this.isWarning = false;
        /**
         * Show helper, error, info, warning text as tooltip
         */
        this.showTextAsTooltip = false;
        /**
         * Show label as required
         */
        this.required = false;
        this.slotRef = makeRef();
    }
    render() {
        const textOptions = {
            invalidText: this.invalidText,
            isInvalid: this.isInvalid,
            isValid: this.isValid,
            validText: this.validText,
            isWarning: this.isWarning,
            warningText: this.warningText,
            isInfo: this.isInfo,
            infoText: this.infoText,
            helperText: this.helperText,
        };
        return (h(Host, { key: '531b08942bf31c36043fb3d7cd3af489f2fe83b0' }, this.label && (h("div", { key: '49b9ef124fc2d69fb2528585d7b232b2e3d6d4b6', class: "field-top" }, h("ix-field-label", { key: '20a089a72de7ee3045c048c19d868ff3b4be2706', required: this.required, htmlFor: this.htmlForLabel, controlRef: this.controlRef, isInvalid: this.isInvalid }, this.label))), h("div", { key: '80c8ce58eb73379755757b2cef6b318cdc279b6b', class: {
                'slot-wrapper': true,
            }, ref: this.slotRef }, h("slot", { key: 'f41f0681742e9f783092bad7484c851e2b987c99' })), h("div", { key: '9db46e23621d508e08fe87cae69a3698491347dc', class: 'field-bottom' }, !this.showTextAsTooltip && h(HelperText, Object.assign({ key: '29802c92675a34cd215740a5df6fae77ed36368f' }, textOptions)), h("div", { key: 'b94a1cc2acd5f8379e768fda9f222af6cb85f2b7', class: "bottom-right" }, h("slot", { key: '3c5ba2d9a6d4671b6fadba6e06455ae229894f94', name: "bottom-right" }))), this.showTextAsTooltip === true && hasAnyText(textOptions) && (h("ix-tooltip", { key: '890c4a39781fbb14b96660691fd6f6fe96a30e0b', for: this.slotRef.waitForCurrent(), showDelay: 500, placement: "bottom" }, h(HelperText, Object.assign({ key: '638410fd78282cf38065635a79cceeb064a2a10e' }, textOptions))))));
    }
    static get is() { return "ix-field-wrapper"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["field-wrapper.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["field-wrapper.css"]
        };
    }
    static get properties() {
        return {
            "helperText": {
                "type": "string",
                "attribute": "helper-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Show text below the field component"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "label": {
                "type": "string",
                "attribute": "label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Label for the field component"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "invalidText": {
                "type": "string",
                "attribute": "invalid-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Error text for the field component"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "validText": {
                "type": "string",
                "attribute": "valid-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Valid text for the field component"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "infoText": {
                "type": "string",
                "attribute": "info-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Info text for the field component"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "warningText": {
                "type": "string",
                "attribute": "warning-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Warning text for the field component"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "isInvalid": {
                "type": "boolean",
                "attribute": "is-invalid",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Is the field component invalid"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "isValid": {
                "type": "boolean",
                "attribute": "is-valid",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Is the field component valid"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "isInfo": {
                "type": "boolean",
                "attribute": "is-info",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Is the field component info"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "isWarning": {
                "type": "boolean",
                "attribute": "is-warning",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Is the field component warning"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "showTextAsTooltip": {
                "type": "boolean",
                "attribute": "show-text-as-tooltip",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show helper, error, info, warning text as tooltip"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "required": {
                "type": "boolean",
                "attribute": "required",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show label as required"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "htmlForLabel": {
                "type": "string",
                "attribute": "html-for-label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The id of the form element that the label is associated with"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "controlRef": {
                "type": "unknown",
                "attribute": "control-ref",
                "mutable": false,
                "complexType": {
                    "original": "| MakeRef<HTMLElement>\n    | MakeRef<HTMLInputElement>\n    | MakeRef<HTMLTextAreaElement>",
                    "resolved": "undefined | { (ref: HTMLElement | null | undefined): void; current: HTMLElement | null; waitForCurrent(): Promise<HTMLElement>; } | { (ref: HTMLInputElement | null | undefined): void; current: HTMLInputElement | null; waitForCurrent(): Promise<HTMLInputElement>; } | { (ref: HTMLTextAreaElement | null | undefined): void; current: HTMLTextAreaElement | null; waitForCurrent(): Promise<HTMLTextAreaElement>; }",
                    "references": {
                        "MakeRef": {
                            "location": "import",
                            "path": "../utils/make-ref",
                            "id": "src/components/utils/make-ref.ts::MakeRef"
                        },
                        "HTMLElement": {
                            "location": "global",
                            "id": "global::HTMLElement"
                        },
                        "HTMLInputElement": {
                            "location": "global",
                            "id": "global::HTMLInputElement"
                        },
                        "HTMLTextAreaElement": {
                            "location": "global",
                            "id": "global::HTMLTextAreaElement"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The control element that the label is associated with"
                },
                "getter": false,
                "setter": false
            }
        };
    }
    static get elementRef() { return "hostElement"; }
}
