// @siemens/ix@2.7.0/dist/cjs/ix-split-button-item.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-757233c7.js');
const makeRef = require('./make-ref-499e1e41.js');

const splitButtonItemCss = ":host{display:contents}";
const IxSplitButtonItemStyle0 = splitButtonItemCss;

const SplitButtonItem = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.itemClick = index.createEvent(this, "itemClick", 7);
        this.wrapperRef = makeRef.makeRef();
        this.icon = undefined;
        this.label = undefined;
    }
    /** @internal */
    async getDropdownItemElement() {
        return this.wrapperRef.waitForCurrent();
    }
    render() {
        return (index.h(index.Host, { key: 'cb34ba65f519984cf68639856699df8471d82f4c' }, index.h("ix-dropdown-item", { key: '54a95bd00935e3246e4d5bf8f9cdbc8da5d2cd82', ref: this.wrapperRef, suppressChecked: true, icon: this.icon, label: this.label, onItemClick: (e) => {
                e.preventDefault();
                e.stopPropagation();
            }, onClick: (e) => this.itemClick.emit(e) })));
    }
    get hostElement() { return index.getElement(this); }
};
SplitButtonItem.style = IxSplitButtonItemStyle0;

exports.ix_split_button_item = SplitButtonItem;
