// @siemens/ix@4.3.0/dist/collection/components/menu-item/menu-item.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { iconDocument } from "@siemens/ix-icons/icons";
import { h, Host } from "@stencil/core";
import { a11yBoolean, a11yHostAttributes } from "../utils/a11y";
import { makeRef } from "../utils/make-ref";
import { menuController } from "../utils/menu-service/menu-service";
import { createMutationObserver } from "../utils/mutation-observer";
import { createSequentialId } from "../utils/uuid";
let sequenceId = 0;
/**
 * @slot menu-item-label Custom label
 */
export class MenuItem {
    constructor() {
        /**
         * Move the Tab to a top position.
         */
        this.home = false;
        /**
         * Caution: this is no longer working. Please use slot="bottom" instead.
         *
         * Place tab on bottom
         */
        this.bottom = false;
        /**
         * State to display active
         */
        this.active = false;
        /**
         * Disable tab and remove event handlers
         */
        this.disabled = false;
        /**
         * Specifies where to open the linked document when href is provided.
         *
         * @since 4.0.0
         */
        this.target = '_self';
        /** @internal */
        this.isCategory = false;
        this.ariaHiddenTooltip = false;
        this.menuExpanded = false;
        this.internalItemId = createSequentialId('ix-menu-item-', sequenceId++);
        this.buttonRef = makeRef();
        this.isHostedInsideCategory = false;
        this.observer = createMutationObserver(() => {
            this.setTooltip();
        });
    }
    componentWillLoad() {
        var _a;
        this.isHostedInsideCategory =
            !!this.hostElement.closest('ix-menu-category');
        this.onIconChange();
        this.menuExpanded = ((_a = menuController.nativeElement) === null || _a === void 0 ? void 0 : _a.expand) || false;
        this.menuExpandedDisposer = menuController.expandChange.on((expand) => (this.menuExpanded = expand));
    }
    componentWillRender() {
        this.setTooltip();
    }
    setTooltip() {
        var _a, _b, _c;
        this.tooltip =
            (_c = (_b = (_a = this.tooltipText) !== null && _a !== void 0 ? _a : this.label) !== null && _b !== void 0 ? _b : this.hostElement.textContent) !== null && _c !== void 0 ? _c : undefined;
        this.ariaHiddenTooltip =
            this.tooltipText === this.label ||
                this.tooltipText === this.hostElement.textContent;
    }
    connectedCallback() {
        this.observer.observe(this.hostElement, {
            subtree: true,
            childList: true,
            characterData: true,
        });
    }
    disconnectedCallback() {
        if (this.observer) {
            this.observer.disconnect();
        }
        if (this.menuExpandedDisposer) {
            this.menuExpandedDisposer.dispose();
        }
    }
    onIconChange() {
        if (!this.isHostedInsideCategory && !this.hostElement.icon) {
            this.icon = iconDocument;
        }
    }
    render() {
        let extendedAttributes = {};
        if (this.home) {
            extendedAttributes = {
                slot: 'home',
            };
        }
        if (this.bottom) {
            extendedAttributes = {
                slot: 'bottom',
            };
        }
        const hostA11y = a11yHostAttributes(this.hostElement);
        const commonAttributes = Object.assign({ class: 'tab', tabIndex: this.disabled ? -1 : 0 }, hostA11y);
        const menuContent = [
            this.icon && (h("ix-icon", { key: '3235cdce2773d33e6f1a75b4df395079246fa993', class: 'tab-icon', name: this.icon, "aria-hidden": "true" })),
            this.notifications ? (h("div", { class: "notification" }, h("div", { class: "pill" }, this.notifications))) : null,
            h("span", { key: '704284ed9ae6b83b2b2d65ebc103e5e8ff766c1b', id: this.internalItemId, class: "tab-text text-default" }, this.label, h("slot", { key: '6b79e39fb7e616088d11bd8b9c57f4406fdd94b6' })),
        ];
        return (h(Host, Object.assign({ key: '331694d5c502b1ed77035e591ffa5748777d8a81', class: {
                disabled: this.disabled,
                'home-tab': this.home,
                'bottom-tab': this.bottom,
                active: this.active,
                'tab-nested': this.isHostedInsideCategory,
            }, "aria-disabled": this.disabled ? 'true' : null }, extendedAttributes), this.href ? (h("a", Object.assign({}, commonAttributes, { href: this.disabled ? undefined : this.href, target: this.target, rel: this.rel, role: "button", ref: this.buttonRef, onClick: (e) => {
                if (this.disabled) {
                    e.preventDefault();
                    e.stopPropagation();
                }
            } }), menuContent)) : (h("button", Object.assign({}, commonAttributes, { ref: this.buttonRef }), menuContent)), h("ix-tooltip", { key: 'd282a68bce72435079cbfea5234b227c92027643', for: this.buttonRef.waitForCurrent(), placement: 'right', showDelay: 1000, interactive: false, "aria-hidden": a11yBoolean(this.ariaHiddenTooltip), "aria-labelledby": this.internalItemId }, this.tooltip)));
    }
    static get is() { return "ix-menu-item"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["menu-item.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["menu-item.css"]
        };
    }
    static get properties() {
        return {
            "label": {
                "type": "string",
                "attribute": "label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Label of the menu item. Will also be used as tooltip text"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "home": {
                "type": "boolean",
                "attribute": "home",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Move the Tab to a top position."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "bottom": {
                "type": "boolean",
                "attribute": "bottom",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Caution: this is no longer working. Please use slot=\"bottom\" instead.\n\nPlace tab on bottom"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "icon": {
                "type": "string",
                "attribute": "icon",
                "mutable": true,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Name of the icon you want to display. Icon names can be resolved from the documentation {@link https://ix.siemens.io/docs/icon-library/icons}"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "notifications": {
                "type": "number",
                "attribute": "notifications",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Show notification count on tab"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "active": {
                "type": "boolean",
                "attribute": "active",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "State to display active"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "disabled": {
                "type": "boolean",
                "attribute": "disabled",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Disable tab and remove event handlers"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "tooltipText": {
                "type": "string",
                "attribute": "tooltip-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "Will be shown as tooltip text, if not provided menu text content will be used."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "href": {
                "type": "string",
                "attribute": "href",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "URL for the button link. When provided, the button will render as an anchor tag."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "target": {
                "type": "string",
                "attribute": "target",
                "mutable": false,
                "complexType": {
                    "original": "AnchorTarget",
                    "resolved": "\"_blank\" | \"_parent\" | \"_self\" | \"_top\" | undefined",
                    "references": {
                        "AnchorTarget": {
                            "location": "import",
                            "path": "../button/button.interface",
                            "id": "src/components/button/button.interface.ts::AnchorTarget"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "Specifies where to open the linked document when href is provided."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'_self'"
            },
            "rel": {
                "type": "string",
                "attribute": "rel",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "Specifies the relationship between the current document and the linked document when href is provided."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "isCategory": {
                "type": "boolean",
                "attribute": "is-category",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": ""
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get states() {
        return {
            "tooltip": {},
            "ariaHiddenTooltip": {},
            "menuExpanded": {}
        };
    }
    static get elementRef() { return "hostElement"; }
    static get watchers() {
        return [{
                "propName": "icon",
                "methodName": "onIconChange"
            }];
    }
}
