// @siemens/ix@4.3.0/components/ix-key-value-list.js
import { p as proxyCustomElement, H, h, d as Host } from './p-D-__gyet.js';

const keyValueListCss = ":host(.keyValueList) ::slotted(ix-key-value){border-bottom:1px solid var(--theme-color-weak-bdr)}:host(.keyValueList.keyValueList--striped) ::slotted(ix-key-value:nth-child(odd)){background:var(--theme-color-ghost-alt)}";

const KeyValueList = /*@__PURE__*/ proxyCustomElement(class KeyValueList extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
        /**
         * Optional striped key value list style
         */
        this.striped = false;
    }
    render() {
        return (h(Host, { key: '8ac8c1edacddbe11072e692c165c338bdaa6fcdf', class: { keyValueList: true, 'keyValueList--striped': this.striped } }, h("slot", { key: '844c573286e6022f4d3f53086b584425306f7f1e' })));
    }
    static get style() { return keyValueListCss; }
}, [257, "ix-key-value-list", {
        "striped": [4]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-key-value-list"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-key-value-list":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, KeyValueList);
            }
            break;
    } });
}

const IxKeyValueList = KeyValueList;
const defineCustomElement = defineCustomElement$1;

export { IxKeyValueList, defineCustomElement };
