// @siemens/ix@4.3.0/components/p-B09CzVld.js
import { p as proxyCustomElement, H, e as createEvent, h, d as Host } from './p-D-__gyet.js';
import { r as requestAnimationFrameNoNgZone } from './p-BEuV0Xpe.js';
import { d as iconChevronRightSmall, m as iconChevronLeftSmall } from './p-BvsadYCu.js';

const tabsCss = ":host{width:auto;display:flex;align-items:center;position:relative}:host *,:host *::after,:host *::before{box-sizing:border-box}:host ::-webkit-scrollbar-button{display:none}@-moz-document url-prefix(){:host *{scrollbar-color:var(--theme-scrollbar-thumb--background) var(--theme-scrollbar-track--background);scrollbar-width:thin}}:host{}:host ::-webkit-scrollbar{width:0.5rem;height:0.5rem}:host{}:host ::-webkit-scrollbar-track{border-radius:5px;background:var(--theme-scrollbar-track--background)}:host ::-webkit-scrollbar-track:hover{background:var(--theme-scrollbar-track--background--hover)}:host{}:host ::-webkit-scrollbar-thumb{border-radius:5px;background:var(--theme-scrollbar-thumb--background)}:host{}:host ::-webkit-scrollbar-thumb:hover{background:var(--theme-scrollbar-thumb--background--hover)}:host ::-webkit-scrollbar-corner{display:none}:host .tab-items{overflow:hidden;scroll-behavior:smooth;width:100%}:host .tab-items .items-content{display:flex;align-items:center}:host .arrow{all:unset;position:absolute;display:flex;align-items:center;justify-content:center;width:32px;height:32px;top:0;bottom:0;left:0;margin:auto 0;border-radius:4px;color:var(--theme-btn-tertiary--color);background-color:var(--theme-btn-tertiary--background);z-index:2}:host .arrow:hover{color:var(--theme-btn-tertiary--color--hover);background-color:var(--theme-btn-tertiary--background--hover)}:host .arrow:active{color:var(--theme-btn-tertiary--color--active);background-color:var(--theme-btn-tertiary--background--active)}:host .arrow.right{left:auto;right:0}:host .overflow-shadow{display:block;position:relative;height:100%;width:100%;pointer-events:all}:host .overflow-shadow.shadow-left{-webkit-mask-image:linear-gradient(90deg, transparent 0px, black 45px);mask-image:linear-gradient(90deg, transparent 0px, black 45px)}:host .overflow-shadow.shadow-right{-webkit-mask-image:linear-gradient(90deg, black calc(100% - 45px), transparent 100%);mask-image:linear-gradient(90deg, black calc(100% - 45px), transparent 100%)}:host .overflow-shadow.shadow-both{-webkit-mask-image:linear-gradient(90deg, transparent 0px, black 45px, black calc(100% - 45px), transparent 100%);mask-image:linear-gradient(90deg, transparent 0px, black 45px, black calc(100% - 45px), transparent 100%)}";

const TAB_MANAGED_CLASSES = {
    SELECTED: 'selected',
    DISABLED: 'disabled',
    SMALL_TAB: 'small-tab',
    ICON: 'icon',
    STRETCHED: 'stretched',
    BOTTOM: 'bottom',
    TOP: 'top',
    CIRCLE: 'circle',
    HYDRATED: 'hydrated',
};
const MANAGED_CLASSES_SET = new Set(Object.values(TAB_MANAGED_CLASSES));
const Tabs = /*@__PURE__*/ proxyCustomElement(class Tabs extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
        this.selectedChange = createEvent(this, "selectedChange", 7);
        /**
         * Set tab items to small size
         */
        this.small = false;
        /**
         * Set rounded tabs
         */
        this.rounded = false;
        /**
         * Set default selected tab by index
         */
        this.selected = 0;
        /**
         * Set layout width style
         */
        this.layout = 'auto';
        /**
         * Set placement style
         */
        this.placement = 'bottom';
        this.totalItems = 0;
        this.currentScrollAmount = 0;
        this.scrollAmount = 100;
        this.scrollActionAmount = 0;
        this.showArrowPrevious = false;
        this.showArrowNext = false;
        this.windowStartSize = window.innerWidth;
        this.ARROW_WIDTH = 32;
        this.updateScheduled = false;
        this.clickAction = {
            timeout: null,
            isClick: true,
        };
    }
    onWindowResize() {
        this.totalItems = 0;
        this.totalItems = this.getTabs().length;
        if (this.windowStartSize === 0)
            return (this.windowStartSize = window.innerWidth);
        this.move(this.windowStartSize - window.innerWidth);
        this.windowStartSize = window.innerWidth;
    }
    getTabs() {
        return Array.from(this.hostElement.querySelectorAll('ix-tab-item'));
    }
    getTab(tabIndex) {
        return this.getTabs()[tabIndex];
    }
    getTabsWrapper() {
        var _a;
        return (_a = this.hostElement.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('.items-content');
    }
    getTabsContainer() {
        var _a;
        return (_a = this.hostElement.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('.tab-items');
    }
    initResizeObserver() {
        const parentElement = this.hostElement.parentElement;
        if (!parentElement)
            return;
        this.resizeObserver = new ResizeObserver(() => {
            this.renderArrows();
        });
        this.resizeObserver.observe(parentElement);
    }
    observeSlotChanges() {
        var _a;
        (_a = this.classObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
        this.classObserver = new MutationObserver(() => {
            this.scheduleTabUpdate();
        });
        this.classObserver.observe(this.hostElement, {
            childList: true,
            subtree: true,
            attributes: true,
            attributeFilter: ['class'],
        });
    }
    scheduleTabUpdate() {
        if (this.updateScheduled)
            return;
        this.updateScheduled = true;
        requestAnimationFrame(() => {
            this.updateTabAttributes();
            this.updateScheduled = false;
        });
    }
    setTabAttributes(element, index) {
        const isSelected = index === this.selected;
        const isDisabled = element.disabled;
        if (this.small)
            element.setAttribute('small', 'true');
        if (this.rounded)
            element.setAttribute('rounded', 'true');
        element.setAttribute('layout', this.layout);
        element.setAttribute('selected', isSelected.toString());
        element.setAttribute('placement', this.placement);
        element.toggleAttribute('disabled', isDisabled);
        this.applyRequiredClasses(element, isSelected, isDisabled);
    }
    applyRequiredClasses(element, isSelected, isDisabled) {
        const requiredClasses = new Set(this.buildRequiredClasses(isSelected, isDisabled));
        const { classList } = element;
        for (const cls of requiredClasses) {
            classList.add(cls);
        }
        for (const managedClass of MANAGED_CLASSES_SET) {
            if (!requiredClasses.has(managedClass)) {
                classList.remove(managedClass);
            }
        }
    }
    buildRequiredClasses(isSelected, isDisabled) {
        const classConditions = {
            [TAB_MANAGED_CLASSES.HYDRATED]: true,
            [TAB_MANAGED_CLASSES.SELECTED]: isSelected,
            [TAB_MANAGED_CLASSES.DISABLED]: isDisabled,
            [TAB_MANAGED_CLASSES.SMALL_TAB]: this.small,
            [TAB_MANAGED_CLASSES.STRETCHED]: this.layout === 'stretched',
            [TAB_MANAGED_CLASSES.BOTTOM]: this.placement === 'bottom',
            [TAB_MANAGED_CLASSES.TOP]: this.placement === 'top',
            [TAB_MANAGED_CLASSES.CIRCLE]: this.rounded,
        };
        return Object.entries(classConditions)
            .filter(([, condition]) => condition)
            .map(([className]) => className);
    }
    ensureSelectedIndex() {
        if (this.totalItems === 0) {
            console.warn('ix-tabs: No tabs available for selection');
            this.selected = -1;
            return;
        }
        if (this.selected < this.totalItems) {
            return;
        }
        const originalIndex = this.selected;
        const previousIndex = originalIndex - 1;
        if (previousIndex >= 0 && previousIndex < this.totalItems) {
            this.updateSelected(previousIndex);
            return;
        }
        if (this.totalItems > 0) {
            this.updateSelected(0);
        }
    }
    updateSelected(index) {
        this.selected = index;
        this.selectedChange.emit(index);
    }
    updateTabAttributes() {
        const tabs = this.getTabs();
        this.totalItems = tabs.length;
        this.ensureSelectedIndex();
        for (const [index, element] of tabs.entries()) {
            this.setTabAttributes(element, index);
        }
        this.renderArrows();
    }
    showArrows() {
        try {
            const tabWrapper = this.getTabsWrapper();
            return (tabWrapper &&
                tabWrapper.scrollWidth >
                    Math.ceil(tabWrapper.getBoundingClientRect().width) &&
                this.layout === 'auto');
        }
        catch (error) {
            return false;
        }
    }
    showPreviousArrow() {
        try {
            return this.showArrows() === true && this.scrollActionAmount < 0;
        }
        catch (error) {
            return false;
        }
    }
    showNextArrow() {
        try {
            const tabWrapper = this.getTabsWrapper();
            if (!tabWrapper) {
                return false;
            }
            const tabWrapperRect = tabWrapper.getBoundingClientRect();
            return (this.showArrows() === true &&
                this.scrollActionAmount >
                    (tabWrapper.scrollWidth - tabWrapperRect.width) * -1);
        }
        catch (error) {
            return false;
        }
    }
    move(amount, click = false) {
        const tabsWrapper = this.getTabsWrapper();
        if (!tabsWrapper) {
            return;
        }
        const tabsWrapperVisibleWidth = tabsWrapper.getBoundingClientRect().width;
        const maxScrollWidth = -this.currentScrollAmount +
            tabsWrapperVisibleWidth -
            tabsWrapper.scrollWidth;
        amount = amount < maxScrollWidth ? maxScrollWidth : amount;
        amount += this.currentScrollAmount;
        amount = Math.min(amount, 0);
        const styles = [
            `transform: translateX(${amount}px);`,
            click ? 'transition: all ease-in-out 400ms;' : '',
        ].join('');
        tabsWrapper.setAttribute('style', styles);
        if (click)
            this.currentScrollAmount = this.scrollActionAmount = amount;
        else
            this.scrollActionAmount = amount;
    }
    onSelectedChange(newValue) {
        if (!this.showArrows())
            return;
        const tab = this.getTab(newValue);
        const container = this.getTabsContainer();
        if (!tab || !container)
            return;
        const containerRect = container.getBoundingClientRect();
        const tabRect = tab.getBoundingClientRect();
        const tabLeftRelative = tabRect.left - containerRect.left;
        const tabRightRelative = tabLeftRelative + tabRect.width;
        if (tabLeftRelative < this.ARROW_WIDTH) {
            this.move(-tabLeftRelative + this.ARROW_WIDTH, true);
        }
        else if (tabRightRelative > containerRect.width - this.ARROW_WIDTH) {
            this.move(containerRect.width - tabRightRelative - this.ARROW_WIDTH, true);
        }
    }
    setSelected(index) {
        this.selected = index;
    }
    clickTab(index) {
        if (!this.clickAction.isClick || this.dragStop()) {
            return;
        }
        const { defaultPrevented } = this.selectedChange.emit(index);
        if (defaultPrevented) {
            return;
        }
        this.setSelected(index);
    }
    dragStart(element, event) {
        if (!this.showArrows())
            return;
        if (event.button > 0)
            return;
        this.clickAction.timeout =
            this.clickAction.timeout === null
                ? setTimeout(() => (this.clickAction.isClick = false), 300)
                : null;
        const tabPositionX = parseFloat(window.getComputedStyle(element).left);
        const mousedownPositionX = event.clientX;
        const move = (event) => this.dragMove(event, tabPositionX, mousedownPositionX);
        const windowClick = () => {
            window.removeEventListener('mousemove', move, false);
            window.removeEventListener('click', windowClick, false);
            this.dragStop();
        };
        window.addEventListener('click', windowClick);
        window.addEventListener('mousemove', move, false);
    }
    dragMove(event, tabX, mousedownX) {
        this.move(event.clientX + tabX - mousedownX);
    }
    dragStop() {
        if (this.clickAction.timeout) {
            clearTimeout(this.clickAction.timeout);
            this.clickAction.timeout = null;
        }
        if (this.clickAction.isClick)
            return false;
        this.currentScrollAmount = this.scrollActionAmount;
        this.clickAction.isClick = true;
        return true;
    }
    componentWillLoad() {
        this.initResizeObserver();
    }
    componentDidRender() {
        this.updateTabAttributes();
    }
    componentWillRender() {
        this.renderArrows();
    }
    renderArrows() {
        requestAnimationFrameNoNgZone(() => {
            this.showArrowNext = this.showNextArrow();
            this.showArrowPrevious = this.showPreviousArrow();
        });
    }
    componentDidLoad() {
        const tabs = this.getTabs();
        tabs.forEach((element) => {
            element.addEventListener('mousedown', (event) => this.dragStart(element, event));
        });
        this.observeSlotChanges();
    }
    disconnectedCallback() {
        var _a, _b;
        (_a = this.resizeObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
        (_b = this.classObserver) === null || _b === void 0 ? void 0 : _b.disconnect();
    }
    onTabClick(event) {
        if (event.defaultPrevented) {
            return;
        }
        const target = event.target;
        const tabs = this.getTabs();
        tabs.forEach((tab, index) => {
            if (!tab.disabled && tab === target) {
                this.clickTab(index);
            }
        });
    }
    render() {
        return (h(Host, { key: 'db27e336ea0432630306faa2e51f2799f1340d2c' }, this.showArrowPrevious && (h("button", { key: 'c9edd1bab0a829dd853a8dc227b59f068ea8b24a', class: "arrow", onClick: () => this.move(this.scrollAmount, true), "aria-label": this.ariaLabelChevronLeftIconButton }, h("ix-icon", { key: '6271e2d4ccbc8d94793b0226c7117c8f1a24561a', name: iconChevronLeftSmall }))), h("div", { key: 'd1ff277b75c488f25821296fc5a7320f822f1bbd', class: {
                'tab-items': true,
                'overflow-shadow': true,
                'shadow-left': this.showArrowPrevious,
                'shadow-right': this.showArrowNext,
                'shadow-both': this.showArrowNext && this.showArrowPrevious,
            } }, h("div", { key: '9f929a8ba8f76c0efe8604218df1fe207eeff9ff', class: "items-content" }, h("slot", { key: 'a906661489fdee6484c871c7c7c46f0f1c5153fe' }))), this.showArrowNext && (h("button", { key: '632c5d16531fe4071368a94cdef569e43fd2ec71', class: "arrow right", onClick: () => this.move(-this.scrollAmount, true), "aria-label": this.ariaLabelChevronRightIconButton }, h("ix-icon", { key: '7105649d79c2a5a6d2c8fd3e6e9a0db540a61767', name: iconChevronRightSmall })))));
    }
    get hostElement() { return this; }
    static get watchers() { return {
        "selected": ["onSelectedChange"]
    }; }
    static get style() { return tabsCss; }
}, [257, "ix-tabs", {
        "small": [4],
        "rounded": [4],
        "selected": [1026],
        "layout": [1],
        "placement": [1],
        "ariaLabelChevronLeftIconButton": [1, "aria-label-chevron-left-icon-button"],
        "ariaLabelChevronRightIconButton": [1, "aria-label-chevron-right-icon-button"],
        "totalItems": [32],
        "currentScrollAmount": [32],
        "scrollAmount": [32],
        "scrollActionAmount": [32],
        "showArrowPrevious": [32],
        "showArrowNext": [32]
    }, [[9, "resize", "onWindowResize"], [0, "tabClick", "onTabClick"]], {
        "selected": ["onSelectedChange"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-tabs"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-tabs":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, Tabs);
            }
            break;
    } });
}

export { Tabs as T, defineCustomElement as d };
