// @siemens/ix@4.3.0/components/p-BHD2m8UF.js
import { p as proxyCustomElement, H, h, d as Host } from './p-D-__gyet.js';
import { b as a11yHostAttributes } from './p-Bb7pDeaQ.js';
import { m as makeRef } from './p-bcj7UEIC.js';
import { d as defineCustomElement$1 } from './p-DBeR3Qpr.js';
import { c as createClassMutationObserver } from './p-Dq_H2flK.js';

function isIxInputFieldComponent(obj) {
    return (obj &&
        'getAssociatedFormElement' in obj &&
        typeof obj.getAssociatedFormElement === 'function' &&
        'getNativeInputElement' in obj &&
        typeof obj.getNativeInputElement === 'function');
}

const fieldLabelCss = ":host{display:inline-block;position:relative;margin-top:0.5rem;margin-bottom:0.25rem}";

const FormFieldLabel = /*@__PURE__*/ proxyCustomElement(class FormFieldLabel extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
        /** @internal */
        this.isInvalid = false;
        this.explicitIsInvalid = undefined;
        this.htmlForObserver = new MutationObserver(() => this.checkForInternalState());
        this.a11yAttributes = {};
        this.labelRef = makeRef();
    }
    isInvalidChanged(newValue) {
        this.explicitIsInvalid = newValue;
    }
    connectedCallback() {
        this.registerHtmlForObserver();
        this.registerControlRefObserver();
    }
    disconnectedCallback() {
        if (this.htmlForObserver) {
            this.htmlForObserver.disconnect();
        }
        if (this.htmlForClassObserver) {
            this.htmlForClassObserver.destroy();
        }
        if (this.controlRefClassObserver) {
            this.controlRefClassObserver.destroy();
        }
    }
    componentWillRender() {
        this.checkForInternalState();
    }
    componentWillLoad() {
        this.a11yAttributes = a11yHostAttributes(this.hostElement);
    }
    registerHtmlForObserver() {
        if (typeof window === 'undefined') {
            return;
        }
        if (this.htmlForObserver) {
            this.htmlForObserver.disconnect();
        }
        if (this.htmlFor) {
            this.htmlForObserver.observe(window.document, {
                childList: true,
                subtree: true,
            });
        }
    }
    async registerControlRefObserver() {
        if (typeof window === 'undefined') {
            return;
        }
        if (this.controlRefClassObserver) {
            this.controlRefClassObserver.destroy();
        }
        if (this.controlRef && !this.htmlFor) {
            const input = await this.controlRef.waitForCurrent();
            this.controlRefClassObserver = createClassMutationObserver(input, () => this.checkForInvalidState(input));
        }
    }
    registerHtmlForClassObserver(forElement) {
        if (this.htmlForClassObserver) {
            this.htmlForClassObserver.destroy();
        }
        this.htmlForClassObserver = createClassMutationObserver(forElement, () => this.checkForInvalidState(forElement));
    }
    checkForInvalidState(elementToCheck) {
        var _a;
        const hasInvalidClass = elementToCheck.classList.contains('is-invalid') ||
            elementToCheck.classList.contains('ix-invalid');
        this.isInvalid = (_a = this.explicitIsInvalid) !== null && _a !== void 0 ? _a : hasInvalidClass;
    }
    async checkForInternalState() {
        if (this.htmlFor) {
            const forElement = document.getElementById(this.htmlFor);
            if (forElement) {
                if (typeof forElement.required === 'boolean') {
                    this.required = forElement.required;
                }
                this.registerHtmlForClassObserver(forElement);
                this.checkForInvalidState(forElement);
                return;
            }
        }
        if (this.controlRef) {
            const input = await this.controlRef.waitForCurrent();
            this.required = input.required;
        }
    }
    async focusOnClick() {
        if (this.htmlFor) {
            const target = document.getElementById(this.htmlFor);
            if (target) {
                let input = null;
                if (isIxInputFieldComponent(target)) {
                    input = await target.getNativeInputElement();
                }
                else {
                    input = target;
                }
                if (typeof input.focus === 'function') {
                    input.focus();
                }
            }
        }
        if (this.controlRef) {
            (await this.controlRef.waitForCurrent()).focus();
        }
    }
    render() {
        return (h(Host, { key: '1a8134a30a1b101f6bfd678f905eb88e927e47d6', onClick: () => this.focusOnClick() }, h("label", Object.assign({ key: 'ef6df6072485fb72ea02a0a9800be3be15e84314', htmlFor: this.htmlFor }, this.a11yAttributes, { ref: this.labelRef }), h("ix-typography", { key: '572975edeae54d15932c3b6091ed63dfff8bb231', textColor: this.isInvalid ? 'alarm' : 'soft', format: "label" }, h("slot", { key: '452847ea7365cefc945bc6166bf2e8fd44d3f530' }), this.required && h("span", { key: '5e85f801d6fe8a01d6595a55bb263d0bcf91a888' }, "*")))));
    }
    get hostElement() { return this; }
    static get watchers() { return {
        "isInvalid": ["isInvalidChanged"],
        "htmlFor": ["registerHtmlForObserver"],
        "controlRef": ["registerControlRefObserver"]
    }; }
    static get style() { return fieldLabelCss; }
}, [257, "ix-field-label", {
        "required": [1540],
        "htmlFor": [513, "html-for"],
        "controlRef": [16, "control-ref"],
        "isInvalid": [1028, "is-invalid"]
    }, undefined, {
        "isInvalid": ["isInvalidChanged"],
        "htmlFor": ["registerHtmlForObserver"],
        "controlRef": ["registerControlRefObserver"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-field-label", "ix-typography"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-field-label":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, FormFieldLabel);
            }
            break;
        case "ix-typography":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}

export { FormFieldLabel as F, defineCustomElement as d };
