// @siemens/ix/dist/collection/components/drawer/drawer.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { iconClose } from "@siemens/ix-icons/icons";
import { h, Host, } from "@stencil/core";
import { animate } from "animejs";
import Animation from "../utils/animation";
/**
 * @deprecated Will be removed with 5.0.0, use ix-pane as successor
 */
export class Drawer {
    constructor() {
        /**
         * Show or hide the drawer
         */
        this.show = false;
        /**
         * Fired in case of an outside click during drawer showed state
         */
        // eslint-disable-next-line @stencil-community/ban-default-true
        this.closeOnClickOutside = true;
        /**
         * Render the drawer with maximum height
         */
        this.fullHeight = false;
        /**
         * Min width interpreted as REM
         */
        this.minWidth = 16;
        /**
         * Max width interpreted as REM
         */
        this.maxWidth = 28;
        /**
         * Width interpreted as REM if not set to 'auto'
         */
        this.width = this.minWidth;
        this.toggle = false;
        this.callback = this.clickedOutside.bind(this);
        this.showContent = true;
    }
    onShowChanged(newValue, oldValue) {
        if (newValue === oldValue) {
            return;
        }
        this.toggleDrawer(newValue);
    }
    /**
     * Toggle or define show state of drawer
     * @param show Overwrite toggle state with boolean
     */
    async toggleDrawer(show) {
        show = show !== null && show !== void 0 ? show : !this.show;
        if (show) {
            const { defaultPrevented } = this.open.emit();
            if (defaultPrevented) {
                return;
            }
            this.show = true;
            if (!this.toggle && this.divElement) {
                this.slideInRight(this.divElement);
                setTimeout(() => {
                    window.addEventListener('mousedown', this.callback);
                }, Animation.mediumTime);
            }
        }
        else {
            const { defaultPrevented } = this.drawerClose.emit();
            if (defaultPrevented) {
                return;
            }
            this.show = false;
            if (this.toggle && this.divElement) {
                this.slideOutRight(this.divElement);
            }
            window.removeEventListener('mousedown', this.callback);
        }
        this.toggle = this.show;
        return Promise.resolve();
    }
    onCloseClicked() {
        this.show = false;
    }
    clickedOutside(evt) {
        if (!this.closeOnClickOutside) {
            return;
        }
        const target = evt.target;
        const closestElement = target.closest('#div-container');
        const btn = target.closest('#drawer-btn');
        if (evt.target.type !== 'button' &&
            closestElement !== this.divElement &&
            target !== btn) {
            this.show = false;
        }
    }
    getConstrainedWidth(width) {
        return Math.min(Math.max(width, this.minWidth), this.maxWidth);
    }
    slideOutRight(el) {
        const initialWidth = `${this.getConstrainedWidth(this.width === 'auto' ? this.minWidth : this.width)}rem`;
        animate(el, {
            duration: Animation.mediumTime,
            width: [initialWidth, 0],
            opacity: [1, 0],
            easing: 'easeInSine',
            onComplete: () => {
                el.classList.add('display-none');
            },
        });
    }
    slideInRight(el) {
        const targetWidth = `${this.getConstrainedWidth(this.width === 'auto' ? this.minWidth : this.width)}rem`;
        animate(el, {
            duration: Animation.mediumTime,
            width: [0, targetWidth],
            opacity: [0, 1],
            easing: 'easeOutSine',
            onBegin: () => {
                el.classList.remove('display-none');
            },
            onComplete: () => {
                this.showContent = true;
            },
        });
    }
    componentDidLoad() {
        this.toggleDrawer(this.show);
    }
    render() {
        return (h(Host, { key: '97ae2273ad9c603d477907a07cba1e1477e9611d', class: {
                'drawer-container': true,
                'display-none': true,
            }, style: {
                width: '0',
                'max-width': `${this.maxWidth}rem`,
                height: this.fullHeight ? '100%' : 'auto',
                overflow: 'hidden',
            }, ref: (el) => (this.divElement = el), "data-testid": "container", id: "div-container" }, h("div", { key: '01ce6e66f3460299967c678cded87b96d6967445', style: {
                width: this.width === 'auto'
                    ? 'auto'
                    : `${this.getConstrainedWidth(this.width)}rem`,
            } }, h("div", { key: 'f4c040460d4917c8289b3746158696ed20c92d20', class: "header" }, h("div", { key: '0a05c91318f6c67c8a802b35970f321dbba55dbe', class: "header-content" }, h("slot", { key: '853415c470b8c34b70ef8e95be2fc44629f6eb62', name: "header" })), h("ix-icon-button", { key: '47cb76743364e504a287c218b62d8ff82f92276b', class: "close-button", style: {
                display: this.showContent ? 'block' : 'none',
            }, icon: iconClose, iconColor: "color-soft-text", size: "24", variant: "tertiary", onClick: () => this.onCloseClicked(), "data-testid": "close-button", "aria-label": this.ariaLabelCloseButton })), h("div", { key: '427f1fea6a16c11291998c40ecfe80599b0dcf75', class: "content", style: {
                display: this.showContent ? 'block' : 'none',
            } }, h("slot", { key: 'b5471ae2a3a59ede24fcf6deeb323bd2ab6acf4a' })))));
    }
    static get is() { return "ix-drawer"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["drawer.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["drawer.css"]
        };
    }
    static get properties() {
        return {
            "show": {
                "type": "boolean",
                "attribute": "show",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show or hide the drawer"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "closeOnClickOutside": {
                "type": "boolean",
                "attribute": "close-on-click-outside",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Fired in case of an outside click during drawer showed state"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "true"
            },
            "fullHeight": {
                "type": "boolean",
                "attribute": "full-height",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Render the drawer with maximum height"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "minWidth": {
                "type": "number",
                "attribute": "min-width",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Min width interpreted as REM"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "16"
            },
            "maxWidth": {
                "type": "number",
                "attribute": "max-width",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Max width interpreted as REM"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "28"
            },
            "width": {
                "type": "any",
                "attribute": "width",
                "mutable": false,
                "complexType": {
                    "original": "number | 'auto'",
                    "resolved": "\"auto\" | number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Width interpreted as REM if not set to 'auto'"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "this.minWidth"
            },
            "ariaLabelCloseButton": {
                "type": "string",
                "attribute": "aria-label-close-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label for the close icon button\nWill be set as aria-label on the nested HTML button element"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            }
        };
    }
    static get states() {
        return {
            "showContent": {}
        };
    }
    static get events() {
        return [{
                "method": "open",
                "name": "open",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Fire event after drawer is open"
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }, {
                "method": "drawerClose",
                "name": "drawerClose",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Fire event after drawer is close"
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }];
    }
    static get methods() {
        return {
            "toggleDrawer": {
                "complexType": {
                    "signature": "(show?: boolean) => Promise<void>",
                    "parameters": [{
                            "name": "show",
                            "type": "boolean | undefined",
                            "docs": "Overwrite toggle state with boolean"
                        }],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Toggle or define show state of drawer",
                    "tags": [{
                            "name": "param",
                            "text": "show Overwrite toggle state with boolean"
                        }]
                }
            }
        };
    }
    static get watchers() {
        return [{
                "propName": "show",
                "methodName": "onShowChanged"
            }];
    }
}
