// @siemens/ix@2.2.3/components/card-title.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';

const cardTitleCss = ":host{display:flex;position:relative;flex-direction:row;align-items:center;margin-top:0.25rem;margin-bottom:0.25rem;gap:1rem;width:100%}:host .title-actions{display:flex;position:relative;flex-direction:row;gap:0.25rem;margin-left:auto;margin-right:0}";
const IxCardTitleStyle0 = cardTitleCss;

const CardHeader = /*@__PURE__*/ proxyCustomElement(class CardHeader extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
    }
    render() {
        return (h(Host, { key: '0f4e8785e6861fa9151cece5946a7b33e85bd7fd' }, h("slot", { key: 'd8909ebe344940cda53d4076ccd1291eb1240f8b' }), h("div", { key: 'f87b52b2b467664d6e9d886d358b1c6b5403aae8', class: "title-actions" }, h("slot", { key: '9e45b897368e305f0ce1e66b47424360c44c148d', name: "title-actions" }))));
    }
    static get style() { return IxCardTitleStyle0; }
}, [1, "ix-card-title"]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-card-title"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-card-title":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, CardHeader);
            }
            break;
    } });
}

export { CardHeader as C, defineCustomElement as d };
