// @siemens/ix@2.2.3/components/field-wrapper.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { m as makeRef } from './make-ref.js';
import { r as renderHelperText, h as hasAnyText } from './helper-text-util.js';
import { d as defineCustomElement$3 } from './field-label.js';
import { d as defineCustomElement$2 } from './tooltip.js';
import { d as defineCustomElement$1 } from './typography.js';

const fieldWrapperCss = ":host{display:flex;position:relative;flex-direction:column}:host .slot-wrapper{display:flex;position:relative;align-items:center;justify-content:flex-start;gap:0.25rem}:host .field-bottom,:host .field-top{display:flex;flex-direction:row;position:relative;justify-content:space-between;gap:1rem;width:-moz-min-content;width:min-content;min-width:100%}:host .field-bottom .bottom-right{margin-left:auto;margin-right:0px}:host .bottom-text{display:flex;position:relative;align-items:flex-start;justify-content:flex-start;gap:0.25rem;margin-right:0.25rem}:host .text-icon{margin:0.125rem}:host .text-icon.invalid{color:var(--theme-helper-icon--color--invalid)}:host .text-icon.info{color:var(--theme-helper-icon--color--info)}:host .text-icon.warning{color:var(--theme-helper-icon--color--warning)}:host .text-icon.valid{color:var(--theme-helper-icon--color--valid)}:host .bottom-text{margin-top:0.25rem;margin-bottom:0.25rem}";
const IxFieldWrapperStyle0 = fieldWrapperCss;

const FieldWrapper = /*@__PURE__*/ proxyCustomElement(class FieldWrapper extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.slotRef = makeRef();
        this.helperText = undefined;
        this.label = undefined;
        this.invalidText = undefined;
        this.validText = undefined;
        this.infoText = undefined;
        this.warningText = undefined;
        this.isInvalid = false;
        this.isValid = false;
        this.isInfo = false;
        this.isWarning = false;
        this.showTextAsTooltip = false;
        this.required = false;
        this.htmlForLabel = undefined;
        this.controlRef = undefined;
    }
    render() {
        const textOptions = {
            invalidText: this.invalidText,
            isInvalid: this.isInvalid,
            isValid: this.isValid,
            validText: this.validText,
            isWarning: this.isWarning,
            warningText: this.warningText,
            isInfo: this.isInfo,
            infoText: this.infoText,
            helperText: this.helperText,
        };
        return (h(Host, { key: 'e572687360d706dbe987ecb2552b573fe4c4d5d4' }, this.label && (h("div", { key: 'c6373751bc6d7d90d7927661332a7ebe07ebe32c', class: "field-top" }, h("ix-field-label", { key: '5c0f33a101ca7fe85307dfb1bcca00583f7af505', required: this.required, htmlFor: this.htmlForLabel, controlRef: this.controlRef, isInvalid: this.isInvalid }, this.label))), h("div", { key: 'a0537918a05ae6cdb31fec15593310fe567f7f29', class: {
                'slot-wrapper': true,
            }, ref: this.slotRef }, h("slot", { key: '9ac61601a8bb8486fb195007affb95efd38a16cc' })), h("div", { key: 'a2ce3a990499d4f44e908504e72e8e38ec6f9bf0', class: 'field-bottom' }, !this.showTextAsTooltip && renderHelperText(textOptions), h("div", { key: '83e807d409fdc1a85f0b4b5533d3f8925ff35fec', class: "bottom-right" }, h("slot", { key: '87702e1c61e5c242a6085bcc4db7f84ee325dacd', name: "bottom-right" }))), this.showTextAsTooltip === true && hasAnyText(textOptions) && (h("ix-tooltip", { key: '66a20de650c5ea67e804f787bd729ac02078ed42', for: this.slotRef.waitForCurrent(), showDelay: 500, placement: "bottom" }, renderHelperText(textOptions)))));
    }
    get hostElement() { return this; }
    static get style() { return IxFieldWrapperStyle0; }
}, [1, "ix-field-wrapper", {
        "helperText": [1, "helper-text"],
        "label": [1],
        "invalidText": [1, "invalid-text"],
        "validText": [1, "valid-text"],
        "infoText": [1, "info-text"],
        "warningText": [1, "warning-text"],
        "isInvalid": [4, "is-invalid"],
        "isValid": [4, "is-valid"],
        "isInfo": [4, "is-info"],
        "isWarning": [4, "is-warning"],
        "showTextAsTooltip": [4, "show-text-as-tooltip"],
        "required": [4],
        "htmlForLabel": [1, "html-for-label"],
        "controlRef": [16]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-field-wrapper", "ix-field-label", "ix-tooltip", "ix-typography"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-field-wrapper":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, FieldWrapper);
            }
            break;
        case "ix-field-label":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "ix-tooltip":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
        case "ix-typography":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}

export { FieldWrapper as F, defineCustomElement as d };
