// @siemens/ix@2.2.3/components/ix-css-grid-item.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';

const cssGridItemCss = ":host{display:block;position:relative}:host ::slotted(*){height:100%;width:100%}";
const IxCssGridItemStyle0 = cssGridItemCss;

const CssGridItem = /*@__PURE__*/ proxyCustomElement(class CssGridItem extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.itemName = undefined;
    }
    render() {
        const style = {};
        style['grid-area'] = this.itemName;
        return (h(Host, { key: 'bc5b71e095a35738800e7b95baca2c4a7874205e', style: style }, h("slot", { key: 'dd2a9bc3a5eecc1d539820a72dd9de51e5466923' })));
    }
    static get style() { return IxCssGridItemStyle0; }
}, [1, "ix-css-grid-item", {
        "itemName": [1, "item-name"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-css-grid-item"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-css-grid-item":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, CssGridItem);
            }
            break;
    } });
}

const IxCssGridItem = CssGridItem;
const defineCustomElement = defineCustomElement$1;

export { IxCssGridItem, defineCustomElement };
