// @siemens/ix@2.7.0/dist/collection/components/group-item/group-item.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host, } from "@stencil/core";
export class GroupItem {
    constructor() {
        this.icon = undefined;
        this.text = undefined;
        this.secondaryText = undefined;
        this.suppressSelection = false;
        this.selected = false;
        this.focusable = true;
        this.index = undefined;
    }
    clickListen() {
        this.selectedChanged.emit(this.hostElement);
    }
    render() {
        return (h(Host, { key: 'bdade798c7b506d739029a27c9b73a90f101bc0e', class: {
                selected: this.selected && !this.suppressSelection,
            } }, h("button", { key: 'aa88b5a436cc197341a9afdf7a5cd0e522fd325a', tabindex: this.focusable ? 0 : -1 }, h("div", { key: 'c5e7c0aeadc32db6a8c6b0deb262620aaaff1fa1', class: "group-entry-selection-indicator" }), this.icon ? h("ix-icon", { size: "16", name: this.icon }) : null, this.text ? (h("span", { class: "group-entry-text" }, h("span", { title: this.text }, this.text))) : null, this.secondaryText ? (h("span", { class: "group-entry-text-secondary" }, h("span", { title: this.secondaryText }, this.secondaryText))) : null, h("slot", { key: 'a6fa0189ef24b3ce6c593ee99f2df83263b97053' }))));
    }
    static get is() { return "ix-group-item"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["group-item.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["group-item.css"]
        };
    }
    static get properties() {
        return {
            "icon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Group item icon"
                },
                "attribute": "icon",
                "reflect": false
            },
            "text": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Group item text"
                },
                "attribute": "text",
                "reflect": false
            },
            "secondaryText": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Group item secondary text"
                },
                "attribute": "secondary-text",
                "reflect": false
            },
            "suppressSelection": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Supress the selection of the group"
                },
                "attribute": "suppress-selection",
                "reflect": false,
                "defaultValue": "false"
            },
            "selected": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show selected state"
                },
                "attribute": "selected",
                "reflect": false,
                "defaultValue": "false"
            },
            "focusable": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The elements tabindex attribute will get set accordingly.\nIf true tabindex will be 0, -1 otherwise."
                },
                "attribute": "focusable",
                "reflect": false,
                "defaultValue": "true"
            },
            "index": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Index"
                },
                "attribute": "index",
                "reflect": false
            }
        };
    }
    static get events() {
        return [{
                "method": "selectedChanged",
                "name": "selectedChanged",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Selection changed"
                },
                "complexType": {
                    "original": "HTMLIxGroupItemElement",
                    "resolved": "HTMLIxGroupItemElement",
                    "references": {
                        "HTMLIxGroupItemElement": {
                            "location": "global",
                            "id": "global::HTMLIxGroupItemElement"
                        }
                    }
                }
            }];
    }
    static get elementRef() { return "hostElement"; }
    static get listeners() {
        return [{
                "name": "click",
                "method": "clickListen",
                "target": undefined,
                "capture": false,
                "passive": true
            }];
    }
}
