// @siemens/ix/dist/collection/components/workflow-step/workflow-step.js
/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { iconCircle, iconCircleDot, iconCircleFilled, iconError, iconSuccess, iconTriangleFilled, iconWarning, } from "@siemens/ix-icons/icons";
import { Fragment, h, Host, } from "@stencil/core";
export class WorkflowStep {
    constructor() {
        /**
         * Select orientation
         */
        this.vertical = false;
        /**
         * Set disabled
         */
        this.disabled = false;
        /**
         * Set status
         */
        this.status = 'open';
        /**
         * Activate navigation click
         */
        this.clickable = false;
        /**
         * Set selected
         */
        this.selected = false;
        /**
         * Activate navigation click
         *
         * @internal
         */
        this.position = 'undefined';
        this.iconColor = 'workflow-step-icon-default--color';
        this.customIconSlot = false;
    }
    selectedHandler() {
        this.setWorkflowStepStyles();
    }
    watchPropHandler() {
        this.setWorkflowStepStyles();
    }
    setWorkflowStepStyles() {
        const selectedStyle = this.selected ? '--selected' : '';
        switch (this.status) {
            case 'open':
                this.iconName = this.selected ? iconCircleDot : iconCircle;
                this.iconColor = `workflow-step-icon-default--color${selectedStyle}`;
                break;
            case 'success':
                this.iconName = iconSuccess;
                this.iconColor = 'color-success';
                break;
            case 'done':
                this.iconName = iconCircleFilled;
                this.iconColor = `workflow-step-icon-done--color${selectedStyle}`;
                break;
            case 'warning':
                this.iconName = iconWarning;
                //TODO(IX-3400): Replace icon colors with proper CSS variables when available
                this.iconColor = 'color-warning-text';
                break;
            case 'error':
                this.iconName = iconError;
                this.iconColor = 'color-alarm';
                break;
            default:
                this.iconName = iconCircle;
                break;
        }
        if (this.disabled) {
            this.iconColor = 'workflow-step-icon-success--color--disabled';
        }
    }
    componentWillLoad() {
        this.watchPropHandler();
        this.selectedHandler();
        this.customIconSlot = !!this.hostElement.querySelector('[slot="custom-icon"]');
    }
    onStepClick() {
        if (!this.disabled && this.clickable) {
            this.selectedChanged.emit(this.hostElement);
        }
    }
    getIconAriaLabel() {
        switch (this.iconName) {
            case iconCircle:
                return 'Circle';
            case iconCircleDot:
                return 'Circle dot';
            case iconCircleFilled:
                return 'Done';
            case iconError:
                return 'Error';
            case iconSuccess:
                return 'Success';
            case iconTriangleFilled:
                return 'Warning';
            case iconWarning:
                return 'Warning';
            default:
                return 'Step';
        }
    }
    render() {
        const icons = !this.customIconSlot ? (h(Fragment, null, h("ix-icon", { color: "color-1", name: this.status === 'warning' ? iconTriangleFilled : iconCircleFilled, class: "absolute", size: "24" }), h("ix-icon", { color: this.iconColor, name: this.iconName, class: "absolute", size: "24", "aria-label": this.getIconAriaLabel() }))) : null;
        return (h(Host, { key: 'bcb3fee2267536717286c691e1b3fcebfa52ba25', class: { 'host-vertical': this.vertical }, onClick: () => this.onStepClick() }, h("div", { key: '54b04c63b7df962caca7cd7c53f8e1e19f1dbdf8', tabIndex: 0, class: {
                step: true,
                selected: this.selected,
                vertical: this.vertical,
                disabled: this.disabled,
                clickable: this.clickable && !this.disabled,
            } }, h("div", { key: '64b8292366bc54e3ddaf92077fe7001d4e263f76', class: "wrapper" }, h("div", { key: 'ce2379628165a5c06885a1eb2c75152da54ffa2f', class: {
                line: true,
                selected: this.selected,
                [this.status]: true,
                [this.position]: true,
            } }), h("div", { key: '11651f719860bc2dfd5e82253cbfc94cd45c0a8f', class: "iconWrapper" }, icons, h("slot", { key: '4a6a1596874064e8efba07583362640950ba46e3', name: "custom-icon" }))), h("div", { key: '1fdbe00dbb8159c8f50d4cbdef4efd86c3bda736', class: "text" }, h("slot", { key: '2ab567c5fe3ff5a917f5f3ea6049bae66a1b2adc' })))));
    }
    static get is() { return "ix-workflow-step"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["workflow-step.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["workflow-step.css"]
        };
    }
    static get properties() {
        return {
            "vertical": {
                "type": "boolean",
                "attribute": "vertical",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Select orientation"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "disabled": {
                "type": "boolean",
                "attribute": "disabled",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Set disabled"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "status": {
                "type": "string",
                "attribute": "status",
                "mutable": false,
                "complexType": {
                    "original": "'open' | 'success' | 'done' | 'warning' | 'error'",
                    "resolved": "\"done\" | \"error\" | \"open\" | \"success\" | \"warning\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Set status"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'open'"
            },
            "clickable": {
                "type": "boolean",
                "attribute": "clickable",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Activate navigation click"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "selected": {
                "type": "boolean",
                "attribute": "selected",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Set selected"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "position": {
                "type": "string",
                "attribute": "position",
                "mutable": false,
                "complexType": {
                    "original": "'first' | 'last' | 'single' | 'undefined'",
                    "resolved": "\"first\" | \"last\" | \"single\" | \"undefined\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": "Activate navigation click"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'undefined'"
            }
        };
    }
    static get states() {
        return {
            "iconName": {},
            "iconColor": {}
        };
    }
    static get events() {
        return [{
                "method": "selectedChanged",
                "name": "selectedChanged",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": ""
                },
                "complexType": {
                    "original": "HTMLIxWorkflowStepElement",
                    "resolved": "HTMLIxWorkflowStepElement",
                    "references": {
                        "HTMLIxWorkflowStepElement": {
                            "location": "global",
                            "id": "global::HTMLIxWorkflowStepElement"
                        }
                    }
                }
            }];
    }
    static get elementRef() { return "hostElement"; }
    static get watchers() {
        return [{
                "propName": "selected",
                "methodName": "selectedHandler"
            }, {
                "propName": "disabled",
                "methodName": "watchPropHandler"
            }, {
                "propName": "status",
                "methodName": "watchPropHandler"
            }];
    }
}
