// @siemens/ix@4.3.0/dist/collection/components/tile/tile.js
/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host } from "@stencil/core";
export class Tile {
    constructor() {
        /**
         * Size of the tile - one of 'small', 'medium' or 'large'
         */
        this.size = 'medium';
        this.hasHeaderSlot = false;
        this.hasFooterSlot = false;
    }
    handleHeaderSlotChange() {
        this.hasHeaderSlot = !!this.hostElement.querySelector('[slot="header"]');
    }
    handleFooterSlotChange() {
        this.hasFooterSlot = !!this.hostElement.querySelector('[slot="footer"]');
    }
    render() {
        return (h(Host, { key: 'e1166a8b4e348942b7ed4258f88f8be64de05933', class: {
                'tile-small': this.size === 'small',
                'tile-medium': this.size === 'medium',
                'tile-big': this.size === 'big',
            } }, h("div", { key: 'a6bd9158342ed546d51c55685afe3d75c2bf31b8', class: {
                'tile-header': true,
                'has-content': this.hasHeaderSlot,
            } }, h("slot", { key: 'c24c054f14e378dc126472b4086b4be4a245cbac', name: "header", onSlotchange: () => this.handleHeaderSlotChange() })), h("div", { key: '975744cdd6f42661636fc382a2f2e3c256ccdc9f', class: "tile-subheader" }, h("slot", { key: 'a867a5e2e3cb8958c819d5d936bf955cb51e808b', name: "subheader" })), h("div", { key: '8ab40a481b778384e6773a6318545e1d0001e669', class: "tile-content" }, h("slot", { key: 'c013dd68db8ce80c3b4cfac26acf9c690fcb2439' })), h("div", { key: '00b012d9f4bce4b623fc30b1f750f4319f4144f9', class: {
                'tile-footer': true,
                'has-content': this.hasFooterSlot,
            } }, h("slot", { key: 'fc565c7344dadeb70cf0002392a0ee98774843df', name: "footer", onSlotchange: () => this.handleFooterSlotChange() }))));
    }
    static get is() { return "ix-tile"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tile.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tile.css"]
        };
    }
    static get properties() {
        return {
            "size": {
                "type": "string",
                "attribute": "size",
                "mutable": false,
                "complexType": {
                    "original": "'small' | 'medium' | 'big'",
                    "resolved": "\"big\" | \"medium\" | \"small\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Size of the tile - one of 'small', 'medium' or 'large'"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'medium'"
            }
        };
    }
    static get states() {
        return {
            "hasHeaderSlot": {},
            "hasFooterSlot": {}
        };
    }
    static get elementRef() { return "hostElement"; }
}
