// @siemens/ix@4.3.0/dist/collection/components/toggle-button/toggle-button.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host } from "@stencil/core";
import { BaseButton } from "../button/base-button";
import { a11yBoolean } from "../utils/a11y";
export class ToggleButton {
    constructor() {
        /**
         * Button variant.
         */
        this.variant = 'subtle-primary';
        /**
         * Disable the button
         */
        this.disabled = false;
        /**
         * Loading button
         */
        this.loading = false;
        /**
         * Show button as pressed
         */
        this.pressed = false;
    }
    dispatchPressedChange() {
        this.pressedChange.emit(!this.pressed);
    }
    render() {
        const baseButtonProps = {
            variant: this.variant,
            iconOnly: false,
            iconOval: false,
            selected: this.pressed,
            disabled: this.disabled || this.loading,
            icon: this.icon,
            iconRight: this.iconRight,
            loading: this.loading,
            onClick: () => this.dispatchPressedChange(),
            type: 'button',
            ariaAttributes: {
                'aria-pressed': a11yBoolean(this.pressed),
                'aria-label': this.ariaLabelButton,
            },
        };
        return (h(Host, { key: '3d0e41a6571193d4b7d8e73701b6d295f4ab5482', class: {
                disabled: this.disabled || this.loading,
            } }, h(BaseButton, Object.assign({ key: '770d156dbc007d146efbd7b5604b3c902197b4c2' }, baseButtonProps), h("slot", { key: '6d2a11909dcf61da1571d2e921e850d9a54ed95c' }))));
    }
    static get is() { return "ix-toggle-button"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["./toggle-button.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["toggle-button.css"]
        };
    }
    static get properties() {
        return {
            "variant": {
                "type": "string",
                "attribute": "variant",
                "mutable": false,
                "complexType": {
                    "original": "ToggleButtonVariant",
                    "resolved": "\"primary\" | \"secondary\" | \"subtle-primary\" | \"subtle-secondary\" | \"subtle-tertiary\" | \"tertiary\"",
                    "references": {
                        "ToggleButtonVariant": {
                            "location": "local",
                            "path": "/home/runner/work/ix/ix/packages/core/src/components/toggle-button/toggle-button.tsx",
                            "id": "src/components/toggle-button/toggle-button.tsx::ToggleButtonVariant"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Button variant."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'subtle-primary'"
            },
            "disabled": {
                "type": "boolean",
                "attribute": "disabled",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Disable the button"
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "defaultValue": "false"
            },
            "loading": {
                "type": "boolean",
                "attribute": "loading",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Loading button"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "icon": {
                "type": "string",
                "attribute": "icon",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Icon name"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "iconRight": {
                "type": "string",
                "attribute": "icon-right",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "Icon name for the right side of the button"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "pressed": {
                "type": "boolean",
                "attribute": "pressed",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show button as pressed"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "ariaLabelButton": {
                "type": "string",
                "attribute": "aria-label-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label that will be set on the native HTML button element"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            }
        };
    }
    static get events() {
        return [{
                "method": "pressedChange",
                "name": "pressedChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Pressed change event"
                },
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                }
            }];
    }
}
