// @siemens/ix@4.3.0/dist/collection/components/event-list-item/event-list-item.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { iconChevronRight } from "@siemens/ix-icons/icons";
import { h, Host, } from "@stencil/core";
import { a11yBoolean } from "../utils/a11y";
export class EventListItem {
    constructor() {
        /**
         * Variant of the event list item
         *
         * @since 4.0.0
         */
        this.variant = 'outline';
        /**
         * Show event list item as selected
         */
        this.selected = false;
        /**
         * Disable event list item
         */
        this.disabled = false;
        /**
         * Show chevron on right side of the event list item
         */
        this.chevron = false;
    }
    handleItemClick() {
        this.itemClick.emit();
    }
    render() {
        var _a;
        let color = ((_a = this.itemColor) === null || _a === void 0 ? void 0 : _a.startsWith('--theme'))
            ? `var(${this.itemColor})`
            : `var(--theme-${this.itemColor})`;
        return (h(Host, { key: '3fa7304c0d114cd65b572e6d7e1a7362d3f63950', class: {
                disabled: this.disabled,
                [`event-list-item-${this.variant}`]: true,
            } }, h("div", { key: '9dc5734ea43068e74efc45d89a6fda3bad040cd1', role: "listitem", "aria-disabled": a11yBoolean(this.disabled), class: {
                'event-list-item': true,
                selected: this.selected,
                disabled: this.disabled,
            } }, h("div", { key: '5fc29826643c430d3f0c82826142b461395d5075', class: `indicator ${!this.itemColor ? 'indicator-empty' : ''}`, style: {
                'background-color': this.itemColor ? color : 'inherit',
                opacity: `${this.disabled ? 0.4 : 1}`,
            } }), h("div", { key: 'c53b6ea99b0744dc6a0d57024853d0627a3963e2', class: "event-list-item-container" }, h("div", { key: '22a50db11d9997caa9df67fea550e1f58224344c', class: "event-content" }, h("slot", { key: '6361d1b0c326919f46ed7eb1b6bcb4b17421258e' })), this.chevron && (h("ix-icon", { key: 'edd5d1e1ef0fa1c85b319f9bdccdf1e2cbe8183a', name: iconChevronRight, size: "16", class: "chevron-icon", "aria-hidden": "true" }))))));
    }
    static get is() { return "ix-event-list-item"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["event-list-item.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["event-list-item.css"]
        };
    }
    static get properties() {
        return {
            "variant": {
                "type": "string",
                "attribute": "variant",
                "mutable": false,
                "complexType": {
                    "original": "'outline' | 'filled'",
                    "resolved": "\"filled\" | \"outline\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "Variant of the event list item"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'outline'"
            },
            "itemColor": {
                "type": "string",
                "attribute": "item-color",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Color of the status indicator.\nYou can find a list of all available colors in our documentation.\nExample values are `--theme-color-alarm` or `color-alarm`\n\n{@link https://ix.siemens.io/docs/styles/colors}"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "selected": {
                "type": "boolean",
                "attribute": "selected",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show event list item as selected"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "disabled": {
                "type": "boolean",
                "attribute": "disabled",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Disable event list item"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "chevron": {
                "type": "boolean",
                "attribute": "chevron",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show chevron on right side of the event list item"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get events() {
        return [{
                "method": "itemClick",
                "name": "itemClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event list item click"
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }];
    }
    static get listeners() {
        return [{
                "name": "click",
                "method": "handleItemClick",
                "target": undefined,
                "capture": false,
                "passive": true
            }];
    }
}
