// @siemens/ix/dist/collection/components/menu-about-news/menu-about-news.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { iconClose, iconShout } from "@siemens/ix-icons/icons";
import { h, Host } from "@stencil/core";
export class MenuAboutNews {
    constructor() {
        /**
         * Show about news
         */
        this.show = false;
        /**
         * i18n label for 'Show more' button
         */
        this.i18nShowMore = 'Show more';
        /** @internal */
        this.expanded = false;
    }
    render() {
        return (h(Host, { key: '7443d209d383de9e46d8c024c1756ae814c5d764', class: {
                expanded: this.expanded,
                show: !this.show,
            } }, h("div", { key: '034c93b01133fa50d28c0e4ce98becc53fe29158', class: "banner-container" }, h("ix-icon", { key: '69df7443eba7107db4d4f20d2a628159a7ed8660', color: "color-inv-contrast-text", name: iconShout, size: "32" }), h("svg", { key: '36cc1e785395307d29813cac4f13b1309ee2d257', viewBox: "0 0 48 56", xmlns: "http://www.w3.org/2000/svg" }, h("polygon", { key: '35d4a86b62c225dc03329e6e53bdd6180d1a6b8e', points: "0 0 48 0 48 56 24 48 0 56" }))), h("div", { key: 'd5f1d76a2bf661fb865fa7c4857b0f20d6f225dc', class: "cui-popover-news-header" }, h("ix-typography", { key: 'd8ef22907758393fe430cc2a4940dea8ee7372e8', format: "label", bold: true }, this.label)), h("ix-icon-button", { key: '2fb7cf52ee3bf3e475fc5fa6d0aaadec96246acc', size: "24", icon: iconClose, iconColor: "color-soft-text", variant: "tertiary", onClick: () => {
                this.show = false;
                this.closePopover.emit();
            } }), h("div", { key: '8875b7027424233f78ac46c52cbeccc3d4fb7560', class: "slot-container" }, h("slot", { key: 'b662f3b8c11a9c3ea639ed6b753a4b80ed038bec' })), this.aboutItemLabel ? (h("div", { class: "cui-popover-news-footer" }, h("ix-button", { variant: "primary", onClick: (e) => {
                this.show = false;
                this.showMore.emit(e);
            } }, this.i18nShowMore))) : null, h("div", { key: '97cbdaf247a6f83b9361121cbac596dc86e2bb28', id: "arrow" })));
    }
    static get is() { return "ix-menu-about-news"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["menu-about-news.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["menu-about-news.css"]
        };
    }
    static get properties() {
        return {
            "show": {
                "type": "boolean",
                "attribute": "show",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show about news"
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "defaultValue": "false"
            },
            "label": {
                "type": "string",
                "attribute": "label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Title of the about news"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "i18nShowMore": {
                "type": "string",
                "attribute": "i18n-show-more",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "i18n label for 'Show more' button"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'Show more'"
            },
            "aboutItemLabel": {
                "type": "string",
                "attribute": "about-item-label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Subtitle of the about news"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "expanded": {
                "type": "boolean",
                "attribute": "expanded",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": ""
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get events() {
        return [{
                "method": "showMore",
                "name": "showMore",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Show More button is pressed"
                },
                "complexType": {
                    "original": "MouseEvent",
                    "resolved": "MouseEvent",
                    "references": {
                        "MouseEvent": {
                            "location": "global",
                            "id": "global::MouseEvent"
                        }
                    }
                }
            }, {
                "method": "closePopover",
                "name": "closePopover",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Popover closed"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }];
    }
}
