// @stencil/react-output-target/dist/index.js
import { Project as R, VariableDeclarationKind as O } from "ts-morph";
import j from "node:path";
const P = (e) => e.toLowerCase().split("-").map((r) => r.charAt(0).toUpperCase() + r.slice(1)).join(""), M = (e) => e.replace(/-([_a-z])/g, (r, s) => s.toUpperCase()), L = (e) => e.replace(/\/([a-z])/g, (r, s) => s.toUpperCase()), k = (e) => {
  const r = L(e);
  return M(`on-${r}`);
}, J = (e) => e.replace(/\/\/.*$/gm, "").replace(/\n/g, " ").replace(/\s{2,}/g, " ").trim(), q = async ({
  components: e,
  project: r,
  outDir: s
}) => {
  const l = r || new R({ useInMemoryFileSystem: !0 }), u = `/* eslint-disable */
`, d = `/**
 * This file was automatically generated by the Stencil React Output Target.
 * Changes to this file may cause incorrect behavior and will be lost if the code is regenerated.
 */

`, f = j.join(s || "", "components.ts"), g = l.createSourceFile(f, d + u, {
    overwrite: !0
  });
  for (const y of e) {
    const C = y.tagName, h = P(C), i = y.tagName;
    g.addExportDeclaration({
      moduleSpecifier: `./${i}.js`,
      namedExports: [h]
    });
  }
  return g.organizeImports(), g.formatText(), await g.save(), g;
}, x = ({
  components: e,
  stencilPackageName: r,
  customElementsDir: s,
  hydrateModule: l,
  clientModule: u,
  serializeShadowRoot: d,
  transformTag: f
}) => {
  const g = new R({ useInMemoryFileSystem: !0 }), y = `'use client';

`, C = `/**
 * This file was automatically generated by the Stencil React Output Target.
 * Changes to this file may cause incorrect behavior and will be lost if the code is regenerated.
 */

`, h = `/* eslint-disable */
`, c = l ? [
    f ? `import { getTagTransformer } from './tag-transformer.js';
` : "",
    "import { createComponent, type SerializeShadowRootOptions, type HydrateModule, type ReactWebComponent } from '@stencil/react-output-target/ssr';"
  ].filter(Boolean).join(`
`) : "import { createComponent } from '@stencil/react-output-target/runtime';", p = f ? `import { transformTag } from './tag-transformer.js';
` : "", n = g.createSourceFile(
    "component.ts",
    `${y}${C}${h}
import React from 'react';
${c}
import type { EventName, StencilReactComponent } from '@stencil/react-output-target/runtime';
${p}
  `
  );
  l && u && n.addImportDeclaration({
    moduleSpecifier: u,
    namespaceImport: "clientComponents"
  }), l && n.addVariableStatement({
    isExported: !0,
    declarationKind: O.Const,
    declarations: [
      {
        name: "serializeShadowRoot",
        type: "SerializeShadowRootOptions",
        initializer: d ? JSON.stringify(d) : '{ default: "declarative-shadow-dom" }'
      }
    ]
  });
  for (const T of e) {
    const a = T.tagName, t = P(a), m = `${t}Element`, b = `${t}CustomEvent`;
    n.addImportDeclaration({
      moduleSpecifier: `${r}/${s}/${a}.js`,
      namedImports: [
        {
          name: t,
          alias: m
        },
        {
          name: "defineCustomElement",
          alias: `define${t}`
        }
      ]
    });
    const w = (T.events || []).filter((o) => o.internal === !1), E = [], v = /* @__PURE__ */ new Set();
    let F = !1;
    for (const o of w) {
      if (Object.keys(o.complexType.references).length > 0)
        for (const N of Object.keys(o.complexType.references))
          !(o.complexType.references[N].location === "global") && !v.has(N) && (v.add(N), n.addImportDeclaration({
            moduleSpecifier: r,
            namedImports: [
              {
                name: N,
                isTypeOnly: !0
              }
            ]
          }));
      F || (F = !0, n.addImportDeclaration({
        moduleSpecifier: r,
        namedImports: [
          {
            name: b,
            isTypeOnly: !0
          }
        ]
      })), E.push({
        originalName: o.name,
        name: k(o.name),
        type: `EventName<${b}<${J(o.complexType.original)}>>`
      });
    }
    const S = `${t}Events`;
    n.addTypeAlias({
      isExported: !0,
      name: S,
      type: E.length > 0 ? `{ ${E.map((o) => `${o.name}: ${o.type}`).join(`,
`)} }` : "NonNullable<unknown>"
    });
    const U = f ? `,
    transformTag` : "", D = `/*@__PURE__*/ createComponent<${m}, ${S}>({
    tagName: '${a}',
    elementClass: ${m},
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {${E.map((o) => `${o.name}: '${o.originalName}'`).join(`,
`)}} as ${S},
    defineCustomElement: define${t}${U}
  })`, z = f ? `,
    getTagTransformer` : "", A = `/*@__PURE__*/ createComponent<${m}, ${S}>({
    tagName: '${a}',
    properties: {${T.properties.filter((o) => !!o.attribute).map((o) => `${o.name}: '${o.attribute}'`).join(`,
`)}},
    hydrateModule: import('${l}') as Promise<HydrateModule>,
    clientModule: clientComponents.${t} as ReactWebComponent<${m}, ${S}>,
    serializeShadowRoot${z}
  })`;
    n.addVariableStatement({
      isExported: !0,
      declarationKind: O.Const,
      // React as never is a hack to by-pass a @types/react issue.
      declarations: [
        {
          name: t,
          type: `StencilReactComponent<${m}, ${S}>`,
          initializer: l ? A : D
        }
      ]
    });
  }
  if (n.organizeImports(), l && u) {
    const T = n.getImportDeclarations().find((a) => {
      var t;
      return ((t = a.getNamespaceImport()) == null ? void 0 : t.getText()) === "clientComponents";
    });
    T && n.insertText(
      T.getStart(),
      `// @ts-ignore - ignore potential type issues as the project is importing itself
`
    );
  }
  return n.formatText(), n.getFullText();
}, G = ({
  stencilPackageName: e,
  customElementsDir: r
}) => `/* eslint-disable */
/* tslint:disable */
import { setTagTransformer as clientSetTagTransformer } from '${e}/${r}/index.js';

let tagTransformer: ((tagName: string) => string) | undefined;

export const setTagTransformer = (transformer: (tagName: string) => string) => {
  clientSetTagTransformer(transformer);
  tagTransformer = transformer;
};

export const transformTag = (tag: string): string => {
  return tagTransformer ? tagTransformer(tag) : tag;
};

export const getTagTransformer = () => tagTransformer;
`, K = async ({
  stencilPackageName: e,
  components: r,
  outDir: s,
  esModules: l,
  customElementsDir: u,
  excludeComponents: d,
  project: f,
  hydrateModule: g,
  clientModule: y,
  excludeServerSideRenderingFor: C,
  serializeShadowRoot: h,
  transformTag: i
}) => {
  const c = [], p = r.filter((a) => !(a.internal === !0 || d != null && d.includes(a.tagName)));
  if (p.length === 0)
    return [];
  const n = {};
  function T(a, t = "components") {
    const m = j.join(s, `${t}.ts`), b = x({
      components: a,
      stencilPackageName: e,
      customElementsDir: u,
      transformTag: i
    });
    if (n[m] = b, i) {
      const w = j.join(s, "tag-transformer.ts");
      n[w] = G({ stencilPackageName: e, customElementsDir: u });
    }
    if (g) {
      const w = j.join(s, `${t}.server.ts`), E = x({
        components: a.filter(
          (v) => !C || !C.includes(v.tagName)
        ),
        stencilPackageName: e,
        customElementsDir: u,
        hydrateModule: g,
        clientModule: y,
        serializeShadowRoot: h,
        transformTag: i
      });
      n[w] = E;
    }
  }
  if (l) {
    for (const t of p)
      T([t], t.tagName);
    const a = await q({ components: p, project: f, outDir: s });
    c.push(a);
  } else
    T(p);
  return await Promise.all(
    Object.entries(n).map(async ([a, t]) => {
      const m = f.createSourceFile(a, t, { overwrite: !0 });
      await m.save(), c.push(m);
    })
  ), c;
}, $ = "react-output-target", H = "dist/components", _ = "dist-custom-elements", I = "dist-hydrate-script", Q = ({
  outDir: e,
  esModules: r,
  stencilPackageName: s,
  excludeComponents: l,
  customElementsDir: u,
  hydrateModule: d,
  clientModule: f,
  excludeServerSideRenderingFor: g,
  serializeShadowRoot: y,
  transformTag: C
}) => {
  let h = H;
  return {
    type: "custom",
    name: $,
    validate(i) {
      if (u)
        h = u;
      else {
        const c = (i.outputTargets || []).find(
          (p) => p.type === _
        );
        if (c == null)
          throw new Error(
            `The '${$}' requires '${_}' output target. Add { type: '${_}' }, to the outputTargets config.`
          );
        if (c.dir !== void 0 && (h = c.dir), c.externalRuntime !== !1)
          throw new Error(
            `The '${$}' requires the '${_}' output target to have 'externalRuntime: false' set in its configuration.`
          );
      }
      if (d) {
        if ((i.outputTargets || []).find((p) => p.type === I) == null)
          throw new Error(
            `The '${$}' requires '${I}' output target when the 'hydrateModule' option is set. Add { type: '${I}' }, to the outputTargets config.`
          );
        if (f == null)
          throw new Error(
            `The '${$}' requires the 'clientModule' option when the 'hydrateModule' option is set. Please provide the clientModule manually to the ${$} output target.`
          );
      }
      if (s === void 0) {
        if (i.sys && i.packageJsonFilePath) {
          const { name: c } = JSON.parse(i.sys.readFileSync(i.packageJsonFilePath, "utf8"));
          s = c;
        }
        if (!s)
          throw new Error(
            `Unable to find the package name in the package.json file: ${i.packageJsonFilePath}. Please provide the stencilPackageName manually to the ${$} output target.`
          );
      }
    },
    async generator(i, c, p) {
      const n = p.createTimeSpan(`generate ${$} started`, !0), T = p.components, a = new R(), t = await K({
        outDir: e,
        components: T,
        stencilPackageName: s,
        customElementsDir: h,
        excludeComponents: l,
        esModules: r === !0,
        project: a,
        hydrateModule: d,
        clientModule: f,
        excludeServerSideRenderingFor: g,
        serializeShadowRoot: y,
        transformTag: C
      });
      await Promise.all(
        t.map((m) => c.fs.writeFile(m.getFilePath(), m.getFullText()))
      ), n.finish(`generate ${$} finished`);
    },
    __internal_getCustomElementsDir() {
      return h;
    }
  };
};
export {
  Q as reactOutputTarget
};
