// @siemens/ix/dist/collection/components/date-dropdown/date-dropdown.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { iconHistory } from "@siemens/ix-icons/icons";
import { Fragment, h, Host, } from "@stencil/core";
import { DateTime } from "luxon";
import { makeRef } from "../utils/make-ref";
export class DateDropdown {
    constructor() {
        /**
         * Disable the button that opens the dropdown containing the date picker.
         */
        this.disabled = false;
        /**
         * Date format string.
         * See {@link https://moment.github.io/luxon/#/formatting?id=table-of-tokens} for all available tokens.
         */
        this.format = 'yyyy/LL/dd';
        /**
         * If true disables date range selection (from/to).
         */
        this.singleSelection = false;
        /**
         * Picker date. If the picker is in range mode this property is the start date.
         * If set to `null` no default start date will be pre-selected.
         *
         * Format is based on `format`
         */
        this.from = '';
        /**
         * Picker date. If the picker is in range mode this property is the end date.
         * If the picker is not in range mode leave this value `null`
         *
         * Format is based on `format`
         */
        this.to = '';
        /**
         * The earliest date that can be selected by the date picker.
         * If not set there will be no restriction.
         */
        this.minDate = '';
        /**
         * The latest date that can be selected by the date picker.
         * If not set there will be no restriction.
         */
        this.maxDate = '';
        /**
         * Used to set the initial select date range as well as the button name,
         * if not set or no according date range label is found, nothing will be selected
         */
        this.dateRangeId = 'custom';
        /**
         * Button variant
         */
        this.variant = 'primary';
        /**
         * Loading button
         */
        this.loading = false;
        /**
         * Shows week numbers displayed on the left side of the date picker
         *
         * @since 3.0.0
         */
        this.showWeekNumbers = false;
        /**
         * Controls whether custom date range selection is disabled in the component.
         * When set to 'false', the user can select a custom date range using the date picker.
         * When set to 'true', only predefined time date ranges are available for selection.
         */
        this.customRangeDisabled = false;
        /**
         * An array of predefined date range options for the date picker.
         * Each option is an object with a label describing the range and a function
         * that returns the start and end dates of the range as a DateRangeOption object.
         *
         * Example format:
         *   {
         *     id: 'some unique id',
         *     label: 'Name of the range',
         *     from: undefined, to: '2023/03/29'
         *   },
         *   // ... other predefined date range options ...
         */
        this.dateRangeOptions = [];
        /**
         * The index of which day to start the week on, based on the Locale#weekdays array.
         * E.g. if the locale is en-us, weekStartIndex = 1 results in starting the week on monday.
         */
        this.weekStartIndex = 0;
        /**
         * Text for custom dropdown item. Will be used for translation.
         */
        this.i18nCustomItem = 'Custom...';
        /**
         * Text for the done button. Will be used for translation.
         */
        this.i18nDone = 'Done';
        /**
         * Text for the done button. Will be used for translation.
         */
        this.i18nNoRange = 'No range set';
        /** @internal */
        this.today = DateTime.now().toISO();
        /**
         * Enable Popover API rendering for dropdown.
         *
         * @default false
         * @since 4.3.0
         */
        this.enableTopLayer = false;
        this.selectedDateRangeId = '';
        this.triggerRef = makeRef();
        this.datePickerTouched = false;
    }
    onDateRangeIdChange() {
        this.onRangeListSelect(this.dateRangeId);
        this.updateCurrentDate();
        this.setDateRangeSelection(this.dateRangeId);
        if (!this.currentRangeValue) {
            return;
        }
        this.onDateSelect({
            from: this.currentRangeValue.from,
            to: this.currentRangeValue.to,
            id: this.currentRangeValue.id,
        });
    }
    onDateRangeOptionsChange() {
        this.initialize();
        this.onDateRangeIdChange();
    }
    onDisabledChange() {
        if (this.disabled) {
            this.closeDropdown();
        }
    }
    componentWillLoad() {
        this.initialize();
        this.setDateRangeSelection(this.dateRangeId);
    }
    /**
     * Retrieves the currently selected date range from the component.
     * This method returns the selected date range as a `DateChangeEvent` object.
     */
    async getDateRange() {
        return this.currentRangeValue || { id: '', from: '', to: '' };
    }
    initialize() {
        const isCustomRange = this.dateRangeId === 'custom' ||
            (!this.dateRangeId && !!this.from && !!this.to);
        if (isCustomRange && !this.customRangeDisabled) {
            this.selectedDateRangeId = 'custom';
            this.updateCurrentDate();
            return;
        }
        const isValidConfiguration = !isCustomRange && !this.from;
        if (!isValidConfiguration) {
            console.warn('"from" and "range-date-id" is provided this is an invalid combination. Using "custom".');
            this.selectedDateRangeId = 'custom';
            this.updateCurrentDate();
            return;
        }
    }
    updateCurrentDate() {
        this.currentRangeValue = {
            id: this.selectedDateRangeId,
            from: this.from,
            to: this.to,
        };
    }
    onDateSelect(rangeValue, preserveDropdown = true) {
        this.dateRangeChange.emit(rangeValue);
        if (preserveDropdown) {
            this.closeDropdown();
        }
        this.datePickerTouched = false;
    }
    onRangeListSelect(id) {
        if (this.setDateRangeSelection(id) && this.currentRangeValue) {
            this.onDateSelect(this.currentRangeValue);
        }
    }
    setDateRangeSelection(id) {
        this.selectedDateRangeId = id;
        const option = this.dateRangeOptions.find((range) => range.id === id);
        if (option) {
            this.currentRangeValue = option;
        }
        return option;
    }
    closeDropdown() {
        const dropdown = this.hostElement.shadowRoot.querySelector('ix-dropdown');
        if (dropdown) {
            dropdown.show = false;
        }
    }
    getButtonLabel() {
        var _a, _b;
        if (this.selectedDateRangeId === 'custom' && ((_a = this.currentRangeValue) === null || _a === void 0 ? void 0 : _a.from)) {
            let range = this.currentRangeValue.from;
            if (this.currentRangeValue.to) {
                range += ` - ${this.currentRangeValue.to}`;
            }
            return range;
        }
        if (!this.selectedDateRangeId) {
            return this.i18nNoRange;
        }
        if (!this.dateRangeOptions || ((_b = this.dateRangeOptions) === null || _b === void 0 ? void 0 : _b.length) === 0) {
            return this.i18nNoRange;
        }
        const option = this.dateRangeOptions.find((option) => option.id === this.selectedDateRangeId);
        if (!option) {
            console.error(`Cannot find range option with id ${this.selectedDateRangeId}`);
            return this.i18nNoRange;
        }
        return option.label;
    }
    render() {
        var _a, _b, _c;
        return (h(Host, { key: '34b878eb8221b6c13f235d1b42ec63d3b6dce25c' }, h("ix-button", { key: 'c40cff0539f79b19c0af903edce79de7d52e704e', "data-testid": "date-dropdown-trigger", "data-date-dropdown-trigger": true, variant: this.variant, loading: this.loading, icon: iconHistory, ref: this.triggerRef, disabled: this.disabled, ariaLabel: this.ariaLabelDropdownButton }, this.getButtonLabel()), h("ix-dropdown", { key: '3a74d5b8d2f4e7c7487b0bac9af67c3d25d6b75d', "data-testid": "date-dropdown", "data-date-dropdown": true, class: "min-width max-height", trigger: this.triggerRef.waitForCurrent(), closeBehavior: "outside", placement: "bottom-start", enableTopLayer: this.enableTopLayer, onShowChanged: ({ detail: show }) => {
                if (!show &&
                    this.selectedDateRangeId === 'custom' &&
                    this.datePickerTouched &&
                    this.currentRangeValue) {
                    this.onDateSelect(this.currentRangeValue);
                }
            } }, h("ix-layout-grid", { key: '0aeda647289d651805bdd7eaac3528b5ee6b649b', noMargin: true }, h("ix-row", { key: '43bed15b884d437de29d9879be99e1cc4b48d41a' }, ((_a = this.dateRangeOptions) === null || _a === void 0 ? void 0 : _a.length) > 1 && (h("ix-col", { key: '3beb46a324a3d12bcb13be28956d7219f708b528', class: {
                'no-margin': true,
                'border-right': this.selectedDateRangeId === 'custom',
            } }, this.dateRangeOptions.map((dateRangeOption) => (h("ix-dropdown-item", { label: dateRangeOption.label, onClick: () => this.onRangeListSelect(dateRangeOption.id), checked: this.selectedDateRangeId === dateRangeOption.id }))), h("div", { key: 'f0e7910986569a41b7d0ca0419361e7c9e81be2d', hidden: this.customRangeDisabled }, h("ix-dropdown-item", { key: 'c1af0808c290045824cbdba19dc72d46d23e2d66', label: this.i18nCustomItem, checked: this.selectedDateRangeId === 'custom', onClick: () => this.onRangeListSelect('custom') })))), h("ix-col", { key: 'b3b1b70c86b309f2c8572751a30007ef1da9e9f3', class: "no-margin" }, this.selectedDateRangeId === 'custom' && (h(Fragment, { key: '70e07f3dec874213296af70b8ca63e631a247541' }, h("ix-date-picker", { key: '386b034422347098aff8b09255e83368f0d8d106', embedded: true, locale: this.locale, onDateChange: (e) => {
                e.stopPropagation();
                this.currentRangeValue = Object.assign(Object.assign({}, e.detail), { id: 'custom' });
                this.datePickerTouched = true;
            }, onDateRangeChange: (e) => e.stopPropagation(), format: this.format, singleSelection: this.singleSelection, from: this.from || ((_b = this.currentRangeValue) === null || _b === void 0 ? void 0 : _b.from), to: this.to || ((_c = this.currentRangeValue) === null || _c === void 0 ? void 0 : _c.to), minDate: this.minDate, maxDate: this.maxDate, today: this.today, weekStartIndex: this.weekStartIndex, showWeekNumbers: this.showWeekNumbers }), h("div", { key: '3f3f115659920a42b192c9f10cdd81af5a0c04ae', class: "pull-right" }, h("ix-button", { key: '2fb0e8b8d6de6244a1d9d60d175226a24c2103ad', onClick: () => {
                if (this.currentRangeValue) {
                    this.onDateSelect(this.currentRangeValue);
                }
            } }, this.i18nDone))))))))));
    }
    static get is() { return "ix-date-dropdown"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["date-dropdown.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["date-dropdown.css"]
        };
    }
    static get properties() {
        return {
            "disabled": {
                "type": "boolean",
                "attribute": "disabled",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Disable the button that opens the dropdown containing the date picker."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "format": {
                "type": "string",
                "attribute": "format",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Date format string.\nSee {@link https://moment.github.io/luxon/#/formatting?id=table-of-tokens} for all available tokens."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'yyyy/LL/dd'"
            },
            "singleSelection": {
                "type": "boolean",
                "attribute": "single-selection",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "If true disables date range selection (from/to)."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "from": {
                "type": "string",
                "attribute": "from",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Picker date. If the picker is in range mode this property is the start date.\nIf set to `null` no default start date will be pre-selected.\n\nFormat is based on `format`"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "to": {
                "type": "string",
                "attribute": "to",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Picker date. If the picker is in range mode this property is the end date.\nIf the picker is not in range mode leave this value `null`\n\nFormat is based on `format`"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "minDate": {
                "type": "string",
                "attribute": "min-date",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The earliest date that can be selected by the date picker.\nIf not set there will be no restriction."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "maxDate": {
                "type": "string",
                "attribute": "max-date",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The latest date that can be selected by the date picker.\nIf not set there will be no restriction."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "dateRangeId": {
                "type": "string",
                "attribute": "date-range-id",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Used to set the initial select date range as well as the button name,\nif not set or no according date range label is found, nothing will be selected"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'custom'"
            },
            "variant": {
                "type": "string",
                "attribute": "variant",
                "mutable": false,
                "complexType": {
                    "original": "ButtonVariant",
                    "resolved": "\"danger-primary\" | \"danger-secondary\" | \"danger-tertiary\" | \"primary\" | \"secondary\" | \"subtle-primary\" | \"subtle-secondary\" | \"subtle-tertiary\" | \"tertiary\"",
                    "references": {
                        "ButtonVariant": {
                            "location": "import",
                            "path": "../button/button",
                            "id": "src/components/button/button.tsx::ButtonVariant"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Button variant"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'primary'"
            },
            "loading": {
                "type": "boolean",
                "attribute": "loading",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Loading button"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "showWeekNumbers": {
                "type": "boolean",
                "attribute": "show-week-numbers",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.0.0"
                        }],
                    "text": "Shows week numbers displayed on the left side of the date picker"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "ariaLabelDropdownButton": {
                "type": "string",
                "attribute": "aria-label-dropdown-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "ARIA label for the dropdown\nWill be set as aria-label on the nested HTML button element that will trigger the dropdown"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "customRangeDisabled": {
                "type": "boolean",
                "attribute": "custom-range-disabled",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Controls whether custom date range selection is disabled in the component.\nWhen set to 'false', the user can select a custom date range using the date picker.\nWhen set to 'true', only predefined time date ranges are available for selection."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "dateRangeOptions": {
                "type": "unknown",
                "attribute": "date-range-options",
                "mutable": false,
                "complexType": {
                    "original": "DateDropdownOption[]",
                    "resolved": "DateDropdownOption[]",
                    "references": {
                        "DateDropdownOption": {
                            "location": "import",
                            "path": "./date-dropdown.types",
                            "id": "src/components/date-dropdown/date-dropdown.types.ts::DateDropdownOption"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "An array of predefined date range options for the date picker.\nEach option is an object with a label describing the range and a function\nthat returns the start and end dates of the range as a DateRangeOption object.\n\nExample format:\n  {\n    id: 'some unique id',\n    label: 'Name of the range',\n    from: undefined, to: '2023/03/29'\n  },\n  // ... other predefined date range options ..."
                },
                "getter": false,
                "setter": false,
                "defaultValue": "[]"
            },
            "locale": {
                "type": "string",
                "attribute": "locale",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Locale identifier (e.g. 'en' or 'de')."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "weekStartIndex": {
                "type": "number",
                "attribute": "week-start-index",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The index of which day to start the week on, based on the Locale#weekdays array.\nE.g. if the locale is en-us, weekStartIndex = 1 results in starting the week on monday."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "0"
            },
            "i18nCustomItem": {
                "type": "string",
                "attribute": "i18n-custom-item",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Text for custom dropdown item. Will be used for translation."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'Custom...'"
            },
            "i18nDone": {
                "type": "string",
                "attribute": "i18n-done",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Text for the done button. Will be used for translation."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'Done'"
            },
            "i18nNoRange": {
                "type": "string",
                "attribute": "i18n-no-range",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Text for the done button. Will be used for translation."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'No range set'"
            },
            "today": {
                "type": "string",
                "attribute": "today",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": ""
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "DateTime.now().toISO()"
            },
            "enableTopLayer": {
                "type": "boolean",
                "attribute": "enable-top-layer",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "default",
                            "text": "false"
                        }, {
                            "name": "since",
                            "text": "4.3.0"
                        }],
                    "text": "Enable Popover API rendering for dropdown."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get states() {
        return {
            "selectedDateRangeId": {},
            "currentRangeValue": {}
        };
    }
    static get events() {
        return [{
                "method": "dateRangeChange",
                "name": "dateRangeChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "EventEmitter for date range change events.\n\nThis event is emitted when the date range changes within the component.\nThe event payload contains information about the selected date range."
                },
                "complexType": {
                    "original": "DateRangeChangeEvent",
                    "resolved": "{ id: string; from?: string | undefined; to?: string | undefined; }",
                    "references": {
                        "DateRangeChangeEvent": {
                            "location": "import",
                            "path": "./date-dropdown.types",
                            "id": "src/components/date-dropdown/date-dropdown.types.ts::DateRangeChangeEvent"
                        }
                    }
                }
            }];
    }
    static get methods() {
        return {
            "getDateRange": {
                "complexType": {
                    "signature": "() => Promise<DateRangeChangeEvent>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "DateRangeChangeEvent": {
                            "location": "import",
                            "path": "./date-dropdown.types",
                            "id": "src/components/date-dropdown/date-dropdown.types.ts::DateRangeChangeEvent"
                        }
                    },
                    "return": "Promise<DateRangeChangeEvent>"
                },
                "docs": {
                    "text": "Retrieves the currently selected date range from the component.\nThis method returns the selected date range as a `DateChangeEvent` object.",
                    "tags": []
                }
            }
        };
    }
    static get elementRef() { return "hostElement"; }
    static get watchers() {
        return [{
                "propName": "dateRangeId",
                "methodName": "onDateRangeIdChange"
            }, {
                "propName": "to",
                "methodName": "onDateRangeIdChange"
            }, {
                "propName": "from",
                "methodName": "onDateRangeIdChange"
            }, {
                "propName": "dateRangeOptions",
                "methodName": "onDateRangeOptionsChange"
            }, {
                "propName": "disabled",
                "methodName": "onDisabledChange"
            }];
    }
}
