// @siemens/ix@2.2.3/dist/collection/components/tile/tile.js
/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host } from "@stencil/core";
export class Tile {
    constructor() {
        this.size = 'medium';
        this.hasHeaderSlot = false;
        this.hasFooterSlot = false;
    }
    handleHeaderSlotChange() {
        this.hasHeaderSlot = !!this.hostElement.querySelector('[slot="header"]');
    }
    handleFooterSlotChange() {
        this.hasFooterSlot = !!this.hostElement.querySelector('[slot="footer"]');
    }
    render() {
        return (h(Host, { key: 'ac3f0fca5c25f7bb7224ef248cca4ee6e6294d85', class: {
                'tile-small': this.size === 'small',
                'tile-medium': this.size === 'medium',
                'tile-big': this.size === 'big',
            } }, h("div", { key: 'c1d65c385b98420f4bdd6d742058200b14634c7b', class: {
                'tile-header': true,
                'has-content': this.hasHeaderSlot,
            } }, h("slot", { key: '38a70fedd15e29e0c238e94bd60f69b31dbf7d2a', name: "header", onSlotchange: () => this.handleHeaderSlotChange() })), h("div", { key: '25d69652c0315f0a434487d523cfbae1d93fda6b', class: "tile-subheader" }, h("slot", { key: '514cbc260eab41c4d400dcbf985a3be3290fb3cd', name: "subheader" })), h("div", { key: '7f4f75b86e8afd4d3e0aa80926d7f7190ca8fa61', class: "tile-content" }, h("slot", { key: 'ae9be816cda00d9b1365403dbce6e254e8c3b808' })), h("div", { key: 'cc46d36240e64c2717aa89c0e7ba7a787d225096', class: {
                'tile-footer': true,
                'has-content': this.hasFooterSlot,
            } }, h("slot", { key: '740503c075e46accb41952baf68f1e1b41dea0fd', name: "footer", onSlotchange: () => this.handleFooterSlotChange() }))));
    }
    static get is() { return "ix-tile"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tile.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tile.css"]
        };
    }
    static get properties() {
        return {
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'small' | 'medium' | 'big'",
                    "resolved": "\"big\" | \"medium\" | \"small\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Size of the tile - one of 'small', 'medium' or 'large'"
                },
                "attribute": "size",
                "reflect": false,
                "defaultValue": "'medium'"
            }
        };
    }
    static get states() {
        return {
            "hasHeaderSlot": {},
            "hasFooterSlot": {}
        };
    }
    static get elementRef() { return "hostElement"; }
}
