// @stencil/react-output-target@4.3.0/dist/ssr.js
import c from "react";
import { c as b } from "./create-component-CXb2e9Nm.js";
var M = {
  animationIterationCount: 1,
  aspectRatio: 1,
  borderImageOutset: 1,
  borderImageSlice: 1,
  borderImageWidth: 1,
  boxFlex: 1,
  boxFlexGroup: 1,
  boxOrdinalGroup: 1,
  columnCount: 1,
  columns: 1,
  flex: 1,
  flexGrow: 1,
  flexPositive: 1,
  flexShrink: 1,
  flexNegative: 1,
  flexOrder: 1,
  gridRow: 1,
  gridRowEnd: 1,
  gridRowSpan: 1,
  gridRowStart: 1,
  gridColumn: 1,
  gridColumnEnd: 1,
  gridColumnSpan: 1,
  gridColumnStart: 1,
  msGridRow: 1,
  msGridRowSpan: 1,
  msGridColumn: 1,
  msGridColumnSpan: 1,
  fontWeight: 1,
  lineHeight: 1,
  opacity: 1,
  order: 1,
  orphans: 1,
  scale: 1,
  tabSize: 1,
  widows: 1,
  zIndex: 1,
  zoom: 1,
  WebkitLineClamp: 1,
  // SVG-related properties
  fillOpacity: 1,
  floodOpacity: 1,
  stopOpacity: 1,
  strokeDasharray: 1,
  strokeDashoffset: 1,
  strokeMiterlimit: 1,
  strokeOpacity: 1,
  strokeWidth: 1
}, w = "px", I = (e) => typeof e == "number" || typeof e == "string";
function A(e) {
  return e.replace(/[A-Z]/g, (t) => `-${t.toLowerCase()}`);
}
function H(e, t, r = w) {
  if (typeof t != "string" && typeof t != "number")
    throw new Error(
      "Invalid input: value of 'cssProperties' must be string or number."
    );
  const n = M[e] === 1;
  if (typeof t == "string" || t === 0 || n)
    return `${t}`;
  const i = (typeof r == "string" ? r : r[e]) || w;
  return `${t}${i}`;
}
function j(e, t) {
  if (typeof e != "object" || e === null)
    throw new TypeError(
      `[stringifyStyleDeclaration]: Expected 'styleDeclaration' to be a non-null object, but received ${e} (type:${typeof e}).`
    );
  const r = t != null && t.important ? "!important" : "";
  return Object.entries(e).filter(([n, i]) => I(i)).map(
    ([n, i]) => `${A(n)}:${H(
      n,
      i,
      t == null ? void 0 : t.unit
    )}${r};`
  ).join("");
}
function U(e, t = !1) {
  if (typeof e != "object" || e === null)
    throw new TypeError(
      `[stringifyCSSProperties]: Expected 'cssProperties' to be a non-null object, but received ${e} (type:${typeof e}).`
    );
  return j(e, typeof t == "boolean" ? {
    important: t
  } : t);
}
const F = {
  acceptCharset: "accept-charset",
  accessKey: "accesskey",
  allowFullScreen: "allowfullscreen",
  autoCapitalize: "autocapitalize",
  autoComplete: "autocomplete",
  autoCorrect: "autocorrect",
  autoFocus: "autofocus",
  autoPlay: "autoplay",
  autoSave: "autosave",
  cellPadding: "cellpadding",
  cellSpacing: "cellspacing",
  charSet: "charset",
  className: "class",
  classID: "classid",
  colSpan: "colspan",
  contentEditable: "contenteditable",
  contextMenu: "contextmenu",
  controlsList: "controlslist",
  crossOrigin: "crossorigin",
  dangerouslySetInnerHTML: "dangerouslysetinnerhtml",
  dateTime: "datetime",
  defaultChecked: "defaultchecked",
  defaultValue: "defaultvalue",
  disablePictureInPicture: "disablepictureinpicture",
  disableRemotePlayback: "disableremoteplayback",
  encType: "enctype",
  enterKeyHint: "enterkeyhint",
  fetchPriority: "fetchpriority",
  htmlFor: "for",
  formMethod: "formmethod",
  formAction: "formaction",
  formEncType: "formenctype",
  formNoValidate: "formnovalidate",
  formTarget: "formtarget",
  frameBorder: "frameborder",
  hrefLang: "hreflang",
  httpEquiv: "http-equiv",
  imageSizes: "imagesizes",
  imageSrcSet: "imagesrcset",
  inputMode: "inputmode",
  itemID: "itemid",
  itemProp: "itemprop",
  itemRef: "itemref",
  itemScope: "itemscope",
  itemType: "itemtype",
  keyParams: "keyparams",
  keyType: "keytype",
  marginWidth: "marginwidth",
  marginHeight: "marginheight",
  maxLength: "maxlength",
  mediaGroup: "mediagroup",
  noModule: "nomodule",
  noValidate: "novalidate",
  playsInline: "playsinline",
  radioGroup: "radiogroup",
  readOnly: "readonly",
  referrerPolicy: "referrerpolicy",
  rowSpan: "rowspan",
  spellCheck: "spellcheck",
  srcDoc: "srcdoc",
  srcLang: "srclang",
  srcSet: "srcset",
  tabIndex: "tabindex",
  useMap: "usemap",
  accentHeight: "accent-height",
  alignmentBaseline: "alignment-baseline",
  arabicForm: "arabic-form",
  attributeName: "attributename",
  attributeType: "attributetype",
  autoReverse: "autoreverse",
  baseFrequency: "basefrequency",
  baselineShift: "baseline-shift",
  baseProfile: "baseprofile",
  calcMode: "calcmode",
  capHeight: "cap-height",
  clipPath: "clip-path",
  clipPathUnits: "clippathunits",
  clipRule: "clip-rule",
  colorInterpolation: "color-interpolation",
  colorInterpolationFilters: "color-interpolation-filters",
  colorProfile: "color-profile",
  colorRendering: "color-rendering",
  contentScriptType: "contentscripttype",
  contentStyleType: "contentstyletype",
  diffuseConstant: "diffuseconstant",
  dominantBaseline: "dominant-baseline",
  enableBackground: "enable-background",
  externalResourcesRequired: "externalresourcesrequired",
  fillOpacity: "fill-opacity",
  fillRule: "fill-rule",
  filterRes: "filterres",
  filterUnits: "filterunits",
  floodOpacity: "flood-opacity",
  floodColor: "flood-color",
  fontFamily: "font-family",
  fontSize: "font-size",
  fontSizeAdjust: "font-size-adjust",
  fontStretch: "font-stretch",
  fontStyle: "font-style",
  fontVariant: "font-variant",
  fontWeight: "font-weight",
  glyphName: "glyph-name",
  glyphOrientationHorizontal: "glyph-orientation-horizontal",
  glyphOrientationVertical: "glyph-orientation-vertical",
  gradientTransform: "gradienttransform",
  gradientUnits: "gradientunits",
  horizAdvX: "horiz-adv-x",
  horizOriginX: "horiz-origin-x",
  imageRendering: "image-rendering",
  kernelMatrix: "kernelmatrix",
  kernelUnitLength: "kernelunitlength",
  keyPoints: "keypoints",
  keySplines: "keysplines",
  keyTimes: "keytimes",
  lengthAdjust: "lengthadjust",
  letterSpacing: "letter-spacing",
  lightingColor: "lighting-color",
  limitingConeAngle: "limitingconeangle",
  markerEnd: "marker-end",
  markerHeight: "markerheight",
  markerMid: "marker-mid",
  markerStart: "marker-start",
  markerUnits: "markerunits",
  markerWidth: "markerwidth",
  maskContentUnits: "maskcontentunits",
  maskUnits: "maskunits",
  numOctaves: "numoctaves",
  overlinePosition: "overline-position",
  overlineThickness: "overline-thickness",
  paintOrder: "paint-order",
  panose1: "panose-1",
  pathLength: "pathlength",
  patternContentUnits: "patterncontentunits",
  patternTransform: "patterntransform",
  patternUnits: "patternunits",
  pointerEvents: "pointer-events",
  pointsAtX: "pointsatx",
  pointsAtY: "pointsaty",
  pointsAtZ: "pointsatz",
  preserveAlpha: "preservealpha",
  preserveAspectRatio: "preserveaspectratio",
  primitiveUnits: "primitiveunits",
  refX: "refx",
  refY: "refy",
  renderingIntent: "rendering-intent",
  repeatCount: "repeatcount",
  repeatDur: "repeatdur",
  requiredExtensions: "requiredextensions",
  requiredFeatures: "requiredfeatures",
  shapeRendering: "shape-rendering",
  slot: "slot",
  specularConstant: "specularconstant",
  specularExponent: "specularexponent",
  spreadMethod: "spreadmethod",
  startOffset: "startoffset",
  stdDeviation: "stddeviation",
  stitchTiles: "stitchtiles",
  stopColor: "stop-color",
  stopOpacity: "stop-opacity",
  strikethroughPosition: "strikethrough-position",
  strikethroughThickness: "strikethrough-thickness",
  strokeDasharray: "stroke-dasharray",
  strokeDashoffset: "stroke-dashoffset",
  strokeLinecap: "stroke-linecap",
  strokeLinejoin: "stroke-linejoin",
  strokeMiterlimit: "stroke-miterlimit",
  strokeWidth: "stroke-width",
  strokeOpacity: "stroke-opacity",
  suppressContentEditableWarning: "suppresscontenteditablewarning",
  suppressHydrationWarning: "suppresshydrationwarning",
  surfaceScale: "surfacescale",
  systemLanguage: "systemlanguage",
  tableValues: "tablevalues",
  targetX: "targetx",
  targetY: "targety",
  textAnchor: "text-anchor",
  textDecoration: "text-decoration",
  textLength: "textlength",
  textRendering: "text-rendering",
  transformOrigin: "transform-origin",
  underlinePosition: "underline-position",
  underlineThickness: "underline-thickness",
  unicodeBidi: "unicode-bidi",
  unicodeRange: "unicode-range",
  unitsPerEm: "units-per-em",
  vectorEffect: "vector-effect",
  vertAdvY: "vert-adv-y",
  vertOriginX: "vert-origin-x",
  vertOriginY: "vert-origin-y",
  wordSpacing: "word-spacing",
  writingMode: "writing-mode",
  xChannelSelector: "xchannelselector",
  xHeight: "x-height",
  xlinkActuate: "xlink:actuate",
  xlinkArcrole: "xlink:arcrole",
  xlinkHref: "xlink:href",
  xlinkRole: "xlink:role",
  xlinkShow: "xlink:show",
  xlinkTitle: "xlink:title",
  xlinkType: "xlink:type",
  xmlBase: "xml:base",
  xmlLang: "xml:lang",
  xmlnsXlink: "xmlns:xlink",
  yChannelSelector: "ychannelselector",
  zoomAndPan: "zoomandpan"
}, _ = "[react-output-target]", m = (e) => typeof e == "string" || typeof e == "number" || typeof e == "boolean", g = (e) => e == null, T = (e) => Array.isArray(e), N = (e) => !!e && /^\s*class\s+/.test(e.toString()), W = (e) => !!e && typeof e == "object" && "_payload" in e, D = (e) => async ({ children: t, ...r } = {}) => {
  if (!("process" in globalThis) || typeof window < "u")
    throw new Error("`createComponentForServerSideRendering` can only be run on the server");
  let n = "";
  for (const [a, o] of Object.entries(r)) {
    if (typeof o == "boolean" && o === !1)
      continue;
    let s = m(o) ? `"${o}"` : typeof o != "function" ? e.serializeProperty(o) : void 0;
    if (a === "style" && typeof o == "object" && o && (s = `"${U(o)}"`), !s)
      continue;
    const d = F[a] || e.properties[a] || a;
    n += ` ${d}=${s}`;
  }
  let i = "";
  const p = e.transformTag ? e.transformTag(e.tagName) : e.tagName, k = `<${p}${n} suppressHydrationWarning="true">`, v = console.error;
  try {
    process.env.STENCIL_SSR_DEBUG || (console.error = () => {
    });
    const a = await y(t), { renderToString: o } = await import("react-dom/server");
    i = o(a);
  } catch (a) {
    if (process.env.STENCIL_SSR_DEBUG) {
      const o = a instanceof Error ? a : new Error("Unknown error");
      console.warn(
        `${_} Failed to serialize light DOM for ${k.slice(0, -1)} />: ${o.message} - this may impact the hydration of the component`
      );
    }
  } finally {
    console.error = v;
  }
  const C = `${k}${i}</${p}>`, { html: u } = await e.renderToString(C, {
    fullDocument: !1,
    serializeShadowRoot: e.serializeShadowRoot ?? "declarative-shadow-dom",
    prettyHtml: !0
  });
  if (!u)
    throw new Error("No HTML returned from renderToString");
  const l = u.split(`
`), E = "<!--r.1-->", x = l[1].includes('shadowrootmode="open"'), z = l[1].includes("shadowrootdelegatesfocus");
  let S;
  x && (S = l.slice(2, l.lastIndexOf("  </template>")).join(`
`).trim());
  const { default: P } = await import("html-react-parser"), R = P, O = () => R(u, {
    transform(a, o) {
      if ("name" in o && (o.name === p || o.name === e.tagName)) {
        const s = a.props, d = `${p}`;
        if (!x) {
          const { children: G, ...$ } = s || {}, L = l.slice(1, -1).join(`
`).trim().replace(new RegExp("(?<=>)\\s+(?=<)", "g"), "");
          return /* @__PURE__ */ c.createElement(d, { ...$, suppressHydrationWarning: !0, dangerouslySetInnerHTML: { __html: L } });
        }
        return /* @__PURE__ */ c.createElement(d, { ...s, suppressHydrationWarning: !0 }, /* @__PURE__ */ c.createElement(
          "template",
          {
            shadowrootmode: "open",
            ...z ? { shadowrootdelegatesfocus: "" } : {},
            suppressHydrationWarning: !0,
            dangerouslySetInnerHTML: { __html: E + S }
          }
        ), t);
      }
    }
  });
  return /* @__PURE__ */ c.createElement(O, null);
};
async function y(e) {
  return m(e) || g(e) ? e : (T(e) || (e = [e]), Promise.all(
    Array.from(e).map(async (t) => {
      if (m(t) || g(t))
        return t;
      if (T(t))
        return y(t);
      if (!c.isValidElement(t))
        return t;
      const { type: r, props: n } = t;
      return {
        ...t,
        props: {
          ...n,
          children: await y(n.children)
        },
        type: await h(r, n)
      };
    })
  ));
}
const h = async (e, t) => {
  let r = null;
  if (typeof e == "string")
    return e;
  if (N(e)) {
    const n = new e(t);
    r = n.render ? n.render() : n;
  } else if (W(e)) {
    const n = e._payload, { deault: i } = n._status === -1 ? await n._result() : n._result;
    r = await h(i, t);
  } else typeof e != "object" && (r = await e(t));
  return !g(r) && !m(r) && typeof r == "object" && r !== null && "type" in r && (r = await h(r.type, t)), r;
};
let f = null;
const V = (e) => {
  if (typeof window < "u") {
    if (e.clientModule)
      return e.clientModule;
    if (b)
      return b({
        tagName: e.tagName,
        properties: e.properties
      });
  }
  return async (t) => {
    let r = !1;
    f || (f = await e.hydrateModule, r = !0);
    const n = f;
    if (e.getTagTransformer && r) {
      const i = e.getTagTransformer();
      i && n.setTagTransformer(i);
    }
    return e.transformTag = n.transformTag, D({
      renderToString: n.renderToString,
      serializeProperty: n.serializeProperty,
      ...e
    })(t);
  };
};
export {
  V as createComponent
};
