// @siemens/ix@4.3.0/dist/collection/components/pill/pill.js
/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host } from "@stencil/core";
import { makeRef } from "../utils/make-ref";
export class Pill {
    constructor() {
        /**
         * Pill variant
         */
        this.variant = 'primary';
        /**
         * Show pill as outline
         */
        this.outline = false;
        /**
         * Align pill content left
         */
        this.alignLeft = false;
        /**
         * Display a tooltip. By default, no tooltip will be displayed.
         * Add the attribute to display the text content of the component as a tooltip or use a string to display a custom text.
         * @since 3.0.0
         */
        this.tooltipText = false;
        this.iconOnly = false;
        this.containerElementRef = makeRef();
    }
    componentWillLoad() {
        this.checkIfContentAvailable();
    }
    checkIfContentAvailable() {
        const hasChildren = this.hostElement.children.length > 0;
        const hasTextContent = !!this.hostElement.textContent;
        this.iconOnly = !hasChildren && !hasTextContent;
    }
    getTooltip() {
        var _a;
        if (!this.tooltipText && !this.hostElement.hasAttribute('tooltip-text')) {
            return null;
        }
        const text = typeof this.tooltipText === 'string' && this.tooltipText.trim()
            ? this.tooltipText
            : (_a = this.hostElement.textContent) === null || _a === void 0 ? void 0 : _a.trim();
        return (h("ix-tooltip", { for: this.containerElementRef.waitForCurrent() }, text));
    }
    render() {
        let customStyle = {};
        if (this.variant === 'custom') {
            customStyle = {
                color: this.pillColor,
                [this.outline ? 'borderColor' : 'backgroundColor']: this.background,
            };
        }
        return (h(Host, { key: '92cc7c2458f3bbf26a62fb641a6a69c279aca3d4', style: this.variant === 'custom'
                ? {
                    '--ix-icon-button-color': this.pillColor,
                }
                : {}, class: {
                'align-left': this.alignLeft,
            } }, h("div", { key: '4b6a3abf6c93b8d8705e9e95738b0e2f1b270465', ref: this.containerElementRef, style: Object.assign({}, customStyle), class: {
                container: true,
                outline: this.outline,
                inactive: false,
                alarm: this.variant === 'alarm',
                critical: this.variant === 'critical',
                info: this.variant === 'info',
                neutral: this.variant === 'neutral',
                primary: this.variant === 'primary',
                success: this.variant === 'success',
                warning: this.variant === 'warning',
                custom: this.variant === 'custom',
                closable: false,
                icon: !!this.icon,
                'with-gap': !this.iconOnly,
            } }, this.icon && (h("ix-icon", { key: '4739a22536b12b5124fc11a6f72f664539c2b662', class: {
                'with-icon': true,
            }, name: this.icon, size: '16', "aria-label": this.ariaLabelIcon })), h("span", { key: '90457e143210a0f070b775ba176aa35946e910c6', class: "slot-container" }, h("slot", { key: 'bd7e9da045b7536b65a4c92ec37197bbfda0b478', onSlotchange: () => this.checkIfContentAvailable() }))), this.getTooltip()));
    }
    static get is() { return "ix-pill"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["pill.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["pill.css"]
        };
    }
    static get properties() {
        return {
            "variant": {
                "type": "string",
                "attribute": "variant",
                "mutable": false,
                "complexType": {
                    "original": "| 'primary'\n    | 'alarm'\n    | 'critical'\n    | 'warning'\n    | 'info'\n    | 'neutral'\n    | 'success'\n    | 'custom'",
                    "resolved": "\"alarm\" | \"critical\" | \"custom\" | \"info\" | \"neutral\" | \"primary\" | \"success\" | \"warning\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Pill variant"
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "defaultValue": "'primary'"
            },
            "outline": {
                "type": "boolean",
                "attribute": "outline",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show pill as outline"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "icon": {
                "type": "string",
                "attribute": "icon",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Show icon"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "ariaLabelIcon": {
                "type": "string",
                "attribute": "aria-label-icon",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label for the icon"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "background": {
                "type": "string",
                "attribute": "background",
                "mutable": false,
                "complexType": {
                    "original": "string | undefined",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Custom color for pill. Only working for `variant='custom'`"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "pillColor": {
                "type": "string",
                "attribute": "pill-color",
                "mutable": false,
                "complexType": {
                    "original": "string | undefined",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Custom font color for pill. Only working for `variant='custom'`"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "alignLeft": {
                "type": "boolean",
                "attribute": "align-left",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Align pill content left"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "tooltipText": {
                "type": "any",
                "attribute": "tooltip-text",
                "mutable": false,
                "complexType": {
                    "original": "string | boolean",
                    "resolved": "boolean | string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.0.0"
                        }],
                    "text": "Display a tooltip. By default, no tooltip will be displayed.\nAdd the attribute to display the text content of the component as a tooltip or use a string to display a custom text."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get states() {
        return {
            "iconOnly": {}
        };
    }
    static get elementRef() { return "hostElement"; }
}
